// fontdue-js/dist/components/CookieNotification/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _store = _interopRequireDefault(require("store"));
var _ComponentsContext = _interopRequireDefault(require("../ComponentsContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const DefaultCookieNotification = _ref => {
  let {
    onDismiss,
    open,
    siteTitle
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cookie-notification",
    "data-open": open
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cookie-notification__container"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cookie-notification__text-container"
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: "cookie-notification__label"
  }, siteTitle, " website uses cookies"), /*#__PURE__*/_react.default.createElement("p", {
    className: "cookie-notification__text"
  }, "We use cookies to ensure that we give you the best experience. By continuing to visit this site you agree to receive cookies from generaltypestudio.com. You can read more about cookies in our ", /*#__PURE__*/_react.default.createElement("a", {
    href: "/privacy-policy"
  }, "privacy policy"), ".")), /*#__PURE__*/_react.default.createElement("div", {
    className: "cookie-notification__button-container"
  }, /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "cookie-notification__button",
    onClick: onDismiss
  }, "Got it"))));
};
const GetComponent = _ref2 => {
  let {
    name,
    fallback,
    children
  } = _ref2;
  return /*#__PURE__*/_react.default.createElement(_ComponentsContext.default.Consumer, null, components => {
    const Component = components[name] || fallback;
    return children(Component);
  });
};
const CookieNotificationBase = _ref3 => {
  let {
    siteTitle,
    dismiss,
    open
  } = _ref3;
  return /*#__PURE__*/_react.default.createElement(GetComponent, {
    name: "CookieNotification",
    fallback: DefaultCookieNotification,
    open: open
  }, Component => /*#__PURE__*/_react.default.createElement(Component, {
    onDismiss: dismiss,
    siteTitle: siteTitle,
    open: open
  }));
};
class CookieNotification extends _react.default.Component {
  constructor(props) {
    super(props);
    this.state = {
      dismissed: _store.default.get('userAgreedCookies') || false
    };
  }
  dismiss = () => {
    _store.default.set('userAgreedCookies', true);
    this.setState({
      dismissed: true
    });
  };
  render() {
    const {
      dismissed
    } = this.state;
    if (dismissed) return null;
    return /*#__PURE__*/_react.default.createElement(CookieNotificationBase, _extends({
      dismiss: this.dismiss,
      open: true
    }, this.props));
  }
}
exports.default = CookieNotification;
