// fontdue-js/dist/hooks/useAutofit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _resizeObserver = _interopRequireDefault(require("@react-hook/resize-observer"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const REF_SIZE = 100;

// Canvas measurement (fast path)
let sharedCanvas = null;
function getCanvasContext() {
  if (!sharedCanvas) sharedCanvas = document.createElement('canvas');
  return sharedCanvas.getContext('2d');
}

// DOM measurement (fallback for variable fonts)
let measureEl = null;
function getMeasureElement() {
  if (!measureEl) {
    measureEl = document.createElement('span');
    measureEl.style.position = 'absolute';
    measureEl.style.top = '-9999px';
    measureEl.style.left = '-9999px';
    measureEl.style.whiteSpace = 'pre';
    measureEl.style.visibility = 'hidden';
    document.body.appendChild(measureEl);
  }
  return measureEl;
}
function measureWithCanvas(text, fontFamily, fontWeight, fontStyleValue, letterSpacing) {
  const ctx = getCanvasContext();
  ctx.font = `${fontStyleValue} ${fontWeight} ${REF_SIZE}px ${fontFamily}`;
  ctx.letterSpacing = `${letterSpacing}em`;
  const width = ctx.measureText(text).width;
  ctx.letterSpacing = '0px';
  return width;
}
function measureWithDOM(text, fontFamily, fontWeight, fontStyleValue, letterSpacing, fontVariationSettings) {
  const el = getMeasureElement();
  el.style.fontFamily = fontFamily;
  el.style.fontWeight = `${fontWeight}`;
  el.style.fontStyle = fontStyleValue;
  el.style.fontSize = `${REF_SIZE}px`;
  el.style.letterSpacing = `${letterSpacing}em`;
  el.style.fontVariationSettings = fontVariationSettings;
  el.textContent = text;
  return el.getBoundingClientRect().width;
}
const useAutofit = _ref => {
  let {
    text,
    fontFamily,
    fontSize: externalFontSize,
    min = 8,
    max = 999,
    letterSpacing = 0,
    fontWeight = 400,
    fontStyle = 'normal',
    fontVariationSettings,
    enabled = true,
    padding = 0
  } = _ref;
  const ref = (0, _react.useRef)(null);
  const [fontSize, setFontSize] = (0, _react.useState)(externalFontSize);
  const [containerWidth, setContainerWidth] = (0, _react.useState)(0);
  const [fontVersion, setFontVersion] = (0, _react.useState)(0);
  const [measured, setMeasured] = (0, _react.useState)(false);
  (0, _resizeObserver.default)(ref, entry => {
    setContainerWidth(entry.contentRect.width);
  });

  // Trigger re-measurement when the font becomes available.
  // If already loaded (e.g. parent waited for font), the promise
  // resolves as a microtask and the re-measurement is near-instant.
  (0, _react.useLayoutEffect)(() => {
    if (!enabled || !fontFamily) return;
    const fontString = `${fontStyle} ${fontWeight} ${REF_SIZE}px ${fontFamily}`;
    document.fonts.load(fontString).then(() => {
      setFontVersion(v => v + 1);
    });
  }, [fontFamily, fontWeight, fontStyle, enabled]);
  (0, _react.useLayoutEffect)(() => {
    if (!enabled) return;
    const availableWidth = containerWidth - padding;
    if (availableWidth <= 0 || !text) return;

    // Split by newlines and measure the widest line, since the rendered
    // text preserves line breaks (Draft.js uses white-space: pre-wrap).
    const lines = text.split('\n');
    const measure = fontVariationSettings ? line => measureWithDOM(line, fontFamily, fontWeight, fontStyle, letterSpacing, fontVariationSettings) : line => measureWithCanvas(line, fontFamily, fontWeight, fontStyle, letterSpacing);
    const refWidth = Math.max(...lines.map(measure));
    if (refWidth <= 0) return;

    // Text width scales linearly with font size.
    // Floor to avoid sub-pixel rounding that causes text to wrap.
    const newSize = Math.min(Math.max(min, Math.floor(availableWidth / (refWidth / REF_SIZE))), max);
    setFontSize(newSize);
    if (!measured) setMeasured(true);
  }, [text, fontFamily, fontWeight, fontStyle, fontVariationSettings, letterSpacing, containerWidth, enabled, min, max, padding, fontVersion]);
  if (!enabled) return {
    ref,
    fontSize: externalFontSize,
    ready: true
  };
  return {
    ref,
    fontSize,
    ready: measured
  };
};
var _default = useAutofit;
exports.default = _default;
