// fontdue-js/dist/hooks/useRefetchOnLicenseChanges.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRefetchOnLicenseChanges = useRefetchOnLicenseChanges;
var _react = require("react");
var _relayRuntime = require("relay-runtime");
var _useLicenseAndOrderVariables = require("./useLicenseAndOrderVariables");
/**
 * Custom hook to handle refetching data when license options or order variables change.
 * Uses a two-step process to prevent UI flickering:
 * 1. Fetches new data into the store
 * 2. Updates the component with a store-only policy
 */
function useRefetchOnLicenseChanges(_ref) {
  let {
    environment,
    refetch,
    query,
    variables = {}
  } = _ref;
  const {
    licenseOptionsSpecs,
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  (0, _react.useEffect)(() => {
    let mounted = true;

    // First fetch the data into the store
    const subscription = (0, _relayRuntime.fetchQuery)(environment, query, {
      ...variables,
      licenseOptions: licenseOptionsSpecs,
      orderVariables: orderVariableSelections
    }).subscribe({
      complete: () => {
        // Only update if component is still mounted
        if (mounted) {
          // Then update from store only to prevent flickering
          refetch({
            ...variables,
            licenseOptions: licenseOptionsSpecs,
            orderVariables: orderVariableSelections
          }, {
            fetchPolicy: 'store-only'
          });
        }
      },
      error: error => {
        if (mounted) {
          console.error('Failed to refetch data:', error);
        }
      }
    });
    return () => {
      mounted = false;
      // Cancel the fetch operation if it's still pending
      subscription.unsubscribe();
    };
  }, [licenseOptionsSpecs, orderVariableSelections, environment, query, variables]);
}
