// fontdue-js/dist/components/FontFamilies/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _FontFamiliesQuery2 = _interopRequireDefault(require("../../__generated__/FontFamiliesQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _reactRelay = require("react-relay");
var _productState = require("../Root/productState");
var _Family = _interopRequireDefault(require("../Family"));
var _useLicenseAndOrderVariables = require("../../hooks/useLicenseAndOrderVariables");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function FontFamilies(_ref) {
  var _node$families;
  let {
    collectionId
  } = _ref;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const store = (0, _reactRedux.useStore)();
  (0, _react.useEffect)(() => {
    // populate as soon as BuyButton mounts. this prepares the StoreModal
    // to be able to apply discounts as soon as the user clicks "BUY" on
    // a style. (this is necessary if the bundle price is <= single style price)
    (0, _productState.populateProductState)(store, environment, collectionId);
  }, [collectionId, store, environment]);
  const {
    licenseOptionsSpecs,
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  const {
    node
  } = (0, _reactRelay.useLazyLoadQuery)((_FontFamiliesQuery2.default.hash && _FontFamiliesQuery2.default.hash !== "18eba6b750d3f714b964af66ecfcfa3b" && console.error("The definition of 'FontFamiliesQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _FontFamiliesQuery2.default), {
    collectionId,
    licenseOptions: licenseOptionsSpecs,
    orderVariables: orderVariableSelections
  });
  if (!node) return null;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "font-families"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "font-families__items"
  }, /*#__PURE__*/_react.default.createElement(_Family.default, {
    node: node,
    size: "wide"
  }), (_node$families = node.families) === null || _node$families === void 0 ? void 0 : _node$families.map((family, index) => /*#__PURE__*/_react.default.createElement(_Family.default, {
    key: index,
    node: family
  }))));
}
var _default = FontFamilies;
exports.default = _default;
