// fontdue-js/dist/components/StoreModal/routes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactRelay = require("react-relay");
var _StoreModalIndex = _interopRequireDefault(require("./StoreModalIndex"));
var _StoreModalProduct = _interopRequireDefault(require("./StoreModalProduct"));
var _StoreModalCart = _interopRequireDefault(require("./StoreModalCart"));
var _StoreModalCheckout = _interopRequireDefault(require("./StoreModalCheckout"));
var _StoreModalIndexQuery = _interopRequireDefault(require("../../__generated__/StoreModalIndexQuery.graphql"));
var _StoreModalProductQuery = _interopRequireDefault(require("../../__generated__/StoreModalProductQuery.graphql"));
var _StoreModalCartQuery = _interopRequireDefault(require("../../__generated__/StoreModalCartQuery.graphql"));
var _StoreModalCheckoutQuery = _interopRequireDefault(require("../../__generated__/StoreModalCheckoutQuery.graphql"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const route = routeConfig => routeConfig;
const routes = {
  index: route({
    Component: _StoreModalIndex.default,
    prepare: (variables, environment) => ({
      indexQuery: (0, _reactRelay.loadQuery)(environment, _StoreModalIndexQuery.default, variables)
    })
  }),
  product: route({
    Component: _StoreModalProduct.default,
    prepare: (variables, environment) => ({
      productQuery: (0, _reactRelay.loadQuery)(environment, _StoreModalProductQuery.default, {
        ...variables,
        licenseOptions: [],
        orderVariables: []
      })
    })
  }),
  cart: route({
    Component: _StoreModalCart.default,
    prepare: (variables, environment) => ({
      cartQuery: (0, _reactRelay.loadQuery)(environment, _StoreModalCartQuery.default, variables, {
        fetchPolicy: 'store-and-network'
      })
    })
  }),
  checkout: route({
    Component: _StoreModalCheckout.default,
    prepare: (variables, environment) => ({
      checkoutQuery: (0, _reactRelay.loadQuery)(environment, _StoreModalCheckoutQuery.default, variables)
    })
  })
};
var _default = routes;
exports.default = _default;
