// fontdue-js/dist/components/StoreModal/StoreModalContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StoreModalContainerQuery2 = _interopRequireDefault(require("../../__generated__/StoreModalContainerQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _StoreModalContainer = _interopRequireDefault(require("../elements/StoreModalContainer"));
var _reactRelay = require("react-relay");
var _StoreModalCheckoutContext = _interopRequireDefault(require("./StoreModalCheckoutContext"));
var _CheckoutSteps = _interopRequireDefault(require("../Cart/CheckoutSteps"));
var _useFontStyle = _interopRequireDefault(require("../useFontStyle"));
var _StoreModalLoadingScreen = _interopRequireDefault(require("../elements/StoreModalLoadingScreen"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const getPageTitle = name => {
  switch (name) {
    case 'index':
      return 'All collections';
    case 'product':
      return 'Buying options';
    case 'cart':
      return 'Cart';
    case 'checkout':
      return 'Checkout';
    case 'review':
      return 'Review';
  }
};
const getBackLink = (history, currentRoute, hideAllCollections) => {
  if (history.length === 0) {
    if (hideAllCollections) {
      return currentRoute === 'cart' ? null : 'Close';
    }
    if (currentRoute !== 'index') return 'All collections';
    return null;
  }
  const previous = history[history.length - 1];
  if (!previous) return null;
  return getPageTitle(previous.name);
};
const StoreModalContainer = _ref => {
  var _settings, _settings2, _currentOrder, _currentOrder$orderIt;
  let {
    children
  } = _ref;
  const data = (0, _reactRelay.useLazyLoadQuery)((_StoreModalContainerQuery2.default.hash && _StoreModalContainerQuery2.default.hash !== "ae784b9ab1842dacc5857b045b399701" && console.error("The definition of 'StoreModalContainerQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalContainerQuery2.default), {});
  const uiFontStyle = (_settings = data.viewer.settings) === null || _settings === void 0 ? void 0 : _settings.uiFontStyle;
  const uiFontStyleBold = (_settings2 = data.viewer.settings) === null || _settings2 === void 0 ? void 0 : _settings2.uiFontStyleBold;
  (0, _useFontStyle.default)({
    fontStyle: uiFontStyle
  });
  (0, _useFontStyle.default)({
    fontStyle: uiFontStyleBold
  });
  const dispatch = (0, _reactRedux.useDispatch)();
  const orderItemCount = ((_currentOrder = data.viewer.currentOrder) === null || _currentOrder === void 0 ? void 0 : (_currentOrder$orderIt = _currentOrder.orderItems) === null || _currentOrder$orderIt === void 0 ? void 0 : _currentOrder$orderIt.length) ?? 0;
  const showCartButton = orderItemCount > 0;
  const handleClose = (0, _react.useCallback)(() => {
    dispatch({
      type: 'CLOSE_CART'
    });
  }, [dispatch]);
  const currentRoute = (0, _reactRedux.useSelector)(state => state.storeModalRoute.name);
  const history = (0, _reactRedux.useSelector)(state => state.storeModalHistory);
  const config = (0, _react.useContext)(_ConfigContext.default);
  const link = getBackLink(history, currentRoute, config.storeModal.hideAllCollections);
  const handleBack = () => {
    if (history.length === 0) {
      if (config.storeModal.hideAllCollections) {
        dispatch({
          type: 'CLOSE_CART'
        });
      } else {
        dispatch({
          type: 'STORE_MODAL_REPLACE',
          route: 'index',
          variables: {
            excludeTags: config.storeModal.indexExcludeTags
          }
        });
      }
    } else {
      dispatch({
        type: 'STORE_MODAL_BACK'
      });
    }
  };
  const handleCart = () => {
    dispatch({
      type: 'STORE_MODAL_NAVIGATE',
      route: 'cart'
    });
  };
  const [checkoutStep, setCheckoutStep] = (0, _react.useState)(_CheckoutSteps.default.Customer);
  const [checkoutError, setCheckoutError] = (0, _react.useState)(null);
  const regularName = uiFontStyle ? `${uiFontStyle.cssFamily} ${uiFontStyle.name}` : null;
  const boldName = uiFontStyleBold ? `${uiFontStyleBold.cssFamily} ${uiFontStyleBold.name}` : null;
  return /*#__PURE__*/_react.default.createElement(_StoreModalCheckoutContext.default.Provider, {
    value: {
      checkoutStep,
      setCheckoutStep,
      checkoutError,
      setCheckoutError
    }
  }, /*#__PURE__*/_react.default.createElement(_StoreModalContainer.default, {
    onClickBackground: handleClose,
    onClickClose: handleClose,
    onClickBack: handleBack,
    onClickCart: handleCart,
    showCartButton: showCartButton,
    orderItemCount: orderItemCount,
    fonts: {
      regular: regularName,
      bold: boldName
    },
    routeName: currentRoute
  }, {
    backText: link,
    pageTitle: getPageTitle(currentRoute),
    body: /*#__PURE__*/_react.default.createElement(_react.Suspense, {
      fallback: /*#__PURE__*/_react.default.createElement(_StoreModalLoadingScreen.default, null)
    }, children)
  }));
};
var _default = StoreModalContainer;
exports.default = _default;
