// fontdue-js/dist/components/StoreModal/StoreModalOrderVariableSelection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalOrderVariableSelection;
var _StoreModalOrderVariableSelection_order2 = _interopRequireDefault(require("../../__generated__/StoreModalOrderVariableSelection_order.graphql"));
var _StoreModalOrderVariableSelection_viewer2 = _interopRequireDefault(require("../../__generated__/StoreModalOrderVariableSelection_viewer.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _OrderVariableSelection = _interopRequireDefault(require("../OrderVariableSelection"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function StoreModalOrderVariableSelection(_ref) {
  var _order$orderVariableS, _viewer$orderVariable, _order$orderVariableS2, _order$orderVariableS3, _viewer$settings;
  let {
    viewer: viewerKey,
    order: orderKey,
    onSelectionChange
  } = _ref;
  const viewer = (0, _reactRelay.useFragment)((_StoreModalOrderVariableSelection_viewer2.default.hash && _StoreModalOrderVariableSelection_viewer2.default.hash !== "b959e46df4274deb8a53ef227787c34a" && console.error("The definition of 'StoreModalOrderVariableSelection_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalOrderVariableSelection_viewer2.default), viewerKey);
  const order = (0, _reactRelay.useFragment)((_StoreModalOrderVariableSelection_order2.default.hash && _StoreModalOrderVariableSelection_order2.default.hash !== "85bc6b91bb3e9b7ee708f776acf670fd" && console.error("The definition of 'StoreModalOrderVariableSelection_order' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalOrderVariableSelection_order2.default), orderKey);

  // Transform order variable selections for the OrderVariableSelection component
  const selections = (_order$orderVariableS = order.orderVariableSelections) === null || _order$orderVariableS === void 0 ? void 0 : _order$orderVariableS.map(selection => {
    var _selection$orderVaria, _selection$orderVaria2;
    return {
      orderVariableId: selection.orderVariable.id,
      orderVariableOptionId: ((_selection$orderVaria = selection.orderVariableOption) === null || _selection$orderVaria === void 0 ? void 0 : _selection$orderVaria.id) || null,
      orderVariableOptionAmount: ((_selection$orderVaria2 = selection.orderVariableOption) === null || _selection$orderVaria2 === void 0 ? void 0 : _selection$orderVaria2.amount) || null,
      countryCode: selection.countryCode || null
    };
  });
  const orderVariableIds = new Set(((_viewer$orderVariable = viewer.orderVariables) === null || _viewer$orderVariable === void 0 ? void 0 : _viewer$orderVariable.map(variable => variable.id)) || []);

  // Create a map of order variable ID to orderVariableDescription for active variables
  const descriptionOverrides = ((_order$orderVariableS2 = order.orderVariableSelections) === null || _order$orderVariableS2 === void 0 ? void 0 : _order$orderVariableS2.reduce((acc, selection) => {
    if (orderVariableIds.has(selection.orderVariable.id)) {
      acc[selection.orderVariable.id] = selection.orderVariableDescription;
    }
    return acc;
  }, {})) ?? {};

  // Find ghost variables (variables that exist in the order but not in the available list)
  const ghostVariables = ((_order$orderVariableS3 = order.orderVariableSelections) === null || _order$orderVariableS3 === void 0 ? void 0 : _order$orderVariableS3.filter(selection => !orderVariableIds.has(selection.orderVariable.id)).map(selection => ({
    orderVariable: {
      ...selection.orderVariable,
      description: selection.orderVariableDescription
    },
    orderVariableOption: selection.orderVariableOption
  }))) ?? [];

  // Handle updating the order when selections change
  const handleSelectionChange = (0, _react.useCallback)(newSelections => {
    if (!selections) return;
    // Combine new selections with ghost variables (we want to keep the ghost variables)

    const ghostVariableSelections = ghostVariables.map(ghost => {
      var _ghost$orderVariableO;
      return {
        orderVariableId: ghost.orderVariable.id,
        orderVariableOptionId: ((_ghost$orderVariableO = ghost.orderVariableOption) === null || _ghost$orderVariableO === void 0 ? void 0 : _ghost$orderVariableO.id) ?? null
      };
    });

    // Get existing ghost variable IDs that are in the selections
    const existingGhostIds = selections.filter(selection => ghostVariables.some(ghost => ghost.orderVariable.id === selection.orderVariableId)).map(selection => selection.orderVariableId);
    // Only add ghost variables that are not already in the selections
    const newGhostSelections = ghostVariableSelections.filter(selection => !existingGhostIds.includes(selection.orderVariableId));
    const combinedSelections = [...newSelections, ...newGhostSelections];
    onSelectionChange(combinedSelections);
  }, [onSelectionChange, ghostVariables, selections]);

  // Handle removing a ghost variable
  const handleRemoveGhostVariable = (0, _react.useCallback)(ghostVariableId => {
    if (!selections) return;
    // Create a new array of selections without the removed ghost variable
    const filteredSelections = selections.filter(selection => selection.orderVariableId !== ghostVariableId);
    onSelectionChange(filteredSelections);
  }, [onSelectionChange, selections]);
  return /*#__PURE__*/_react.default.createElement(_OrderVariableSelection.default, {
    disabled: false,
    variables: viewer,
    selections: selections ?? [],
    onSelectionChange: handleSelectionChange,
    ghostVariables: ghostVariables,
    onRemoveGhostVariable: handleRemoveGhostVariable,
    descriptionOverrides: descriptionOverrides
  }, {
    label: /*#__PURE__*/_react.default.createElement("div", {
      dangerouslySetInnerHTML: {
        __html: ((_viewer$settings = viewer.settings) === null || _viewer$settings === void 0 ? void 0 : _viewer$settings.storeModalSelectOrderVariablesLabel) ?? ''
      }
    })
  });
}
