// fontdue-js/dist/components/StoreModal/StoreModalFamilyButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FamilyButton;
var _StoreModalFamilyButton_collection2 = _interopRequireDefault(require("../../__generated__/StoreModalFamilyButton_collection.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _Price = require("../Price");
var _reducer = require("../../reducer");
var _StoreModalFamilyButton = _interopRequireDefault(require("../elements/StoreModalFamilyButton"));
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
var _useTotalStyles = _interopRequireDefault(require("../useTotalStyles"));
var _utils = require("../../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const getFamilyType = collection => {
  if (collection.collectionType === 'superfamily') return 'Collection';
  return 'Family';
};
const getSuffix = collection => {
  if (collection.isVariableFont) return;
  return ' ' + getFamilyType(collection);
};
function FamilyButton(_ref) {
  var _collection$featureSt, _collection$featureSt2;
  let {
    collection: collectionKey,
    onSelectSku,
    ...rest
  } = _ref;
  const collection = (0, _reactRelay.useFragment)((_StoreModalFamilyButton_collection2.default.hash && _StoreModalFamilyButton_collection2.default.hash !== "3ca5466b4680374b74cf161a9bdef313" && console.error("The definition of 'StoreModalFamilyButton_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalFamilyButton_collection2.default), collectionKey);
  const selector = collection.sku ? (0, _reducer.isSelected)(collection.sku.id) : () => false;
  const selected = (0, _reactRedux.useSelector)(selector);
  const handleClick = () => {
    if (!collection.sku) return;
    onSelectSku(collection.sku.id, !selected);
  };
  const {
    totalStyles,
    variableFonts
  } = (0, _useTotalStyles.default)({
    fontCollection: collection
  });
  let label = [];
  if (totalStyles > 0) {
    label.push((0, _utils.pluralize)(totalStyles, 'style'));
  }
  if (variableFonts > 0) {
    label.push((0, _utils.pluralize)(variableFonts, 'variable font'));
  }
  const supportsEnglish = (_collection$featureSt = collection.featureStyle) === null || _collection$featureSt === void 0 ? void 0 : (_collection$featureSt2 = _collection$featureSt.supportedLanguages) === null || _collection$featureSt2 === void 0 ? void 0 : _collection$featureSt2.includes('en');
  return /*#__PURE__*/_react.default.createElement(_StoreModalFamilyButton.default, _extends({
    selected: selected,
    onClick: collection.sku ? handleClick : undefined,
    supportsEnglish: supportsEnglish ?? true
  }, rest), {
    name: collection.featureStyle && supportsEnglish ? /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      Component: "span",
      fontStyle: collection.featureStyle
    }, collection.name, getSuffix(collection)) : /*#__PURE__*/_react.default.createElement("span", null, collection.name, " ", getSuffix(collection)),
    label: label.join(', '),
    price: collection.sku && /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
      sku: collection.sku,
      format: _Price.PriceFormat.Truncated
    })
  });
}
