// fontdue-js/dist/components/StoreModal/StoreModalCart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StoreModalCartQuery2 = _interopRequireDefault(require("../../__generated__/StoreModalCartQuery.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _CartItem = _interopRequireDefault(require("../Cart/CartItem"));
var _StoreModalPageContainer = _interopRequireDefault(require("../elements/StoreModalPageContainer"));
var _CartState = _interopRequireDefault(require("../Cart/CartState"));
var _CartTotals = _interopRequireDefault(require("../Cart/CartTotals"));
var _CouponCodeInput = _interopRequireDefault(require("../Cart/CouponCodeInput"));
var _StoreModalCartLayout = _interopRequireDefault(require("../elements/StoreModalCartLayout"));
var _EmptyCart = _interopRequireDefault(require("../elements/EmptyCart"));
var _StoreModalOrderVariableSelection = _interopRequireDefault(require("./StoreModalOrderVariableSelection"));
var _StoreModalUnifiedCheckout = _interopRequireDefault(require("../elements/StoreModalUnifiedCheckout"));
var _StripeProvider = _interopRequireDefault(require("../StripeProvider"));
var _utils = require("../Cart/utils");
var _ShareCart = _interopRequireDefault(require("../Cart/ShareCart"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// @ts-ignore

const StoreModalCart = _ref => {
  var _order$orderItems;
  let {
    prepared
  } = _ref;
  const data = (0, _reactRelay.usePreloadedQuery)((_StoreModalCartQuery2.default.hash && _StoreModalCartQuery2.default.hash !== "cfbbbef5d65b35eacebf30617fe2538e" && console.error("The definition of 'StoreModalCartQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalCartQuery2.default), prepared.cartQuery);
  const order = data.viewer.currentOrder;
  const orderHasItems = ((order === null || order === void 0 ? void 0 : (_order$orderItems = order.orderItems) === null || _order$orderItems === void 0 ? void 0 : _order$orderItems.length) ?? 0) > 0;
  const checkoutEnabled = (0, _utils.textVariablesAllHaveText)(order);
  return /*#__PURE__*/_react.default.createElement(_StoreModalPageContainer.default, null, {
    body: order && orderHasItems ? /*#__PURE__*/_react.default.createElement(_CartState.default, {
      order: order
    }, _ref2 => {
      var _data$viewer, _data$viewer$currentO, _data$viewer$currentO2, _order$orderItems2, _order$total, _data$viewer2, _data$viewer2$setting;
      let {
        onUpdateOrderVariableSelections,
        onRemoveDiscount,
        onUpdateOrderItem,
        errors
      } = _ref2;
      return /*#__PURE__*/_react.default.createElement(_StoreModalCartLayout.default, {
        checkoutEnabled: checkoutEnabled
      }, {
        orderVariables: (_data$viewer = data.viewer) !== null && _data$viewer !== void 0 && (_data$viewer$currentO = _data$viewer.currentOrder) !== null && _data$viewer$currentO !== void 0 && (_data$viewer$currentO2 = _data$viewer$currentO.orderVariableSelections) !== null && _data$viewer$currentO2 !== void 0 && _data$viewer$currentO2.length ? /*#__PURE__*/_react.default.createElement(_StoreModalOrderVariableSelection.default, {
          viewer: data.viewer,
          order: order,
          onSelectionChange: onUpdateOrderVariableSelections
        }) : null,
        items: order === null || order === void 0 ? void 0 : (_order$orderItems2 = order.orderItems) === null || _order$orderItems2 === void 0 ? void 0 : _order$orderItems2.map((item, i) => {
          var _errors$orderItems;
          return /*#__PURE__*/_react.default.createElement(_CartItem.default, {
            node: item,
            key: item.id,
            onChange: onUpdateOrderItem,
            errors: errors === null || errors === void 0 ? void 0 : (_errors$orderItems = errors.orderItems) === null || _errors$orderItems === void 0 ? void 0 : _errors$orderItems[i]
          });
        }),
        couponCodeInput: !(order && order.discount) && /*#__PURE__*/_react.default.createElement(_CouponCodeInput.default, null),
        shareCart: /*#__PURE__*/_react.default.createElement(_ShareCart.default, {
          key: (_order$total = order.total) === null || _order$total === void 0 ? void 0 : _order$total.amount
        }),
        totals: /*#__PURE__*/_react.default.createElement(_CartTotals.default, {
          order: order,
          onRemoveDiscount: onRemoveDiscount,
          billingIdentity: order.billingIdentity,
          taxSystem: (_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : (_data$viewer2$setting = _data$viewer2.settings) === null || _data$viewer2$setting === void 0 ? void 0 : _data$viewer2$setting.taxSystem
        }),
        checkout: /*#__PURE__*/_react.default.createElement(_StripeProvider.default, {
          viewer: data.viewer,
          providerType: "elements"
        }, /*#__PURE__*/_react.default.createElement(_StoreModalUnifiedCheckout.default, {
          viewer: data.viewer,
          order: order,
          onScrollToTop: () => {},
          onUpdateOrderVariableSelections: onUpdateOrderVariableSelections
        }))
      });
    }) : /*#__PURE__*/_react.default.createElement(_EmptyCart.default, null)
  });
};
var _default = StoreModalCart;
exports.default = _default;
