// fontdue-js/dist/components/StoreModal/StoreModalStyleButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalStyleButton;
var _StoreModalStyleButton_fontStyle2 = _interopRequireDefault(require("../../__generated__/StoreModalStyleButton_fontStyle.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _Price = require("../Price");
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
var _reducer = require("../../reducer");
var _StoreModalStyleButton = _interopRequireDefault(require("../elements/StoreModalStyleButton"));
var _utils = require("../../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const THIN_SPACE = '\u2009';
function StoreModalStyleButton(_ref) {
  var _fontStyle$variableIn;
  let {
    fontStyle: fontStyleKey,
    onSelectSku,
    isVariableFont,
    ...rest
  } = _ref;
  const fontStyle = (0, _reactRelay.useFragment)((_StoreModalStyleButton_fontStyle2.default.hash && _StoreModalStyleButton_fontStyle2.default.hash !== "3497be385e11b764fa6795bc3f3a0f1f" && console.error("The definition of 'StoreModalStyleButton_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalStyleButton_fontStyle2.default), fontStyleKey);
  const selected = (0, _reactRedux.useSelector)((0, _reducer.isSelected)(fontStyle.sku ? fontStyle.sku.id : fontStyle.id));
  const handleSelect = () => {
    if (!fontStyle.sku) return;
    onSelectSku(fontStyle.sku.id, !selected);
  };
  return /*#__PURE__*/_react.default.createElement(_StoreModalStyleButton.default, _extends({}, rest, {
    onClick: fontStyle.sku ? handleSelect : undefined,
    selected: selected
  }), {
    featureGlyphs: /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      fontStyle: fontStyle,
      Component: "span"
    }, isVariableFont ? (_fontStyle$variableIn = fontStyle.variableInstances) === null || _fontStyle$variableIn === void 0 ? void 0 : _fontStyle$variableIn.map((instance, i) => /*#__PURE__*/_react.default.createElement("span", {
      key: instance.name,
      style: (0, _utils.variableInstanceCSS)(instance)
    }, i !== 0 && THIN_SPACE, (0, _utils.getFeatureGlyphs)(fontStyle.supportedLanguages))) : (0, _utils.getFeatureGlyphs)(fontStyle.supportedLanguages)),
    // for variable fonts, only show the name of the style if there are
    // multiple styles
    name: isVariableFont ? fontStyle.sku ? fontStyle.name : undefined : fontStyle.name,
    price: fontStyle.sku ? /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
      sku: fontStyle.sku,
      format: _Price.PriceFormat.Truncated,
      showCollectionSavings: !selected
    }) : null
  });
}
