// fontdue-js/dist/components/StoreModal/StoreModalProduct.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StoreModalProductContent_collection2 = _interopRequireDefault(require("../../__generated__/StoreModalProductContent_collection.graphql"));
var _StoreModalProductQuery2 = _interopRequireDefault(require("../../__generated__/StoreModalProductQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _StoreModalFamily = _interopRequireDefault(require("./StoreModalFamily"));
var _StoreModalProductSummary = _interopRequireDefault(require("../StoreModalProductSummary"));
var _StoreModalPageContainer = _interopRequireDefault(require("../elements/StoreModalPageContainer"));
var _StoreModalProductLicenseSelection = _interopRequireDefault(require("../StoreModalProductLicenseSelection"));
var _OrderVariableSelectionRedux = _interopRequireDefault(require("../OrderVariableSelection/OrderVariableSelectionRedux"));
var _productState = require("../Root/productState");
var _StoreModalProductRefetchQuery = _interopRequireDefault(require("../../__generated__/StoreModalProductRefetchQuery.graphql"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _useRefetchOnLicenseChanges = require("../../hooks/useRefetchOnLicenseChanges");
var _StoreModalLicenseeIsBillingSelection = _interopRequireDefault(require("./StoreModalLicenseeIsBillingSelection"));
var _useLicenseAndOrderVariables = require("../../hooks/useLicenseAndOrderVariables");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const StoreModalProduct = _ref => {
  var _config$storeModal, _viewer$orderVariable, _viewer$settings, _viewer$settings2, _viewer$settings3, _viewer$settings4, _viewer$settings5, _viewer$settings6, _viewer$settings7, _viewer$settings8, _collection$children;
  let {
    prepared
  } = _ref;
  const config = (0, _react.useContext)(_ConfigContext.default);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const initialData = (0, _reactRelay.usePreloadedQuery)((_StoreModalProductQuery2.default.hash && _StoreModalProductQuery2.default.hash !== "051d45b088e87330de54058b441ad280" && console.error("The definition of 'StoreModalProductQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductQuery2.default), prepared.productQuery);
  const [data, refetch] = (0, _reactRelay.useRefetchableFragment)((_StoreModalProductContent_collection2.default.hash && _StoreModalProductContent_collection2.default.hash !== "0684ec2633ccc5d4d0866e7780df28cc" && console.error("The definition of 'StoreModalProductContent_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductContent_collection2.default), initialData.collection);
  const collection = data;
  const {
    viewer
  } = initialData;
  const refetchVariables = (0, _react.useMemo)(() => collection !== null && collection !== void 0 && collection.id ? {
    id: collection.id
  } : undefined, [collection === null || collection === void 0 ? void 0 : collection.id]);
  (0, _useRefetchOnLicenseChanges.useRefetchOnLicenseChanges)({
    environment,
    refetch,
    query: _StoreModalProductRefetchQuery.default,
    variables: refetchVariables
  });
  const store = (0, _reactRedux.useStore)();
  const dispatch = (0, _reactRedux.useDispatch)();
  const handleSelectProduct = (0, _react.useCallback)((skuId, selected) => {
    dispatch({
      type: 'SELECT_SKU_ID',
      skuId,
      selected
    });
  }, [dispatch]);
  (0, _react.useEffect)(() => {
    if (collection !== null && collection !== void 0 && collection.id) (0, _productState.populateProductState)(store, environment, collection.id);
  }, [collection === null || collection === void 0 ? void 0 : collection.id]);
  (0, _react.useEffect)(() => {
    return () => {
      // unselect the skus when this component unmounts,
      // otherwise the user may have selected items from other products
      // without an easy way to unselect them
      dispatch({
        type: 'UNSELECT_SKUS'
      });
    };
  }, []);
  const licensesPosition = config === null || config === void 0 ? void 0 : (_config$storeModal = config.storeModal) === null || _config$storeModal === void 0 ? void 0 : _config$storeModal.productLicensesPosition;
  const licenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions, _reactRedux.shallowEqual);
  const anyLicenseOptionsSelected = licenseOptions.length > 0;
  const selectedSkuIds = (0, _reactRedux.useSelector)(state => state.selectedSkuIds, _reactRedux.shallowEqual);
  const anySkusSelected = Object.keys(selectedSkuIds).filter(skuId => selectedSkuIds[skuId]).length > 0;
  const licenseeIsBillingIdentity = (0, _reactRedux.useSelector)(state => state.licenseeIsBillingIdentity);
  const hasSelectionForLicenseeIsBillingIdentity = licenseeIsBillingIdentity !== null;
  const {
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  const requiredOrderVariablesNotYetSelected = (_viewer$orderVariable = viewer.orderVariables) === null || _viewer$orderVariable === void 0 ? void 0 : _viewer$orderVariable.filter(variable => variable.variableType === 'select').find(variable => !orderVariableSelections.find(selection => selection.orderVariableId === variable.id));
  if (!collection) {
    return /*#__PURE__*/_react.default.createElement(_StoreModalPageContainer.default, null, {
      body: 'Collection not found'
    });
  }
  const selectLicenseeLabel = /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings = viewer.settings) === null || _viewer$settings === void 0 ? void 0 : _viewer$settings.storeModalSelectLicenseeLabel) ?? ''
    }
  });
  const licenseeSelfLabel = /*#__PURE__*/_react.default.createElement("span", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings2 = viewer.settings) === null || _viewer$settings2 === void 0 ? void 0 : _viewer$settings2.storeModalLicenseeSelfLabel) ?? 'Yourself'
    }
  });
  const licenseeSelfDescription = (_viewer$settings3 = viewer.settings) !== null && _viewer$settings3 !== void 0 && _viewer$settings3.storeModalLicenseeSelfDescription ? /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: viewer.settings.storeModalLicenseeSelfDescription
    }
  }) : null;
  const licenseeClientLabel = /*#__PURE__*/_react.default.createElement("span", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings4 = viewer.settings) === null || _viewer$settings4 === void 0 ? void 0 : _viewer$settings4.storeModalLicenseeClientLabel) ?? 'Your client'
    }
  });
  const licenseeClientDescription = (_viewer$settings5 = viewer.settings) !== null && _viewer$settings5 !== void 0 && _viewer$settings5.storeModalLicenseeClientDescription ? /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: viewer.settings.storeModalLicenseeClientDescription
    }
  }) : null;
  const selectOrderVariablesLabel = /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings6 = viewer.settings) === null || _viewer$settings6 === void 0 ? void 0 : _viewer$settings6.storeModalSelectOrderVariablesLabel) ?? ''
    }
  });
  const selectLicensesLabel = /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings7 = viewer.settings) === null || _viewer$settings7 === void 0 ? void 0 : _viewer$settings7.storeModalSelectLicensesLabel) ?? ''
    }
  });
  const selectFontsLabel = /*#__PURE__*/_react.default.createElement("div", {
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings8 = viewer.settings) === null || _viewer$settings8 === void 0 ? void 0 : _viewer$settings8.storeModalSelectFontsLabel) ?? ''
    }
  });
  return /*#__PURE__*/_react.default.createElement(_StoreModalPageContainer.default, null, {
    body: /*#__PURE__*/_react.default.createElement(_react.Fragment, null, licensesPosition === 'top' && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_StoreModalLicenseeIsBillingSelection.default, {
      disabled: false
    }, {
      label: selectLicenseeLabel,
      selfLabel: licenseeSelfLabel,
      selfDescription: licenseeSelfDescription,
      clientLabel: licenseeClientLabel,
      clientDescription: licenseeClientDescription
    }), /*#__PURE__*/_react.default.createElement(_OrderVariableSelectionRedux.default, {
      viewer: viewer,
      disabled: !hasSelectionForLicenseeIsBillingIdentity
    }, {
      label: selectOrderVariablesLabel
    }), /*#__PURE__*/_react.default.createElement(_StoreModalProductLicenseSelection.default, {
      collection: collection,
      disabled: !hasSelectionForLicenseeIsBillingIdentity || Boolean(requiredOrderVariablesNotYetSelected)
    }, {
      label: selectLicensesLabel
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: "store-modal__families",
      "data-disabled": licensesPosition === 'top' ? !anyLicenseOptionsSelected || !hasSelectionForLicenseeIsBillingIdentity || !!requiredOrderVariablesNotYetSelected : false
    }, /*#__PURE__*/_react.default.createElement(_StoreModalFamily.default, {
      title: selectFontsLabel,
      collection: collection,
      onSelect: handleSelectProduct
    }), (_collection$children = collection.children) === null || _collection$children === void 0 ? void 0 : _collection$children.map(child => /*#__PURE__*/_react.default.createElement(_StoreModalFamily.default, {
      key: child.id,
      collection: child,
      onSelect: handleSelectProduct
    }))), licensesPosition === 'bottom' && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_StoreModalLicenseeIsBillingSelection.default, {
      disabled: !anySkusSelected
    }, {
      label: selectLicenseeLabel,
      selfLabel: licenseeSelfLabel,
      selfDescription: licenseeSelfDescription,
      clientLabel: licenseeClientLabel,
      clientDescription: licenseeClientDescription
    }), /*#__PURE__*/_react.default.createElement(_OrderVariableSelectionRedux.default, {
      viewer: viewer,
      disabled: !anySkusSelected || !hasSelectionForLicenseeIsBillingIdentity
    }, {
      label: selectOrderVariablesLabel
    }), /*#__PURE__*/_react.default.createElement(_StoreModalProductLicenseSelection.default, {
      collection: collection,
      disabled: !anySkusSelected || !hasSelectionForLicenseeIsBillingIdentity || Boolean(requiredOrderVariablesNotYetSelected)
    }, {
      label: selectLicensesLabel
    }))),
    stickyFooter: viewer.stripeAccountId ? /*#__PURE__*/_react.default.createElement(_StoreModalProductSummary.default, {
      viewer: viewer
    }) : /*#__PURE__*/_react.default.createElement("div", null, "Please", ' ', /*#__PURE__*/_react.default.createElement("a", {
      href: `${viewer.endpoint}/admin/settings/payments`,
      target: "_blank"
    }, "connect your Stripe account"), ' ', "to enable the cart.")
  });
};
var _default = StoreModalProduct;
exports.default = _default;
