// fontdue-js/dist/components/StoreModal/StoreModalFamily.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalFamily;
var _StoreModalFamily_collection2 = _interopRequireDefault(require("../../__generated__/StoreModalFamily_collection.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _StoreModalFamilyButton = _interopRequireDefault(require("./StoreModalFamilyButton"));
var _StoreModalStyleButton = _interopRequireDefault(require("./StoreModalStyleButton"));
var _StoreModalFamily = _interopRequireDefault(require("../elements/StoreModalFamily"));
var _StoreModalBundleButton = _interopRequireDefault(require("./StoreModalBundleButton"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// chunk items if they match condition, but only chunk if they match in
// consecutive order. this prevents the list from being reordered
function chunkConsecutiveBy(collection, compare) {
  return collection.reduce((acc, item) => {
    const last = acc[acc.length - 1];
    if (last && compare(last[0], item)) {
      return [...acc.slice(0, acc.length - 1), last.concat(item)];
    }
    return acc.concat([[item]]);
  }, []);
}
const isSlanted = cssStyle => cssStyle === 'italic' || cssStyle === 'oblique';
const stylesGrouped = collection => {
  if (!collection.fontStyles) return [];
  const chunks = chunkConsecutiveBy(collection.fontStyles, (left, right) => left.cssFamily === right.cssFamily && left.cssWeight === right.cssWeight && left.cssStretch === right.cssStretch && isSlanted(left.cssStyle) !== isSlanted(right.cssStyle));

  // Only keep a chunk paired when it's exactly one upright + one slanted
  // (italic or oblique) style. Anything else (3+ consecutive matches, or
  // two styles that aren't a clean upright/slanted pair) is split back
  // into individual entries.
  return chunks.flatMap(chunk => {
    if (chunk.length !== 2) return chunk.map(style => [style]);
    const slantedCount = chunk.filter(s => isSlanted(s.cssStyle)).length;
    return slantedCount === 1 ? [chunk] : chunk.map(style => [style]);
  });
};
function StoreModalFamily(_ref) {
  var _collection$bundles;
  let {
    collection: collectionKey,
    onSelect,
    title
  } = _ref;
  const collection = (0, _reactRelay.useFragment)((_StoreModalFamily_collection2.default.hash && _StoreModalFamily_collection2.default.hash !== "d2d429ebada7593dd5debcce7a69317d" && console.error("The definition of 'StoreModalFamily_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalFamily_collection2.default), collectionKey);
  const fontStyles = (0, _react.useMemo)(() => stylesGrouped(collection), [collection]);
  if ((fontStyles === null || fontStyles === void 0 ? void 0 : fontStyles.length) === 0 && collection.collectionType === 'superfamily' && !collection.sku) return null;
  return /*#__PURE__*/_react.default.createElement(_StoreModalFamily.default, null, {
    title,
    familyButton: /*#__PURE__*/_react.default.createElement(_StoreModalFamilyButton.default, {
      collection: collection,
      onSelectSku: onSelect
    }),
    bundles: collection === null || collection === void 0 ? void 0 : (_collection$bundles = collection.bundles) === null || _collection$bundles === void 0 ? void 0 : _collection$bundles.filter(bundle => !bundle.hidden).map(bundle => /*#__PURE__*/_react.default.createElement(_StoreModalBundleButton.default, {
      key: bundle.id,
      bundle: bundle,
      onSelectSku: onSelect
    })),
    styleGroups: fontStyles.map(group => group.map(fontStyle => /*#__PURE__*/_react.default.createElement(_StoreModalStyleButton.default, {
      key: fontStyle.id,
      fontStyle: fontStyle,
      onSelectSku: onSelect,
      isVariableFont: collection.isVariableFont
    })))
  });
}
