// fontdue-js/dist/components/StoreModal/StoreModalIndexItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StoreModalIndexItem_fontCollection2 = _interopRequireDefault(require("../../__generated__/StoreModalIndexItem_fontCollection.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _reactRelay = require("react-relay");
var _StoreModalIndexItem = _interopRequireDefault(require("../elements/StoreModalIndexItem"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
var _useTotalStyles = _interopRequireDefault(require("../useTotalStyles"));
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const StoreModalIndexItem = _ref => {
  var _data$featureStyle;
  let {
    collection
  } = _ref;
  const {
    storeModal: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const data = (0, _reactRelay.useFragment)((_StoreModalIndexItem_fontCollection2.default.hash && _StoreModalIndexItem_fontCollection2.default.hash !== "8a990fad8dc50c9f00b391902e9bc482" && console.error("The definition of 'StoreModalIndexItem_fontCollection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalIndexItem_fontCollection2.default), collection);
  const dispatch = (0, _reactRedux.useDispatch)();
  const handleClickProduct = productId => {
    dispatch({
      type: 'STORE_MODAL_NAVIGATE',
      route: 'product',
      variables: {
        productId
      }
    });
  };
  const {
    totalStyles,
    variableFonts
  } = (0, _useTotalStyles.default)({
    fontCollection: data
  });
  let label = [];
  if (totalStyles > 0) {
    label.push((0, _utils.pluralize)(totalStyles, 'style'));
  }
  if (variableFonts > 0) {
    label.push((0, _utils.pluralize)(variableFonts, 'variable font'));
  }
  return /*#__PURE__*/_react.default.createElement(_StoreModalIndexItem.default, {
    key: data.id,
    onClick: handleClickProduct.bind(null, data.id)
  }, {
    aa: config.indexLayout === 'styled-aa' && (data.featureStyle ? /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      fontStyle: data.featureStyle
    }, (0, _utils.getFeatureGlyphs)((_data$featureStyle = data.featureStyle) === null || _data$featureStyle === void 0 ? void 0 : _data$featureStyle.supportedLanguages)) : null),
    name: config.indexLayout === 'styled-font-names' ? data.featureStyle && /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      fontStyle: data.featureStyle
    }, data.name) : data.name,
    label: label.join(', ')
  });
};
var _default = StoreModalIndexItem;
exports.default = _default;
