// fontdue-js/dist/components/StoreModal/StoreModalBundleButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalBundleButton;
var _StoreModalBundleButton_bundle2 = _interopRequireDefault(require("../../__generated__/StoreModalBundleButton_bundle.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _Price = require("../Price");
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
var _reducer = require("../../reducer");
var _StoreModalStyleButton = _interopRequireDefault(require("../elements/StoreModalStyleButton"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const THIN_SPACE = '\u2009';
function StoreModalBundleButton(_ref) {
  var _bundle$sku;
  let {
    bundle: bundleKey,
    onSelectSku,
    ...rest
  } = _ref;
  const bundle = (0, _reactRelay.useFragment)((_StoreModalBundleButton_bundle2.default.hash && _StoreModalBundleButton_bundle2.default.hash !== "524b936dedb01e6669ca020b96e28ea6" && console.error("The definition of 'StoreModalBundleButton_bundle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalBundleButton_bundle2.default), bundleKey);
  const selected = (0, _reactRedux.useSelector)((0, _reducer.isSelected)((bundle === null || bundle === void 0 ? void 0 : (_bundle$sku = bundle.sku) === null || _bundle$sku === void 0 ? void 0 : _bundle$sku.id) ?? null));
  if (!bundle.sku || !bundle.fontStyles) return null;
  const handleSelect = () => {
    if (!bundle.sku) return;
    onSelectSku(bundle.sku.id, !selected);
  };
  return /*#__PURE__*/_react.default.createElement(_StoreModalStyleButton.default, _extends({}, rest, {
    onClick: handleSelect,
    selected: selected
  }), {
    featureGlyphs: bundle.fontStyles.map((fontStyle, i) => /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      key: fontStyle.id,
      fontStyle: fontStyle,
      Component: "span"
    }, i !== 0 && THIN_SPACE, "Aa")),
    name: bundle.name,
    price: /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
      sku: bundle.sku,
      format: _Price.PriceFormat.Truncated,
      showCollectionSavings: !selected
    })
  });
}
