// fontdue-js/dist/components/BuyButton/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BuyButtonPreloadedIDQueryRenderer = BuyButtonPreloadedIDQueryRenderer;
exports.BuyButtonPreloadedSlugQueryRenderer = BuyButtonPreloadedSlugQueryRenderer;
exports.default = BuyButton;
var _BuyButtonSlugQuery2 = _interopRequireDefault(require("../../__generated__/BuyButtonSlugQuery.graphql"));
var _BuyButtonIDQuery2 = _interopRequireDefault(require("../../__generated__/BuyButtonIDQuery.graphql"));
var _BuyButton_collection2 = _interopRequireDefault(require("../../__generated__/BuyButton_collection.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _ComponentsContext = _interopRequireDefault(require("../ComponentsContext"));
var _productState = require("../Root/productState");
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function BuyButtonComponent(_ref) {
  let {
    collection: collectionKey,
    label,
    collectionName
  } = _ref;
  const collection = (0, _reactRelay.useFragment)((_BuyButton_collection2.default.hash && _BuyButton_collection2.default.hash !== "9c47030d7450494449d4aa5bda16e27e" && console.error("The definition of 'BuyButton_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _BuyButton_collection2.default), collectionKey);
  const store = (0, _reactRedux.useStore)();
  const components = (0, _react.useContext)(_ComponentsContext.default);
  const dispatch = (0, _reactRedux.useDispatch)();
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const handleClick = (0, _react.useCallback)(() => {
    if (!collection) return;
    dispatch({
      type: 'STORE_MODAL_NAVIGATE',
      route: 'product',
      variables: {
        productId: collection.id
      }
    });
    dispatch({
      type: 'OPEN_CART'
    });
  }, [dispatch, collection]);
  (0, _react.useEffect)(() => {
    if (!collection) return;
    // populate as soon as BuyButton mounts. this prepares the StoreModal
    // to be able to apply discounts as soon as the user clicks "BUY" on
    // a style. (this is necessary if the bundle price is <= single style price)
    (0, _productState.populateProductState)(store, environment, collection.id);
  }, [collection === null || collection === void 0 ? void 0 : collection.id]);
  if (components.BuyButton) {
    return /*#__PURE__*/_react.default.createElement(components.BuyButton, {
      innerProps: {
        onClick: handleClick
      }
    });
  }
  return /*#__PURE__*/_react.default.createElement("button", {
    className: "buy-button",
    onClick: handleClick,
    type: "button"
  }, label ? label : `Buy ${collectionName ?? ''}`);
}
const idQuery = (_BuyButtonIDQuery2.default.hash && _BuyButtonIDQuery2.default.hash !== "4fbf1dbf9e6c530a5d38c697b174d8b0" && console.error("The definition of 'BuyButtonIDQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _BuyButtonIDQuery2.default);
function BuyButtonIDQueryRenderer(_ref2) {
  let {
    collectionId,
    ...rest
  } = _ref2;
  const data = (0, _reactRelay.useLazyLoadQuery)(idQuery, {
    collectionId
  });
  return /*#__PURE__*/_react.default.createElement(BuyButtonComponent, _extends({}, rest, data));
}
function BuyButtonPreloadedIDQueryRenderer(_ref3) {
  let {
    preloadedQuery,
    ...rest
  } = _ref3;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(idQuery, queryRef);
  return /*#__PURE__*/_react.default.createElement(BuyButtonComponent, _extends({}, rest, data));
}
const slugQuery = (_BuyButtonSlugQuery2.default.hash && _BuyButtonSlugQuery2.default.hash !== "6e750adea09698f7cb61f435cd88fd26" && console.error("The definition of 'BuyButtonSlugQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _BuyButtonSlugQuery2.default);
function BuyButtonSlugQueryRenderer(_ref4) {
  var _data$viewer, _data$viewer$slug;
  let {
    collectionSlug,
    ...rest
  } = _ref4;
  const data = (0, _reactRelay.useLazyLoadQuery)(slugQuery, {
    collectionSlug
  });
  return /*#__PURE__*/_react.default.createElement(BuyButtonComponent, _extends({}, rest, {
    collection: ((_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : (_data$viewer$slug = _data$viewer.slug) === null || _data$viewer$slug === void 0 ? void 0 : _data$viewer$slug.collection) ?? null
  }));
}
function BuyButtonPreloadedSlugQueryRenderer(_ref5) {
  var _data$viewer2, _data$viewer2$slug;
  let {
    preloadedQuery,
    ...rest
  } = _ref5;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(slugQuery, queryRef);
  return /*#__PURE__*/_react.default.createElement(BuyButtonComponent, _extends({}, rest, {
    collection: ((_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : (_data$viewer2$slug = _data$viewer2.slug) === null || _data$viewer2$slug === void 0 ? void 0 : _data$viewer2$slug.collection) ?? null
  }));
}

// default client export
function BuyButton(_ref6) {
  let {
    collectionId,
    collectionSlug,
    ...rest
  } = _ref6;
  if (collectionId) {
    return /*#__PURE__*/_react.default.createElement(BuyButtonIDQueryRenderer, _extends({
      collectionId: collectionId
    }, rest));
  }
  if (collectionSlug) {
    return /*#__PURE__*/_react.default.createElement(BuyButtonSlugQueryRenderer, _extends({
      collectionSlug: collectionSlug
    }, rest));
  }
  throw new Error('BuyButton expected either a collectionId or collectionSlug prop');
}
