// fontdue-js/dist/components/elements/EmptyCart/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _Button = _interopRequireDefault(require("../Button"));
var _reactRedux = require("react-redux");
var _ConfigContext = _interopRequireDefault(require("../../ConfigContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const EmptyCart = () => {
  const dispatch = (0, _reactRedux.useDispatch)();
  const config = (0, _react.useContext)(_ConfigContext.default);
  const {
    hideAllCollections
  } = config.storeModal;
  const handleContinueShopping = () => {
    if (hideAllCollections) {
      dispatch({
        type: 'CLOSE_CART'
      });
    } else {
      dispatch({
        type: 'STORE_MODAL_NAVIGATE',
        route: 'index',
        variables: {
          excludeTags: config.storeModal.indexExcludeTags
        }
      });
    }
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "empty-cart__container"
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: "empty-cart__message"
  }, "Your cart is empty."), /*#__PURE__*/_react.default.createElement(_Button.default, {
    onClick: handleContinueShopping
  }, "Continue shopping"));
};
var _default = EmptyCart;
exports.default = _default;
