// fontdue-js/dist/components/StickyNav/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _CartButton = _interopRequireDefault(require("../CartButton"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function NavItem(_ref) {
  let {
    label,
    link,
    text
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "sticky-nav__item",
    "data-label": label
  }, /*#__PURE__*/_react.default.createElement("a", {
    className: "sticky-nav__link",
    href: link
  }, text));
}
function StickyNav(_ref2) {
  let {
    aboutPath,
    homePath,
    typefacesPath,
    siteTitle
  } = _ref2;
  const $stickyNav = (0, _react.useRef)(null);
  const [visible, setVisible] = (0, _react.useState)(false);
  (0, _react.useEffect)(() => {
    function onScroll() {
      var _$stickyNav$current;
      const rect = (_$stickyNav$current = $stickyNav.current) === null || _$stickyNav$current === void 0 ? void 0 : _$stickyNav$current.getBoundingClientRect();

      // there is an issue when the page initially loads, it takes some time
      // to paint the div, and incorrectly reporting the top value as zero.
      // so we keep trying it until it has a non-zero width
      if ((rect === null || rect === void 0 ? void 0 : rect.width) !== 0) {
        var _$stickyNav$current2;
        setVisible(((_$stickyNav$current2 = $stickyNav.current) === null || _$stickyNav$current2 === void 0 ? void 0 : _$stickyNav$current2.getBoundingClientRect().top) === 0);
      } else {
        requestAnimationFrame(onScroll);
      }
    }
    window.addEventListener('scroll', onScroll);
    window.addEventListener('resize', onScroll);
    onScroll();
    return () => {
      window.removeEventListener('scroll', onScroll);
      window.removeEventListener('resize', onScroll);
    };
  }, [$stickyNav, setVisible]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "sticky-nav",
    ref: $stickyNav,
    "data-visible": visible
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "sticky-nav__wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "sticky-nav__links"
  }, /*#__PURE__*/_react.default.createElement(NavItem, {
    label: "home",
    link: homePath,
    text: siteTitle
  }), /*#__PURE__*/_react.default.createElement(NavItem, {
    label: "typefaces",
    link: typefacesPath,
    text: "Typefaces"
  }), /*#__PURE__*/_react.default.createElement(NavItem, {
    label: "about",
    link: aboutPath,
    text: "About & contact"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "sticky-nav__item",
    "data-label": "cart"
  }, /*#__PURE__*/_react.default.createElement(_CartButton.default, null))));
}
var _default = StickyNav;
exports.default = _default;
