// fontdue-js/dist/vanilla/FontStyle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _fontfaceobserver = _interopRequireDefault(require("fontfaceobserver"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const FALLBACK_FAMILY = 'Fallback';
const setStyle = ($el, _ref) => {
  let {
    fontFamily,
    fontWeight,
    fontStyle
  } = _ref;
  $el.style.fontFamily = `"${fontFamily}"`;
  $el.style.fontWeight = fontWeight;
  $el.style.fontStyle = fontStyle;
};
class FontStyle {
  constructor(props) {
    this.$fontStyles = [...document.querySelectorAll('[data-font-style]')];
    this.render(props);
  }
  render() {
    let {
      onAllFontsLoaded
    } = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    let loaded = 0;
    const toLoad = this.$fontStyles.length;
    this.$fontStyles.forEach($fontStyle => {
      const data = JSON.parse($fontStyle.dataset.fontStyle);
      const fontFamily = data['font-family'];
      const fontWeight = data['font-weight'];
      const fontStyle = data['font-style'];
      setStyle($fontStyle, {
        fontFamily: FALLBACK_FAMILY,
        fontWeight: 'normal',
        fontStyle: 'normal'
      });
      const observer = new _fontfaceobserver.default(fontFamily, {
        weight: fontWeight,
        style: fontStyle
      });
      observer.load(null, 90000).then(out => {
        loaded++;
        if (typeof onAllFontsLoaded === 'function' && loaded === toLoad) {
          onAllFontsLoaded();
        }
        setStyle($fontStyle, {
          fontFamily,
          fontWeight,
          fontStyle
        });
      }).catch(console.error);
    });
  }
}
var _default = FontStyle;
exports.default = _default;
