// fontdue-js/dist/components/Family/FamilyList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FamilyList;
var _FamilyList_node2 = _interopRequireDefault(require("../../__generated__/FamilyList_node.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const styleName = (normal, localized) => {
  if (normal === localized) return normal;
  return `${normal} ${localized}`;
};
function FamilyList(_ref) {
  var _node$fontStyles;
  let {
    node: nodeKey,
    text
  } = _ref;
  const node = (0, _reactRelay.useFragment)((_FamilyList_node2.default.hash && _FamilyList_node2.default.hash !== "d52770ab2e7aaa83f392a5d80340f3ad" && console.error("The definition of 'FamilyList_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _FamilyList_node2.default), nodeKey);
  return /*#__PURE__*/_react.default.createElement("ul", {
    className: "font-family-block__list",
    style: {
      fontFamily: node.name
    }
  }, (_node$fontStyles = node.fontStyles) === null || _node$fontStyles === void 0 ? void 0 : _node$fontStyles.map((style, i) => {
    var _style$variableInstan, _style$variableInstan2;
    return /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
      fontStyle: style,
      Component: "li",
      key: i,
      className: "font-family-block__list__item"
    }, (((_style$variableInstan = style.variableInstances) === null || _style$variableInstan === void 0 ? void 0 : _style$variableInstan.length) ?? 0) > 0 ? (_style$variableInstan2 = style.variableInstances) === null || _style$variableInstan2 === void 0 ? void 0 : _style$variableInstan2.map(_ref2 => {
      let {
        name,
        coordinates
      } = _ref2;
      return /*#__PURE__*/_react.default.createElement("span", {
        key: name,
        style: {
          fontVariationSettings: coordinates.map(_ref3 => {
            let {
              axis,
              value
            } = _ref3;
            return `'${axis}' ${value}`;
          }).join(',')
        }
      }, name, ' ');
    }) : /*#__PURE__*/_react.default.createElement("span", null, text || styleName(style.name, style.name_localized), " "));
  }));
}
