// fontdue-js/dist/components/Family/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Family;
var _Family_node2 = _interopRequireDefault(require("../../__generated__/Family_node.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _SelectButton = _interopRequireDefault(require("../SelectButton"));
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _Price = require("../Price");
var _FamilyList = _interopRequireDefault(require("./FamilyList"));
var _reducer = require("../../reducer");
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function Family(_ref) {
  var _node$fontStyles, _node$fontStyles2, _node$families;
  let {
    node: nodeKey,
    size
  } = _ref;
  const node = (0, _reactRelay.useFragment)((_Family_node2.default.hash && _Family_node2.default.hash !== "5c49167e8bf2458ab72604ff960636b2" && console.error("The definition of 'Family_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _Family_node2.default), nodeKey);
  const precartOpen = (0, _reactRedux.useSelector)(state => state.precartOpen);
  const selector = node.sku ? (0, _reducer.isSelected)(node.sku.id) : () => false;
  const selected = (0, _reactRedux.useSelector)(selector);
  const [buyButtonIsHovered, setBuyButtonIsHovered] = (0, _react.useState)(false);
  const isSuperFamily = node.collectionType === 'superfamily';
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "select-button-hover-parent font-family-block",
    "data-size": size,
    "data-selected": selected,
    "data-buy-button-hovered": buyButtonIsHovered
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "font-family-block__wrapper"
  }, node.sku && /*#__PURE__*/_react.default.createElement("div", {
    className: "font-family-block__select-button"
  }, /*#__PURE__*/_react.default.createElement(_SelectButton.default, {
    sku: node.sku,
    productId: node.sku.id,
    onMouseEnter: () => setBuyButtonIsHovered(true),
    onMouseLeave: () => setBuyButtonIsHovered(false),
    className: "font-family-block__select-button-element"
  }), precartOpen && !selected && /*#__PURE__*/_react.default.createElement("div", {
    className: "font-family-block__price"
  }, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: null
  }, /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
    sku: node.sku,
    format: _Price.PriceFormat.Truncated
  })))), /*#__PURE__*/_react.default.createElement("div", {
    className: "font-family-block__label"
  }, /*#__PURE__*/_react.default.createElement("h3", {
    className: "font-family-block__name"
  }, isSuperFamily ? `${node.name} Super family` : node.name), /*#__PURE__*/_react.default.createElement("span", {
    className: "font-family-block__separator"
  }), /*#__PURE__*/_react.default.createElement("p", {
    className: "font-family-block__count"
  }, (0, _utils.pluralize)(node.totalStyles ?? 0, node.isVariableFont ? 'font' : 'style')), node.isVariableFont ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
    className: "font-family-block__separator"
  }), /*#__PURE__*/_react.default.createElement("p", {
    className: "font-family-block__count"
  }, (0, _utils.pluralize)(((_node$fontStyles = node.fontStyles) === null || _node$fontStyles === void 0 ? void 0 : _node$fontStyles.reduce((total, style) => {
    var _style$variableAxes;
    return total + (((_style$variableAxes = style.variableAxes) === null || _style$variableAxes === void 0 ? void 0 : _style$variableAxes.length) ?? 0);
  }, 0)) ?? 0, 'variable axis', 'variable axes')), /*#__PURE__*/_react.default.createElement("span", {
    className: "font-family-block__separator"
  }), /*#__PURE__*/_react.default.createElement("p", {
    className: "font-family-block__count"
  }, (0, _utils.pluralize)(((_node$fontStyles2 = node.fontStyles) === null || _node$fontStyles2 === void 0 ? void 0 : _node$fontStyles2.reduce((total, style) => {
    var _style$variableInstan;
    return total + (((_style$variableInstan = style.variableInstances) === null || _style$variableInstan === void 0 ? void 0 : _style$variableInstan.length) ?? 0);
  }, 0)) ?? 0, 'named instance'))) : null), /*#__PURE__*/_react.default.createElement("div", {
    className: "font-family-block__lists"
  }, isSuperFamily ? (_node$families = node.families) === null || _node$families === void 0 ? void 0 : _node$families.map((family, index) => {
    var _family$featureStyle;
    return /*#__PURE__*/_react.default.createElement(_FamilyList.default, {
      key: index,
      node: family,
      text: (_family$featureStyle = family.featureStyle) === null || _family$featureStyle === void 0 ? void 0 : _family$featureStyle.keyCharacters_localized
    });
  }) : /*#__PURE__*/_react.default.createElement(_FamilyList.default, {
    node: node
  }))));
}
