// fontdue-js/dist/components/SelectButton/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SelectButton;
var _SelectButton_sku2 = _interopRequireDefault(require("../../__generated__/SelectButton_sku.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _ComponentsContext = _interopRequireDefault(require("../ComponentsContext"));
var _utils = require("../../utils");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const DefaultSelectButton = _ref => {
  let {
    productId,
    innerProps: {
      onClick,
      ...rest
    }
  } = _ref;
  const dispatch = (0, _reactRedux.useDispatch)();
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const handleClick = e => {
    onClick === null || onClick === void 0 ? void 0 : onClick(e);
    dispatch({
      type: 'STORE_MODAL_NAVIGATE',
      route: 'product',
      variables: {
        productId
      }
    });
    dispatch({
      type: 'OPEN_CART'
    });
  };
  return /*#__PURE__*/_react.default.createElement("button", _extends({}, rest, {
    onClick: handleClick,
    className: "select-button"
  }), config.selectButtonLabel);
};
function SelectButton(_ref2) {
  let {
    productId,
    sku: skuKey,
    ...rest
  } = _ref2;
  const sku = (0, _reactRelay.useFragment)((_SelectButton_sku2.default.hash && _SelectButton_sku2.default.hash !== "ce7cb4cf209fc2d1076d29ecbda4ebf3" && console.error("The definition of 'SelectButton_sku' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _SelectButton_sku2.default), skuKey);
  const dispatch = (0, _reactRedux.useDispatch)();
  const components = (0, _react.useContext)(_ComponentsContext.default);
  const ButtonComponent = components.SelectButton || DefaultSelectButton;
  const precartOpen = (0, _reactRedux.useSelector)(state => state.precartOpen);
  const selected = (0, _reactRedux.useSelector)(state => sku && (0, _utils.isSelected)(sku.id)(state));
  const onSelectSkuId = (skuId, selected) => dispatch({
    type: 'SELECT_SKU_ID',
    skuId,
    selected
  });
  return /*#__PURE__*/_react.default.createElement(ButtonComponent, {
    productId: productId,
    selected: selected,
    precartOpen: precartOpen,
    innerProps: {
      onClick: () => onSelectSkuId(sku.id, !selected),
      ...rest
    }
  }, {
    price: /*#__PURE__*/_react.default.createElement(_react.Suspense, {
      fallback: null
    }, /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
      sku: sku
    }))
  });
}
