// fontdue-js/dist/components/StoreModalProductLicenseSelection/StoreModalProductLicense.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalProduceLicense;
var _StoreModalProductLicense_license2 = _interopRequireDefault(require("../../__generated__/StoreModalProductLicense_license.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _LicenseElement = _interopRequireDefault(require("./LicenseElement"));
var _LicenseVariableElement = _interopRequireDefault(require("./LicenseVariableElement"));
var _VariableTableAmounts = _interopRequireDefault(require("../VariableTableAmounts"));
var _LicenseVariableRadioElement = _interopRequireDefault(require("./LicenseVariableRadioElement"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const specFromLicense = license => {
  let licenseOptionSpec = {
    licenseId: license.id
  };
  if (license.variables) {
    const options = license.variables.reduce((res, variable) => {
      if (variable.options && variable.options.length > 0) {
        res[variable.id] = variable.options[0].id;
      } else {
        // FIXME this is only necessary to submit the variable,
        // since it's required to have a selection for every variable
        res[variable.id] = null;
      }
      return res;
    }, {});
    licenseOptionSpec = {
      ...licenseOptionSpec,
      licenseOptions: options
    };
  }
  return licenseOptionSpec;
};
function StoreModalProduceLicense(_ref) {
  var _license$variables;
  let {
    license: licenseKey
  } = _ref;
  const license = (0, _reactRelay.useFragment)((_StoreModalProductLicense_license2.default.hash && _StoreModalProductLicense_license2.default.hash !== "87cd41e76dcd8952d7096940a383b9f3" && console.error("The definition of 'StoreModalProductLicense_license' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductLicense_license2.default), licenseKey);
  const licenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions);
  const selectedOptions = licenseOptions.find(_ref2 => {
    let {
      licenseId
    } = _ref2;
    return licenseId === license.id;
  });
  const isSelected = Boolean(selectedOptions);
  const dispatch = (0, _reactRedux.useDispatch)();
  const handleSelectLicense = (0, _react.useCallback)(() => {
    var _license$excludedLice;
    dispatch({
      type: 'PRECART_TOGGLE_LICENSE',
      licenseId: license.id,
      checked: !isSelected,
      licenseOptionSpec: specFromLicense(license),
      excludedLicenseIds: (_license$excludedLice = license.excludedLicenses) === null || _license$excludedLice === void 0 ? void 0 : _license$excludedLice.map(l => l.id)
    });
  }, [dispatch, license, isSelected]);
  const handleLicenseOptionChange = (0, _react.useCallback)(_ref3 => {
    let {
      variableId,
      optionId
    } = _ref3;
    if (!isSelected) {
      var _license$excludedLice2;
      dispatch({
        type: 'PRECART_SELECT_LICENSE',
        licenseOptionSpec: specFromLicense(license),
        excludedLicenseIds: (_license$excludedLice2 = license.excludedLicenses) === null || _license$excludedLice2 === void 0 ? void 0 : _license$excludedLice2.map(l => l.id)
      });
    }
    dispatch({
      type: 'PRECART_SELECT_LICENSE_OPTION',
      licenseId: license.id,
      variableId,
      optionId
    });
  }, [dispatch, license, isSelected]);
  (0, _react.useEffect)(() => {
    if (license.defaultSelected) {
      var _license$excludedLice3;
      dispatch({
        type: 'PRECART_SELECT_LICENSE',
        licenseOptionSpec: specFromLicense(license),
        excludedLicenseIds: (_license$excludedLice3 = license.excludedLicenses) === null || _license$excludedLice3 === void 0 ? void 0 : _license$excludedLice3.map(l => l.id)
      });
    }
  }, []);
  return /*#__PURE__*/_react.default.createElement(_LicenseElement.default, {
    onClick: handleSelectLicense,
    isSelected: isSelected
  }, {
    name: license.name,
    description: license.description,
    variables: ((_license$variables = license.variables) === null || _license$variables === void 0 ? void 0 : _license$variables.map(variable => {
      if (variable.variableType === 'text') return null;
      if (!variable.options) return null;
      const selectedLicenseOption = selectedOptions && selectedOptions.licenseOptions[variable.id];
      const selectedOption = variable.options.find(_ref4 => {
        let {
          id
        } = _ref4;
        return id === selectedLicenseOption;
      });
      if (variable.variableType === 'table') {
        return /*#__PURE__*/_react.default.createElement(_LicenseVariableRadioElement.default, {
          key: variable.id,
          onChange: value => handleLicenseOptionChange({
            variableId: variable.id,
            optionId: value
          }),
          value: selectedLicenseOption,
          options: variable.options.map(option => ({
            value: option.id,
            text: option.name,
            help: /*#__PURE__*/_react.default.createElement(_VariableTableAmounts.default, {
              variable: variable,
              option: option
            })
          }))
        });
      }
      if (variable.variableType === 'select') {
        return /*#__PURE__*/_react.default.createElement(_LicenseVariableElement.default, {
          key: variable.id,
          onChange: e => handleLicenseOptionChange({
            variableId: variable.id,
            optionId: e.target.value
          }),
          value: selectedLicenseOption || '',
          options: variable.options.map(option => ({
            value: option.id,
            text: option.amount
          }))
        }, {
          selectionText: selectedOption ? selectedOption.amount : `Select ${variable.name}`
        });
      }
      return null;
    }).filter(Boolean)) ?? null
  });
}
