// fontdue-js/dist/components/elements/StoreModalContainer/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _Icons = require("../../Icons");
var _StoreModalContainerContext = _interopRequireDefault(require("./StoreModalContainerContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const StoreModalContainer = _ref => {
  let {
    onClickBackground,
    onClickClose,
    onClickBack,
    onClickCart,
    showCartButton,
    orderItemCount,
    fonts,
    routeName,
    children: {
      backText,
      pageTitle,
      body
    }
  } = _ref;
  const container = (0, _react.useRef)(null);
  const handleScrollToTop = (0, _react.useCallback)(() => {
    var _container$current, _container$current$sc;
    (_container$current = container.current) === null || _container$current === void 0 ? void 0 : (_container$current$sc = _container$current.scrollTo) === null || _container$current$sc === void 0 ? void 0 : _container$current$sc.call(_container$current, {
      top: 0,
      left: 0
    });
  }, [container]);
  return /*#__PURE__*/_react.default.createElement(_StoreModalContainerContext.default.Provider, {
    value: {
      scrollToTop: handleScrollToTop
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__overlay",
    "data-route": routeName
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__background",
    onClick: onClickBackground
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__container",
    ref: container,
    style: {
      '--font-family': fonts.regular ? `"${fonts.regular}"` : undefined,
      '--font-family__bold': fonts.bold ? `"${fonts.bold}"` : undefined,
      fontFamily: 'var(--font-family, inherit)'
    }
  }, /*#__PURE__*/_react.default.createElement("nav", {
    className: "store-modal__container__nav",
    "data-has-back-text": Boolean(backText)
  }, backText && /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("button", {
    className: "store-modal__container__back-button",
    onClick: onClickBack
  }, /*#__PURE__*/_react.default.createElement(_Icons.ArrowLeft, null), " ", backText)), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__title"
  }, pageTitle), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__buttons"
  }, showCartButton && /*#__PURE__*/_react.default.createElement("button", {
    className: "store-modal__container__cart-button",
    onClick: onClickCart,
    "aria-label": "Cart",
    "data-count": orderItemCount
  }, /*#__PURE__*/_react.default.createElement(_Icons.CartNew, null)), /*#__PURE__*/_react.default.createElement("button", {
    className: "store-modal__container__close-button",
    onClick: onClickClose,
    "aria-label": "Close"
  }, /*#__PURE__*/_react.default.createElement(_Icons.X, null)))), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__container__body"
  }, body))));
};
var _default = StoreModalContainer;
exports.default = _default;
