// fontdue-js/dist/components/Precart/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _PrecartQuery2 = _interopRequireDefault(require("../../__generated__/PrecartQuery.graphql"));
var _PrecartAddToCartMutation2 = _interopRequireDefault(require("../../__generated__/PrecartAddToCartMutation.graphql"));
var _PrecartClearCartMutation2 = _interopRequireDefault(require("../../__generated__/PrecartClearCartMutation.graphql"));
var _Precart_collection2 = _interopRequireDefault(require("../../__generated__/Precart_collection.graphql"));
var _Precart_viewer2 = _interopRequireDefault(require("../../__generated__/Precart_viewer.graphql"));
var _Precart_license2 = _interopRequireDefault(require("../../__generated__/Precart_license.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _License = _interopRequireDefault(require("./License"));
var _Price = require("../Price");
var _utils = require("../../utils");
var _ComponentsContext = _interopRequireDefault(require("../ComponentsContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function skuName(sku) {
  if (sku.product && 'name' in sku.product) return sku.product.name;
  return null;
}
const countStyles = _ref => {
  let {
    precart
  } = _ref;
  if (!(precart !== null && precart !== void 0 && precart.skus)) return 0;
  return precart.skus.reduce((acc, sku) => {
    var _sku$product, _sku$product2;
    if (((_sku$product = sku.product) === null || _sku$product === void 0 ? void 0 : _sku$product.__typename) === 'FontCollection' || ((_sku$product2 = sku.product) === null || _sku$product2 === void 0 ? void 0 : _sku$product2.__typename) === 'Bundle') {
      var _sku$product3;
      return acc + (((_sku$product3 = sku.product) === null || _sku$product3 === void 0 ? void 0 : _sku$product3.totalStyles) ?? 0);
    }
    return acc + 1;
  }, 0);
};
const countStylesPrice = _ref2 => {
  let {
    precart
  } = _ref2;
  if (!(precart !== null && precart !== void 0 && precart.skus)) return 0;
  return precart.skus.reduce((acc, sku) => {
    var _sku$product4, _sku$product5;
    if (((_sku$product4 = sku.product) === null || _sku$product4 === void 0 ? void 0 : _sku$product4.__typename) === 'FontCollection' || ((_sku$product5 = sku.product) === null || _sku$product5 === void 0 ? void 0 : _sku$product5.__typename) === 'Bundle') {
      var _sku$product6, _sku$product6$totalSt;
      return acc + (((_sku$product6 = sku.product) === null || _sku$product6 === void 0 ? void 0 : (_sku$product6$totalSt = _sku$product6.totalStylesPrice) === null || _sku$product6$totalSt === void 0 ? void 0 : _sku$product6$totalSt.amount) ?? sku.price.amount);
    }
    return acc + sku.price.amount;
  }, 0);
};
const calculateBundleDiscounts = viewer => {
  var _viewer$precart, _viewer$precart$subto;
  const stylesPrice = countStylesPrice(viewer);
  const subtotal = ((_viewer$precart = viewer.precart) === null || _viewer$precart === void 0 ? void 0 : (_viewer$precart$subto = _viewer$precart.subtotal) === null || _viewer$precart$subto === void 0 ? void 0 : _viewer$precart$subto.amount) ?? 0;
  return subtotal - stylesPrice;
};
const specFromLicense = license => {
  let licenseOptionSpec = {
    licenseId: license.id
  };
  if (license.variables) {
    const options = license.variables.reduce((res, variable) => {
      if (variable.options && variable.options.length > 0) {
        res[variable.id] = variable.options[0].id;
      } else {
        // FIXME this is only necessary to submit the variable,
        // since it's required to have a selection for every variable
        res[variable.id] = null;
      }
      return res;
    }, {});
    licenseOptionSpec = {
      ...licenseOptionSpec,
      licenseOptions: options
    };
  }
  return licenseOptionSpec;
};
_Precart_license2.default.hash && _Precart_license2.default.hash !== "db49280bb01f8bbef6283c0d3fba99e1" && console.error("The definition of 'Precart_license' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _Precart_license2.default;
function Precart(_ref3) {
  var _viewer$precart2, _collection$licenses3;
  let {
    viewer: viewerKey,
    collection: collectionKey,
    isLoading
  } = _ref3;
  const viewer = (0, _reactRelay.useFragment)((_Precart_viewer2.default.hash && _Precart_viewer2.default.hash !== "fe66a5cdbee38cae7bfb11922f945fd5" && console.error("The definition of 'Precart_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _Precart_viewer2.default), viewerKey);
  const collection = (0, _reactRelay.useFragment)((_Precart_collection2.default.hash && _Precart_collection2.default.hash !== "dd2c3385a0fc5149dbfa230cfa5e3e99" && console.error("The definition of 'Precart_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _Precart_collection2.default), collectionKey);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const subtotal = viewer === null || viewer === void 0 ? void 0 : (_viewer$precart2 = viewer.precart) === null || _viewer$precart2 === void 0 ? void 0 : _viewer$precart2.subtotal;
  const precartOpen = (0, _reactRedux.useSelector)(state => state.precartOpen);
  const licenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions);
  const selectedSkuIds = (0, _reactRedux.useSelector)(state => state.selectedSkuIds);
  const dispatch = (0, _reactRedux.useDispatch)();
  const onPrecartClose = () => dispatch({
    type: 'CLOSE_PRECART'
  });
  const onUnselectSkuIds = () => dispatch({
    type: 'UNSELECT_SKUS'
  });
  const [isConfirming, setIsConfirming] = (0, _react.useState)(false);
  const [error, setError] = (0, _react.useState)(null);
  const clearCart = (0, _react.useCallback)(() => {
    const mutation = (_PrecartClearCartMutation2.default.hash && _PrecartClearCartMutation2.default.hash !== "af8f953028924bdc7866e645c2d0a903" && console.error("The definition of 'PrecartClearCartMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _PrecartClearCartMutation2.default);
    (0, _reactRelay.commitMutation)(environment, {
      mutation,
      variables: {
        input: {
          orderItems: []
        }
      },
      onCompleted: () => {
        setError(null);
      }
    });
  }, [environment]);
  const onSelectLicense = (id, checked) => {
    var _collection$licenses, _license$excludedLice;
    const license = collection === null || collection === void 0 ? void 0 : (_collection$licenses = collection.licenses) === null || _collection$licenses === void 0 ? void 0 : _collection$licenses.find(node => node.id === id);
    if (!license) return;
    const licenseOptionSpec = specFromLicense(license);
    dispatch({
      type: 'PRECART_TOGGLE_LICENSE',
      licenseId: id,
      checked,
      licenseOptionSpec,
      excludedLicenseIds: (_license$excludedLice = license.excludedLicenses) === null || _license$excludedLice === void 0 ? void 0 : _license$excludedLice.map(l => l.id)
    });
  };
  const onSelectLicenseOption = (licenseId, variableId, optionId) => {
    dispatch({
      type: 'PRECART_SELECT_LICENSE_OPTION',
      licenseId,
      variableId,
      optionId
    });
  };
  const selectedItemsArray = () => Object.keys(selectedSkuIds).filter(id => selectedSkuIds[id] === true);
  const licenseSelectionsArray = () => licenseOptions.reduce((res, _ref4) => {
    let {
      licenseId,
      licenseOptions
    } = _ref4;
    const licenseVariableIds = Object.keys(licenseOptions);
    if (licenseVariableIds.length > 0) {
      return res.concat(licenseVariableIds.map(licenseVariableId => ({
        licenseId,
        licenseVariableId,
        licenseOptionId: licenseOptions[licenseVariableId]
      })));
    }
    return res.concat({
      licenseId
    });
  }, []);
  const addToCart = () => {
    const variables = {
      input: {
        skuIds: selectedItemsArray(),
        licenseSelections: licenseSelectionsArray()
      }
    };
    const mutation = (_PrecartAddToCartMutation2.default.hash && _PrecartAddToCartMutation2.default.hash !== "c5ead46ecc07099bee68d4f4f1ecb5bf" && console.error("The definition of 'PrecartAddToCartMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _PrecartAddToCartMutation2.default);
    (0, _reactRelay.commitMutation)(environment, {
      mutation,
      variables,
      onError: err => {
        console.error('Failed:', err);
        setError(err.message || 'An error occurred');
      },
      onCompleted: (res, errors) => {
        if (errors) {
          var _errors$;
          console.error('Failed:', errors);
          setError(((_errors$ = errors[0]) === null || _errors$ === void 0 ? void 0 : _errors$.message) || 'An error occurred');
        } else {
          setError(null);
          onUnselectSkuIds();
        }
      }
    });
  };
  const addToCartConfirmed = () => {
    onPrecartClose();
    addToCart();
    deactivateConfirming();
  };
  const clearCartConfirming = () => {
    onPrecartClose();
    onUnselectSkuIds();
    deactivateConfirming();
  };
  const backToBrowse = subtotal => {
    if (subtotal !== null && subtotal !== void 0 && subtotal.amount && (licenseOptions === null || licenseOptions === void 0 ? void 0 : licenseOptions.length) > 0) {
      return activateConfirming();
    }
    onUnselectSkuIds();
    return onPrecartClose();
  };
  const activateConfirming = () => {
    setIsConfirming(true);
  };
  const deactivateConfirming = () => {
    setIsConfirming(false);
  };
  (0, _react.useEffect)(() => {
    function setDataAttr() {
      requestAnimationFrame(() => {
        document.body.dataset.precartOpen = `${precartOpen}`;
      });
    }
    setDataAttr();
  }, [precartOpen]);
  (0, _react.useEffect)(() => {
    var _collection$licenses2;
    collection === null || collection === void 0 ? void 0 : (_collection$licenses2 = collection.licenses) === null || _collection$licenses2 === void 0 ? void 0 : _collection$licenses2.filter(license => license.defaultSelected).forEach(license => {
      var _license$excludedLice2;
      return dispatch({
        type: 'PRECART_SELECT_LICENSE',
        licenseOptionSpec: specFromLicense(license),
        excludedLicenseIds: (_license$excludedLice2 = license.excludedLicenses) === null || _license$excludedLice2 === void 0 ? void 0 : _license$excludedLice2.map(l => l.id)
      });
    });
  }, []);
  function renderConfirmingModal(subtotal) {
    var _viewer$precart3, _viewer$precart3$skus;
    if (!isConfirming) {
      return null;
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart-confirm"
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: "pre-cart-confirm__header"
    }, "Do you want to add your selection to your cart?"), /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart-confirm__buttons"
    }, /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      className: "pre-cart-confirm__button",
      "data-label": "confirm",
      onClick: addToCartConfirmed
    }, /*#__PURE__*/_react.default.createElement("div", null, "Add selection to cart:", /*#__PURE__*/_react.default.createElement("br", null), "(", (_viewer$precart3 = viewer.precart) === null || _viewer$precart3 === void 0 ? void 0 : (_viewer$precart3$skus = _viewer$precart3.skus) === null || _viewer$precart3$skus === void 0 ? void 0 : _viewer$precart3$skus.map(skuName).join(', '), ")", /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement(_Price.Price, {
      price: subtotal
    }), /*#__PURE__*/_react.default.createElement("span", {
      className: "pre-cart-confirm__button__arrow"
    }, " \u2192"))), /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      className: "pre-cart-confirm__button",
      "data-label": "clear",
      onClick: clearCartConfirming
    }, "Clear selection")));
  }
  function renderPreCartBottom(subtotal) {
    const licenseSelected = licenseOptions.length > 0;
    if (!subtotal || subtotal.amount === 0) {
      return null;
    }
    const bundleDiscounts = calculateBundleDiscounts(viewer);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__bottom"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__section pre-cart__your-selection"
    }, /*#__PURE__*/_react.default.createElement("h3", {
      className: "pre-cart__your-selection__label"
    }, "Your Selection"), /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__your-selection__items"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__your-selection__item"
    }, /*#__PURE__*/_react.default.createElement("div", null, collection === null || collection === void 0 ? void 0 : collection.name, /*#__PURE__*/_react.default.createElement("br", null), (0, _utils.pluralize)(countStyles(viewer), 'font')), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_Price.Price, {
      price: {
        amount: countStylesPrice(viewer),
        currency: subtotal.currency
      }
    }))), bundleDiscounts < 0 && /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__your-selection__item"
    }, "Bundle discounts:", /*#__PURE__*/_react.default.createElement(_Price.Price, {
      price: {
        amount: bundleDiscounts,
        currency: subtotal.currency
      }
    })))), licenseSelected ? /*#__PURE__*/_react.default.createElement(_ComponentsContext.default.Consumer, null, components => components.PrecartCta && /*#__PURE__*/_react.default.createElement(components.PrecartCta, {
      onClick: addToCartConfirmed,
      subtotal: subtotal
    })) : /*#__PURE__*/_react.default.createElement("div", {
      className: "pre-cart__cta",
      "data-label": "select-license"
    }, "Please select a license"));
  }
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart",
    "data-open": precartOpen,
    "data-loading": isLoading
  }, /*#__PURE__*/_react.default.createElement(_ComponentsContext.default.Consumer, null, components => components.PrecartTopBar && /*#__PURE__*/_react.default.createElement(components.PrecartTopBar, {
    backToBrowse: () => backToBrowse(subtotal)
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__middle pre-cart__licenses"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__section"
  }, /*#__PURE__*/_react.default.createElement("ul", null, collection === null || collection === void 0 ? void 0 : (_collection$licenses3 = collection.licenses) === null || _collection$licenses3 === void 0 ? void 0 : _collection$licenses3.map(license => /*#__PURE__*/_react.default.createElement("li", {
    key: license.id
  }, /*#__PURE__*/_react.default.createElement(_License.default, {
    node: license,
    selected: licenseOptions.some(_ref5 => {
      let {
        licenseId
      } = _ref5;
      return licenseId === license.id;
    }),
    onSelectLicense: onSelectLicense,
    onSelectLicenseOption: onSelectLicenseOption
  }))))), selectedItemsArray().length === 0 && /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__section pre-cart__selection-empty"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "pre-cart__selection-empty__arrow"
  }, '\u25c0 '), "Start selecting fonts to add to your cart")), renderPreCartBottom(subtotal), error && /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__error"
  }, error, ' ', /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: clearCart,
    className: "pre-cart__error-button"
  }, "Empty cart and try again")), renderConfirmingModal(subtotal)));
}
function PrecartQueryRenderer(_ref6) {
  let {
    collectionId,
    ...rest
  } = _ref6;
  const selectedSkuIdsObject = (0, _reactRedux.useSelector)(state => state.selectedSkuIds);
  const selectedSkuIds = Object.keys(selectedSkuIdsObject).filter(id => selectedSkuIdsObject[id] === true);
  const deferredSelectedSkuIds = (0, _react.useDeferredValue)(selectedSkuIds);
  const licenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions);
  const licenseOptionsSpecs = licenseOptions.map(spec => ({
    licenseId: spec.licenseId,
    licenseOptionIds: Object.values(spec.licenseOptions).filter(Boolean)
  }));

  // useDeferredValue allows us to show the stale data instead of falling back
  // to the Suspense fallback in the parent, basically unmounting everything
  const deferredLicenseOptionsSpecs = (0, _react.useDeferredValue)(licenseOptionsSpecs);
  const data = (0, _reactRelay.useLazyLoadQuery)((_PrecartQuery2.default.hash && _PrecartQuery2.default.hash !== "517fbe407b2c13334c4101126038982d" && console.error("The definition of 'PrecartQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _PrecartQuery2.default), {
    licenseOptionsSpecs: deferredLicenseOptionsSpecs,
    selectedSkuIds: deferredSelectedSkuIds,
    collectionId
  });
  return /*#__PURE__*/_react.default.createElement(Precart, _extends({}, data, rest, {
    isLoading: licenseOptionsSpecs !== deferredLicenseOptionsSpecs || selectedSkuIds !== deferredSelectedSkuIds
  }));
}
var _default = PrecartQueryRenderer;
exports.default = _default;
