// fontdue-js/dist/components/Precart/License.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _License_node2 = _interopRequireDefault(require("../../__generated__/License_node.graphql"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _react = _interopRequireDefault(require("react"));
var _Icons = require("../Icons");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const License = _ref => {
  let {
    node: nodeKey,
    selected,
    onSelectLicense,
    onSelectLicenseOption
  } = _ref;
  const node = (0, _reactRelay.useFragment)((_License_node2.default.hash && _License_node2.default.hash !== "1550b152731d8412ffc55716cb536407" && console.error("The definition of 'License_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _License_node2.default), nodeKey);
  if (!node) return null;
  const selectedLicenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions);
  const selectedOptions = selectedLicenseOptions.find(spec => spec.licenseId === node.id);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__license-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__license",
    "data-selected": selected
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__license__name-wrapper"
  }, /*#__PURE__*/_react.default.createElement("label", {
    className: "pre-cart__license__name"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__license__checkbox"
  }, selected && /*#__PURE__*/_react.default.createElement(_Icons.Check, null)), node.name, /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    checked: selected,
    value: node.id,
    onChange: e => onSelectLicense(node.id, e.target.checked)
  })), selected && /*#__PURE__*/_react.default.createElement("a", {
    href: node.url,
    "aria-label": "Info",
    className: "pre-cart__license__icon",
    "data-label": "info"
  }, /*#__PURE__*/_react.default.createElement(_Icons.CircledInfo, null))), selected && /*#__PURE__*/_react.default.createElement("ul", null, node.variables && node.variables.map(variable => {
    var _variable$options, _variable$options$fin;
    const selectedLicenseOption = selectedOptions === null || selectedOptions === void 0 ? void 0 : selectedOptions.licenseOptions[variable.id];
    if (variable.variableType === 'text') return null;
    const selectedLicenseAmount = variable === null || variable === void 0 ? void 0 : (_variable$options = variable.options) === null || _variable$options === void 0 ? void 0 : (_variable$options$fin = _variable$options.find(option => option.id === selectedLicenseOption)) === null || _variable$options$fin === void 0 ? void 0 : _variable$options$fin.amount;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: variable.id
    }, /*#__PURE__*/_react.default.createElement("label", {
      className: "pre-cart__license__option"
    }, "\u21B3 ", selectedLicenseAmount, /*#__PURE__*/_react.default.createElement(_Icons.CarrotDown, null), /*#__PURE__*/_react.default.createElement("select", {
      value: selectedLicenseOption || '',
      onChange: e => onSelectLicenseOption(node.id, variable.id, e.target.value)
    }, /*#__PURE__*/_react.default.createElement("option", {
      value: "",
      disabled: true
    }, "Select one"), variable.options && variable.options.map(option => /*#__PURE__*/_react.default.createElement("option", {
      key: option.id,
      value: option.id
    }, option.amount)))));
  }))), node.description && /*#__PURE__*/_react.default.createElement("div", {
    className: "pre-cart__license__description",
    dangerouslySetInnerHTML: {
      __html: node.description
    }
  }));
};
var _default = License;
exports.default = _default;
