// fontdue-js/dist/components/ConfigContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeConfig = exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const makeTypeTesterConfig = config => {
  var _config$openTypeFeatu, _config$openTypeFeatu2, _config$openTypeFeatu3, _config$openTypeFeatu4, _config$size, _config$size2, _config$size3, _config$lineHeight, _config$lineHeight2, _config$lineHeight3, _config$letterSpacing, _config$letterSpacing2, _config$letterSpacing3, _config$columns, _config$columns2, _config$columns3;
  let shy = (config === null || config === void 0 ? void 0 : config.shy) ?? false;
  if (shy === true) shy = 'hover';
  let toolsPosition = (config === null || config === void 0 ? void 0 : config.toolsPosition) ?? 'inline';
  if (shy !== 'focus') toolsPosition = 'inline';
  return {
    autofitOnChange: (config === null || config === void 0 ? void 0 : config.autofitOnChange) ?? false,
    truncate: (config === null || config === void 0 ? void 0 : config.truncate) ?? false,
    priceText: (config === null || config === void 0 ? void 0 : config.priceText) ?? false,
    selectButtonStyle: (config === null || config === void 0 ? void 0 : config.selectButtonStyle) ?? 'inline',
    selectButtonLabel: (config === null || config === void 0 ? void 0 : config.selectButtonLabel) ?? 'Select',
    selectButton: (config === null || config === void 0 ? void 0 : config.selectButton) ?? (config === null || config === void 0 ? void 0 : config.buyButton) ?? true,
    shy,
    toolsPosition,
    selectable: (config === null || config === void 0 ? void 0 : config.selectable) ?? false,
    priceBar: (config === null || config === void 0 ? void 0 : config.priceBar) ?? false,
    textInput: (config === null || config === void 0 ? void 0 : config.textInput) ?? false,
    initialMode: (config === null || config === void 0 ? void 0 : config.initialMode) ?? 'local',
    groupEdit: (config === null || config === void 0 ? void 0 : config.groupEdit) ?? false,
    bulletStyle: (config === null || config === void 0 ? void 0 : config.bulletStyle) ?? 'square',
    openTypeFeatures: {
      buttonLabel: (config === null || config === void 0 ? void 0 : (_config$openTypeFeatu = config.openTypeFeatures) === null || _config$openTypeFeatu === void 0 ? void 0 : _config$openTypeFeatu.buttonLabel) ?? 'OT Features',
      interactionStyle: (config === null || config === void 0 ? void 0 : (_config$openTypeFeatu2 = config.openTypeFeatures) === null || _config$openTypeFeatu2 === void 0 ? void 0 : _config$openTypeFeatu2.interactionStyle) ?? 'panel',
      columns: config === null || config === void 0 ? void 0 : (_config$openTypeFeatu3 = config.openTypeFeatures) === null || _config$openTypeFeatu3 === void 0 ? void 0 : _config$openTypeFeatu3.columns,
      selectionStyle: (config === null || config === void 0 ? void 0 : (_config$openTypeFeatu4 = config.openTypeFeatures) === null || _config$openTypeFeatu4 === void 0 ? void 0 : _config$openTypeFeatu4.selectionStyle) ?? 'bullet'
    },
    variableAxesPosition: (config === null || config === void 0 ? void 0 : config.variableAxesPosition) ?? 'auto',
    alignmentButtons: (config === null || config === void 0 ? void 0 : config.alignmentButtons) ?? false,
    initialAlignment: (config === null || config === void 0 ? void 0 : config.initialAlignment) ?? 'left',
    size: {
      label: (config === null || config === void 0 ? void 0 : (_config$size = config.size) === null || _config$size === void 0 ? void 0 : _config$size.label) ?? 'Size',
      min: (config === null || config === void 0 ? void 0 : (_config$size2 = config.size) === null || _config$size2 === void 0 ? void 0 : _config$size2.min) ?? (config === null || config === void 0 ? void 0 : config.min) ?? 10,
      max: (config === null || config === void 0 ? void 0 : (_config$size3 = config.size) === null || _config$size3 === void 0 ? void 0 : _config$size3.max) ?? (config === null || config === void 0 ? void 0 : config.max) ?? 200
    },
    lineHeight: config !== null && config !== void 0 && config.lineHeight ? {
      label: (config === null || config === void 0 ? void 0 : (_config$lineHeight = config.lineHeight) === null || _config$lineHeight === void 0 ? void 0 : _config$lineHeight.label) ?? 'Line height',
      min: (config === null || config === void 0 ? void 0 : (_config$lineHeight2 = config.lineHeight) === null || _config$lineHeight2 === void 0 ? void 0 : _config$lineHeight2.min) ?? 0.7,
      max: (config === null || config === void 0 ? void 0 : (_config$lineHeight3 = config.lineHeight) === null || _config$lineHeight3 === void 0 ? void 0 : _config$lineHeight3.max) ?? 2
    } : undefined,
    letterSpacing: config !== null && config !== void 0 && config.letterSpacing ? {
      label: (config === null || config === void 0 ? void 0 : (_config$letterSpacing = config.letterSpacing) === null || _config$letterSpacing === void 0 ? void 0 : _config$letterSpacing.label) ?? 'Letter spacing',
      min: (config === null || config === void 0 ? void 0 : (_config$letterSpacing2 = config.letterSpacing) === null || _config$letterSpacing2 === void 0 ? void 0 : _config$letterSpacing2.min) ?? -0.05,
      max: (config === null || config === void 0 ? void 0 : (_config$letterSpacing3 = config.letterSpacing) === null || _config$letterSpacing3 === void 0 ? void 0 : _config$letterSpacing3.max) ?? 0.2
    } : undefined,
    columns: (config === null || config === void 0 ? void 0 : config.columns) === false ? false : {
      count: (config === null || config === void 0 ? void 0 : (_config$columns = config.columns) === null || _config$columns === void 0 ? void 0 : _config$columns.count) ?? 'auto',
      width: (config === null || config === void 0 ? void 0 : (_config$columns2 = config.columns) === null || _config$columns2 === void 0 ? void 0 : _config$columns2.width) ?? '40ch',
      gap: (config === null || config === void 0 ? void 0 : (_config$columns3 = config.columns) === null || _config$columns3 === void 0 ? void 0 : _config$columns3.gap) ?? '1em'
    }
  };
};
const makeConfig = config => {
  var _config$form, _config$storeModal, _config$storeModal2, _config$storeModal3, _config$storeModal4, _config$stripe, _config$tracking, _config$tracking2, _config$tracking3, _config$tracking4;
  return {
    typeTester: makeTypeTesterConfig(config === null || config === void 0 ? void 0 : config.typeTester),
    form: {
      checkboxStyle: (config === null || config === void 0 ? void 0 : (_config$form = config.form) === null || _config$form === void 0 ? void 0 : _config$form.checkboxStyle) ?? 'check'
    },
    storeModal: {
      indexLayout: (config === null || config === void 0 ? void 0 : (_config$storeModal = config.storeModal) === null || _config$storeModal === void 0 ? void 0 : _config$storeModal.indexLayout) ?? 'styled-aa',
      indexExcludeTags: config === null || config === void 0 ? void 0 : (_config$storeModal2 = config.storeModal) === null || _config$storeModal2 === void 0 ? void 0 : _config$storeModal2.indexExcludeTags,
      productLicensesPosition: (config === null || config === void 0 ? void 0 : (_config$storeModal3 = config.storeModal) === null || _config$storeModal3 === void 0 ? void 0 : _config$storeModal3.productLicensesPosition) ?? 'top',
      hideAllCollections: (config === null || config === void 0 ? void 0 : (_config$storeModal4 = config.storeModal) === null || _config$storeModal4 === void 0 ? void 0 : _config$storeModal4.hideAllCollections) ?? false
    },
    stripe: {
      appearance: (config === null || config === void 0 ? void 0 : (_config$stripe = config.stripe) === null || _config$stripe === void 0 ? void 0 : _config$stripe.appearance) ?? null
    },
    tracking: {
      enabled: (config === null || config === void 0 ? void 0 : (_config$tracking = config.tracking) === null || _config$tracking === void 0 ? void 0 : _config$tracking.enabled) ?? false,
      // Explicit opt-in required
      consentRequired: (config === null || config === void 0 ? void 0 : (_config$tracking2 = config.tracking) === null || _config$tracking2 === void 0 ? void 0 : _config$tracking2.consentRequired) ?? false,
      consentMessage: config === null || config === void 0 ? void 0 : (_config$tracking3 = config.tracking) === null || _config$tracking3 === void 0 ? void 0 : _config$tracking3.consentMessage,
      segment: config === null || config === void 0 ? void 0 : (_config$tracking4 = config.tracking) === null || _config$tracking4 === void 0 ? void 0 : _config$tracking4.segment
    },
    corsErrorModal: (config === null || config === void 0 ? void 0 : config.corsErrorModal) ?? true
  };
};
exports.makeConfig = makeConfig;
var _default = /*#__PURE__*/_react.default.createContext(makeConfig());
exports.default = _default;
