// fontdue-js/dist/components/CorsErrorModal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderCorsErrorModal = renderCorsErrorModal;
var _react = _interopRequireDefault(require("react"));
var _client = require("react-dom/client");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const code = {
  background: '#fef2f2',
  padding: '2px 6px',
  borderRadius: 3,
  fontSize: 13
};
function CorsErrorModal(_ref) {
  let {
    origin,
    fontdueUrl
  } = _ref;
  let settingsOrigin = fontdueUrl;
  try {
    settingsOrigin = new URL(fontdueUrl).origin;
  } catch {
    // keep as-is
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'fixed',
      inset: 0,
      zIndex: 99999,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      background: 'rgba(0,0,0,0.4)',
      fontFamily: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif',
      fontSize: 14,
      lineHeight: 1.6
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: 'white',
      color: '#1f2937',
      borderRadius: 12,
      border: '2px solid #fca5a5',
      padding: '28px 32px',
      maxWidth: 520,
      width: '90%',
      boxShadow: '0 8px 30px rgba(0,0,0,0.12)',
      textAlign: 'left'
    }
  }, /*#__PURE__*/_react.default.createElement("strong", {
    style: {
      fontSize: 16,
      color: '#dc2626'
    }
  }, "Cross-origin request blocked"), /*#__PURE__*/_react.default.createElement("p", {
    style: {
      margin: '12px 0 0'
    }
  }, "Your website (", /*#__PURE__*/_react.default.createElement("code", {
    style: code
  }, origin), ") is not allowed to connect to your Fontdue store."), /*#__PURE__*/_react.default.createElement("p", {
    style: {
      margin: '12px 0 0'
    }
  }, "To fix this, go to", ' ', /*#__PURE__*/_react.default.createElement("a", {
    href: `${settingsOrigin}/admin/settings/security`,
    target: "_blank",
    rel: "noreferrer",
    style: {
      color: '#dc2626',
      fontWeight: 500
    }
  }, "Settings \u2192 Security"), ' ', "in your Fontdue dashboard, and add ", /*#__PURE__*/_react.default.createElement("code", {
    style: code
  }, origin), " to the", ' ', /*#__PURE__*/_react.default.createElement("strong", null, "Cross-origin API access"), " field."), /*#__PURE__*/_react.default.createElement("p", {
    style: {
      margin: '16px 0 0',
      color: '#6b7280',
      fontSize: 13
    }
  }, "This page will reload automatically once the origin is added.")));
}
function mount(origin, fontdueUrl) {
  const container = document.createElement('div');
  document.body.appendChild(container);
  (0, _client.createRoot)(container).render( /*#__PURE__*/_react.default.createElement(CorsErrorModal, {
    origin: origin,
    fontdueUrl: fontdueUrl
  }));
}
function renderCorsErrorModal(origin, fontdueUrl) {
  if (document.body) {
    mount(origin, fontdueUrl);
  } else {
    document.addEventListener('DOMContentLoaded', () => mount(origin, fontdueUrl));
  }
}
