// fontdue-js/dist/components/TestFontsForm/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TestFontsFormPreloadedQueryRenderer = TestFontsFormPreloadedQueryRenderer;
exports.default = TestFontsForm;
var _TestFontsForm_Query2 = _interopRequireDefault(require("../../__generated__/TestFontsForm_Query.graphql"));
var _TestFontsFormUpdateCustomerMutation2 = _interopRequireDefault(require("../../__generated__/TestFontsFormUpdateCustomerMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactGoogleRecaptcha = _interopRequireDefault(require("react-google-recaptcha"));
var _TextField = _interopRequireDefault(require("../TextField"));
var _Icons = require("../Icons");
var _Checkbox = _interopRequireDefault(require("../Checkbox"));
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const updateCustomerMutation = (_TestFontsFormUpdateCustomerMutation2.default.hash && _TestFontsFormUpdateCustomerMutation2.default.hash !== "ba56958399f0893bd667ff02c33a6975" && console.error("The definition of 'TestFontsFormUpdateCustomerMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TestFontsFormUpdateCustomerMutation2.default);
const TestFontsDownloading = _ref => {
  let {
    archiveUrl
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__downloading"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__downloading__icon"
  }, /*#__PURE__*/_react.default.createElement(_Icons.DownloadFonts, null)), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, "The download will start automatically."), /*#__PURE__*/_react.default.createElement("p", null, "If it didn\u2019t start, click ", /*#__PURE__*/_react.default.createElement("a", {
    href: archiveUrl
  }, "here"), ".")));
};
const TestFontsFormComponent = _ref2 => {
  var _data$viewer$settings, _data$viewer$settings2, _data$viewer$testFont;
  let {
    data,
    agreementLabel,
    downloadLabel = 'Download test fonts',
    newsletterCheckboxChecked = false
  } = _ref2;
  const [email, setEmail] = (0, _react.useState)('');
  const [name, setName] = (0, _react.useState)('');
  const [newsletterOptIn, setNewsletterOptIn] = (0, _react.useState)(newsletterCheckboxChecked);
  const [error, setError] = (0, _react.useState)(null);
  const [eulaAgreed, setEulaAgreed] = (0, _react.useState)(false);
  const [eulaFail, setEulaFail] = (0, _react.useState)(false);
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const [submitted, setSubmitted] = (0, _react.useState)(false);
  const [recaptchaToken, setRecaptchaToken] = (0, _react.useState)(null);
  const [pendingSubmit, setPendingSubmit] = (0, _react.useState)(false);
  const downloadForm = (0, _react.useRef)(null);
  const recaptchaRef = (0, _react.useRef)(null);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  if (!data.viewer) return null;
  const recaptchaEnabled = ((_data$viewer$settings = data.viewer.settings) === null || _data$viewer$settings === void 0 ? void 0 : _data$viewer$settings.recaptchaEnabled) ?? false;
  const recaptchaSiteKey = (_data$viewer$settings2 = data.viewer.settings) === null || _data$viewer$settings2 === void 0 ? void 0 : _data$viewer$settings2.recaptchaSiteKey;
  const submitForm = (0, _react.useCallback)(token => {
    setSubmitting(true);
    setPendingSubmit(false);
    setError(null);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: updateCustomerMutation,
      variables: {
        input: {
          name,
          email,
          newsletterOptIn,
          recaptchaToken: token ?? undefined
        }
      },
      onCompleted: (res, errors) => {
        if (errors && errors.length > 0) {
          var _recaptchaRef$current;
          setError(errors[0].message);
          setSubmitting(false);
          // Reset reCAPTCHA on error so user can retry
          (_recaptchaRef$current = recaptchaRef.current) === null || _recaptchaRef$current === void 0 ? void 0 : _recaptchaRef$current.reset();
          setRecaptchaToken(null);
        } else {
          var _downloadForm$current;
          (_downloadForm$current = downloadForm.current) === null || _downloadForm$current === void 0 ? void 0 : _downloadForm$current.submit();
          setSubmitted(true);
          setSubmitting(false);
          setError(null);
        }
      },
      onError: _ref3 => {
        var _recaptchaRef$current2;
        let {
          message
        } = _ref3;
        setError(message);
        setSubmitting(false);
        // Reset reCAPTCHA on error so user can retry
        (_recaptchaRef$current2 = recaptchaRef.current) === null || _recaptchaRef$current2 === void 0 ? void 0 : _recaptchaRef$current2.reset();
        setRecaptchaToken(null);
      }
    });
  }, [environment, name, email, newsletterOptIn]);
  const handleRecaptchaChange = (0, _react.useCallback)(token => {
    setRecaptchaToken(token);
    // If we were waiting for a token to submit, do it now
    if (token && pendingSubmit) {
      submitForm(token);
    }
  }, [pendingSubmit, submitForm]);
  const handleRecaptchaExpired = (0, _react.useCallback)(() => {
    setRecaptchaToken(null);
  }, []);
  const handleSubmit = e => {
    e.preventDefault();
    if (!eulaAgreed) {
      setEulaFail(true);
      return;
    }

    // If reCAPTCHA is enabled but no token, execute it and wait for callback
    if (recaptchaEnabled && recaptchaSiteKey && !recaptchaToken) {
      var _recaptchaRef$current3;
      setPendingSubmit(true);
      setError(null);
      (_recaptchaRef$current3 = recaptchaRef.current) === null || _recaptchaRef$current3 === void 0 ? void 0 : _recaptchaRef$current3.execute();
      return;
    }
    submitForm(recaptchaToken);
  };

  // Button is disabled only while actively submitting or waiting for reCAPTCHA
  const disabled = !eulaAgreed || !email || !name;
  const settings = data.viewer.settings;
  const newsletterOptInLabel = settings === null || settings === void 0 ? void 0 : settings.newsletterOptInLabel;
  const eulaAgreementLabel = settings === null || settings === void 0 ? void 0 : settings.eulaAgreementLabel;
  const archiveUrl = (_data$viewer$testFont = data.viewer.testFonts) === null || _data$viewer$testFont === void 0 ? void 0 : _data$viewer$testFont.archiveUrl;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts"
  }, error && /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__errors"
  }, error), submitted ? /*#__PURE__*/_react.default.createElement(TestFontsDownloading, {
    archiveUrl: archiveUrl
  }) : /*#__PURE__*/_react.default.createElement("form", {
    className: "test-fonts__form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Name *",
    value: name,
    onChange: value => setName(value),
    autocomplete: "name",
    placeholder: "Your name"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Email *",
    type: "email",
    value: email,
    onChange: value => setEmail(value),
    autocomplete: "email",
    placeholder: "Your email"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__agreement test-fonts__checkbox-field",
    "data-highlighted": eulaFail && !eulaAgreed
  }, /*#__PURE__*/_react.default.createElement(_Checkbox.default, {
    id: "test-fonts-eula-agreement",
    checked: eulaAgreed,
    onChange: event => setEulaAgreed(event.target.checked)
  }), /*#__PURE__*/_react.default.createElement("label", {
    htmlFor: "test-fonts-eula-agreement",
    dangerouslySetInnerHTML: {
      __html: (agreementLabel || eulaAgreementLabel) ?? ''
    }
  })), newsletterOptInLabel && /*#__PURE__*/_react.default.createElement("div", {
    className: "test-fonts__newsletter-opt-in test-fonts__checkbox-field"
  }, /*#__PURE__*/_react.default.createElement(_Checkbox.default, {
    id: "newsletter-opt-in",
    checked: newsletterOptIn,
    onChange: event => setNewsletterOptIn(event.target.checked)
  }), /*#__PURE__*/_react.default.createElement("label", {
    htmlFor: "newsletter-opt-in",
    dangerouslySetInnerHTML: {
      __html: newsletterOptInLabel
    }
  })), recaptchaEnabled && recaptchaSiteKey && /*#__PURE__*/_react.default.createElement(_reactGoogleRecaptcha.default, {
    ref: recaptchaRef,
    sitekey: recaptchaSiteKey,
    size: "invisible",
    onChange: handleRecaptchaChange,
    onExpired: handleRecaptchaExpired
  }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("button", {
    className: "submit-button",
    type: "submit",
    disabled: submitting || pendingSubmit || disabled
  }, submitting || pendingSubmit ? 'Submitting...' : downloadLabel, /*#__PURE__*/_react.default.createElement("span", {
    className: "submit-button__arrow"
  }, " \u2192")))), /*#__PURE__*/_react.default.createElement("form", {
    action: archiveUrl ?? undefined,
    ref: downloadForm
  }));
};
const query = (_TestFontsForm_Query2.default.hash && _TestFontsForm_Query2.default.hash !== "cd43f0cacc4dcf01cf94fb1ff97197ca" && console.error("The definition of 'TestFontsForm_Query' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TestFontsForm_Query2.default);
function TestFontsFormPreloadedQueryRenderer(_ref4) {
  let {
    preloadedQuery,
    ...rest
  } = _ref4;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(query, queryRef);
  return /*#__PURE__*/_react.default.createElement(TestFontsFormComponent, _extends({
    data: data
  }, rest));
}
function TestFontsForm(props) {
  const data = (0, _reactRelay.useLazyLoadQuery)(query, {});
  return /*#__PURE__*/_react.default.createElement(TestFontsFormComponent, _extends({
    data: data
  }, props));
}
