// fontdue-js/dist/components/Price/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Price = Price;
exports.default = exports.PriceFormat = void 0;
var _Price_price2 = _interopRequireDefault(require("../../__generated__/Price_price.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
let PriceFormat;
exports.PriceFormat = PriceFormat;
(function (PriceFormat) {
  PriceFormat["Default"] = "default";
  PriceFormat["Truncated"] = "truncated";
})(PriceFormat || (exports.PriceFormat = PriceFormat = {}));
const currencySymbols = {
  USD: '$',
  AUD: '$',
  CAD: '$',
  EUR: '€',
  GBP: '£',
  JPY: '¥',
  INR: '₹',
  BRL: 'R$',
  CHF: 'CHF '
};
const formatMoney = (priceInCents, format, currency) => {
  let value = priceInCents / 100;
  let precision = format === PriceFormat.Truncated ? 0 : 2;
  if (currency === 'JPY') {
    // always show yen as "cents" without any fraction
    value = priceInCents;
    precision = 0;
  }
  try {
    const formatter = new window.Intl.NumberFormat(window.navigator.language, {
      style: 'currency',
      currency,
      currencyDisplay: 'symbol',
      minimumFractionDigits: precision
    });
    return formatter.format(value);
  } catch (e) {
    // fallback for browsers without support for Intl.NumberFormat
    const symbol = currencySymbols[currency] ?? '';
    return `${symbol}${value.toFixed(precision)}`;
  }
};
const calculateAmount = _ref => {
  let {
    amount,
    invert,
    subtractFrom,
    multiplier
  } = _ref;
  let subtracted = amount;
  if (subtractFrom) {
    subtracted = subtractFrom - amount;
  }
  if (invert) {
    return subtracted * -1 * multiplier;
  }
  return subtracted * multiplier;
};
function Price(_ref2) {
  let {
    price,
    format = PriceFormat.Default,
    invert = false,
    // multiply by -1
    subtractFrom,
    // amount to subtract from
    multiplier = 1,
    // multiply the result amount?
    isLoading = false,
    ...rest
  } = _ref2;
  let priceBeforeDiscount = null;
  let priceFormatted = '-';
  if (price) {
    priceFormatted = formatMoney(calculateAmount({
      amount: price.amount,
      subtractFrom: subtractFrom === null || subtractFrom === void 0 ? void 0 : subtractFrom.amount,
      invert,
      multiplier
    }), format, price.currency);
    priceBeforeDiscount = formatMoney(calculateAmount({
      amount: price.amount,
      subtractFrom: subtractFrom === null || subtractFrom === void 0 ? void 0 : subtractFrom.amount,
      invert,
      multiplier: 1
    }), format, price.currency);
  }
  return /*#__PURE__*/_react.default.createElement("span", _extends({
    className: "price"
  }, rest, {
    "data-loading": isLoading
  }), priceFormatted !== priceBeforeDiscount ? '+' : null, priceFormatted);
}
function RelayPrice(_ref3) {
  let {
    price: priceKey,
    ...rest
  } = _ref3;
  const price = (0, _reactRelay.useFragment)((_Price_price2.default.hash && _Price_price2.default.hash !== "7d3425945fe5f6df382005630e30ce86" && console.error("The definition of 'Price_price' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _Price_price2.default), priceKey);
  return /*#__PURE__*/_react.default.createElement(Price, _extends({
    price: price
  }, rest));
}
var _default = RelayPrice;
exports.default = _default;
