// fontdue-js/dist/components/StoreModalProductSummary/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StoreModalProductSummary_viewer2 = _interopRequireDefault(require("../../__generated__/StoreModalProductSummary_viewer.graphql"));
var _StoreModalProductSummaryClearCartMutation2 = _interopRequireDefault(require("../../__generated__/StoreModalProductSummaryClearCartMutation.graphql"));
var _StoreModalProductSummaryAddToCartMutation2 = _interopRequireDefault(require("../../__generated__/StoreModalProductSummaryAddToCartMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _Price = _interopRequireWildcard(require("../Price"));
var _Button = _interopRequireDefault(require("../elements/Button"));
var _StoreModalProductSummaryRefetchQuery = _interopRequireDefault(require("../../__generated__/StoreModalProductSummaryRefetchQuery.graphql"));
var _utils = require("../../utils");
var _useLicenseAndOrderVariables = require("../../hooks/useLicenseAndOrderVariables");
var _useRefetchOnLicenseChanges = require("../../hooks/useRefetchOnLicenseChanges");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const addToCartMutation = (_StoreModalProductSummaryAddToCartMutation2.default.hash && _StoreModalProductSummaryAddToCartMutation2.default.hash !== "91ea762e3842af2919da3a4c710a48cc" && console.error("The definition of 'StoreModalProductSummaryAddToCartMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductSummaryAddToCartMutation2.default);
const clearCartMutation = (_StoreModalProductSummaryClearCartMutation2.default.hash && _StoreModalProductSummaryClearCartMutation2.default.hash !== "b5c37f74432030a297c1bcf2be63b6f7" && console.error("The definition of 'StoreModalProductSummaryClearCartMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductSummaryClearCartMutation2.default);
const countStyles = viewer => {
  const {
    precart
  } = viewer;
  if (!precart || !precart.skus) return 0;
  return precart.skus.reduce((acc, sku) => {
    if (!sku.product) return 0;
    if ('totalStyles' in sku.product && sku.product.totalStyles) {
      return acc + sku.product.totalStyles;
    }
    return acc + 1;
  }, 0);
};
const countStylesPrice = viewer => {
  const {
    precart
  } = viewer;
  if (!precart || !precart.skus) return 0;
  return precart.skus.reduce((acc, sku) => {
    if (!sku || !sku.product) return acc;
    if ('totalStylesPrice' in sku.product && sku.product.totalStylesPrice) {
      return acc + sku.product.totalStylesPrice.amount;
    }
    return acc + sku.price.amount;
  }, 0);
};
const StoreModalProductSummary = _ref => {
  var _viewer$orderVariable;
  let {
    viewer: viewerKey
  } = _ref;
  const dispatch = (0, _reactRedux.useDispatch)();
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const selectedSkuIds = (0, _reactRedux.useSelector)(state => state.selectedSkuIds, _reactRedux.shallowEqual);
  const {
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  const licenseeIsBillingIdentity = (0, _reactRedux.useSelector)(state => state.licenseeIsBillingIdentity);
  const licenseOptions = (0, _reactRedux.useSelector)(state => state.licenseOptions, _reactRedux.shallowEqual);
  const selectedSkuIdsArray = (0, _react.useMemo)(() => Object.keys(selectedSkuIds).filter(id => selectedSkuIds[id] === true), [selectedSkuIds]);
  const refetchVariables = (0, _react.useMemo)(() => ({
    selectedSkuIds: selectedSkuIdsArray
  }), [selectedSkuIdsArray]);
  const [viewer, refetch] = (0, _reactRelay.useRefetchableFragment)((_StoreModalProductSummary_viewer2.default.hash && _StoreModalProductSummary_viewer2.default.hash !== "81a5f120d76cbb84753d158a7ff38c19" && console.error("The definition of 'StoreModalProductSummary_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StoreModalProductSummary_viewer2.default), viewerKey);
  (0, _useRefetchOnLicenseChanges.useRefetchOnLicenseChanges)({
    environment,
    refetch,
    query: _StoreModalProductSummaryRefetchQuery.default,
    variables: refetchVariables
  });
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const [error, setError] = (0, _react.useState)(null);
  const [commitAddToCart] = (0, _reactRelay.useMutation)(addToCartMutation);
  const clearCart = (0, _react.useCallback)(() => {
    (0, _reactRelay.commitMutation)(environment, {
      mutation: clearCartMutation,
      variables: {
        input: {
          orderItems: []
        }
      },
      onCompleted: () => {
        setError(null);
      },
      updater: store => {
        store.invalidateStore();
      }
    });
  }, [environment]);
  const handleAddToCart = (0, _react.useCallback)(() => {
    setSubmitting(true);
    const variables = {
      input: {
        skuIds: Object.keys(selectedSkuIds).filter(id => selectedSkuIds[id] === true),
        licenseSelections: licenseOptions.reduce((res, _ref2) => {
          let {
            licenseId,
            licenseOptions
          } = _ref2;
          const licenseVariableIds = Object.keys(licenseOptions);
          if (licenseVariableIds.length > 0) {
            return res.concat(licenseVariableIds.map(licenseVariableId => ({
              licenseId,
              licenseVariableId,
              licenseOptionId: licenseOptions[licenseVariableId]
            })));
          }
          return res.concat({
            licenseId
          });
        }, []),
        orderVariableSelections,
        licenseeIsBillingIdentity
      }
    };
    commitAddToCart({
      variables,
      onError: err => {
        console.error('Failed:', err);
        setError(err.message || 'An error occurred');
        setSubmitting(false);
      },
      onCompleted: (_res, errors) => {
        if (errors) {
          var _errors$;
          console.error('Failed:', errors);
          setError(((_errors$ = errors[0]) === null || _errors$ === void 0 ? void 0 : _errors$.message) || 'An error occurred');
          setSubmitting(false);
          return;
        }
        setError(null);
        dispatch({
          type: 'STORE_MODAL_NAVIGATE',
          route: 'cart'
        });
      },
      updater: store => {
        store.invalidateStore();
      }
    });
  }, [selectedSkuIds, licenseOptions, orderVariableSelections, licenseeIsBillingIdentity, dispatch, commitAddToCart, setError]);
  if (!viewer.precart) return null;
  const anySkusSelected = Object.keys(selectedSkuIds).filter(skuId => selectedSkuIds[skuId]).length > 0;
  const stylesPrice = countStylesPrice(viewer);
  let taxLabel = null;
  const {
    taxSystem,
    stripeTaxSettings
  } = viewer.settings ?? {};
  if (taxSystem === 'stripe') {
    // For Stripe Tax, use the explicit tax behavior setting
    if ((stripeTaxSettings === null || stripeTaxSettings === void 0 ? void 0 : stripeTaxSettings.taxBehavior) === 'exclusive') {
      taxLabel = ' excl. Tax';
    }
    // No label for 'inclusive' or 'inferred_by_currency'
  } else {
    // Legacy fontdue tax system - show label if any tax is not inclusive
    if (viewer.taxCollections.some(_ref3 => {
      let {
        inclusive
      } = _ref3;
      return !inclusive;
    })) {
      taxLabel = ' excl. Tax';
    }
  }
  let message = null;
  const requiredOrderVariablesNotYetSelected = (_viewer$orderVariable = viewer.orderVariables) === null || _viewer$orderVariable === void 0 ? void 0 : _viewer$orderVariable.filter(variable => variable.variableType === 'select').find(variable => !orderVariableSelections.find(selection => selection.orderVariableId === variable.id));
  const hasSelectionForLicenseeIsBillingIdentity = licenseeIsBillingIdentity !== null;
  const validSelections = anySkusSelected && licenseOptions.length > 0 && !requiredOrderVariablesNotYetSelected && hasSelectionForLicenseeIsBillingIdentity;
  const {
    storeModal: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const licensesPosition = config.productLicensesPosition;
  if (validSelections) {
    if (viewer.precart.subtotal) {
      message = /*#__PURE__*/_react.default.createElement("div", {
        className: "store-modal__product-summary__selection-text"
      }, (0, _utils.pluralize)(countStyles(viewer), 'font'), ' for ', viewer.precart.subtotal.amount < stylesPrice && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        className: "store-modal__product-summary__styles-price"
      }, /*#__PURE__*/_react.default.createElement(_Price.Price, {
        price: {
          amount: stylesPrice,
          currency: viewer.precart.subtotal.currency
        }
      })), ' '), /*#__PURE__*/_react.default.createElement(_Price.default, {
        price: viewer.precart.subtotal
      }), taxLabel);
    } else {
      message = /*#__PURE__*/_react.default.createElement("span", {
        className: "store-modal__product-summary__nothing-selected"
      }, "Calculating price...");
    }
  } else {
    let validationMessage;
    if (!anySkusSelected && licensesPosition === 'top') validationMessage = 'Please select some fonts';
    if (licenseOptions.length === 0) validationMessage = 'Please select a license';
    if (requiredOrderVariablesNotYetSelected) validationMessage = `Please select a ${requiredOrderVariablesNotYetSelected.name.toLocaleLowerCase()}`;
    if (!hasSelectionForLicenseeIsBillingIdentity) validationMessage = 'Please select the licensee';
    if (!anySkusSelected && licensesPosition === 'bottom') validationMessage = 'Please select some fonts';
    message = /*#__PURE__*/_react.default.createElement("span", {
      className: "store-modal__product-summary__nothing-selected"
    }, validationMessage);
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__product-summary__container"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__product-summary__row"
  }, message, /*#__PURE__*/_react.default.createElement(_Button.default, {
    className: "store-modal__product-summary__add-to-cart-button",
    onClick: handleAddToCart,
    disabled: submitting || !validSelections
  }, submitting ? 'Adding...' : 'Add to cart')), error && /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__product-summary__error"
  }, error, ' ', /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: clearCart,
    className: "store-modal__product-summary__error-button"
  }, "Empty cart and try again")));
};
var _default = StoreModalProductSummary;
exports.default = _default;
