// fontdue-js/dist/components/elements/StoreModalLicenseeIsBillingIdentityElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = StoreModalLicenseeIsBillingIdentityElement;
var _react = _interopRequireWildcard(require("react"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _Button = _interopRequireDefault(require("./Button"));
var _Icons = require("../Icons");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function CheckboxButton(_ref) {
  let {
    label,
    selected,
    onClick
  } = _ref;
  const {
    form: config
  } = (0, _react.useContext)(_ConfigContext.default);
  return /*#__PURE__*/_react.default.createElement(_Button.default, {
    className: "store-modal__order-variable-selection__button",
    onClick: onClick,
    "data-selected": selected
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__label"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "store-modal__order-variable-selection__icon-pair"
  }, selected === true ? config.checkboxStyle === 'check' ? /*#__PURE__*/_react.default.createElement(_Icons.CheckboxChecked, null) : /*#__PURE__*/_react.default.createElement(_Icons.CheckboxCrossed, null) : /*#__PURE__*/_react.default.createElement(_Icons.Checkbox, null), label)));
}
function StoreModalLicenseeIsBillingIdentityElement(_ref2) {
  let {
    disabled,
    onChange,
    value,
    children: {
      label,
      selfLabel,
      selfDescription,
      clientLabel,
      clientDescription
    }
  } = _ref2;
  const handleClick = (0, _react.useCallback)(licenseeIsBilling => () => {
    onChange(licenseeIsBilling);
  }, [onChange]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__container",
    "data-disabled": disabled
  }, /*#__PURE__*/_react.default.createElement("h3", {
    className: "store-modal__licensee-is-billing-identity__title"
  }, label), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__buttons"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__option"
  }, /*#__PURE__*/_react.default.createElement(CheckboxButton, {
    label: selfLabel ?? 'Yourself',
    selected: value === true,
    onClick: handleClick(true)
  }), selfDescription ? /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__description"
  }, selfDescription) : null), /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__option"
  }, /*#__PURE__*/_react.default.createElement(CheckboxButton, {
    label: clientLabel ?? 'Your client',
    selected: value === false,
    onClick: handleClick(false)
  }), clientDescription ? /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__licensee-is-billing-identity__description"
  }, clientDescription) : null)));
}
