// fontdue-js/dist/components/PriceBar/PriceBarSection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _PriceBarSection_node2 = _interopRequireDefault(require("../../__generated__/PriceBarSection_node.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _Price = _interopRequireWildcard(require("../Price"));
var _SelectButton = _interopRequireDefault(require("../SelectButton"));
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function PriceBarSection(_ref) {
  var _node$sku;
  let {
    node: nodeKey
  } = _ref;
  const node = (0, _reactRelay.useFragment)((_PriceBarSection_node2.default.hash && _PriceBarSection_node2.default.hash !== "f7d785b4599958bb699c2a706ab763cc" && console.error("The definition of 'PriceBarSection_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _PriceBarSection_node2.default), nodeKey);
  const selected = (0, _reactRedux.useSelector)((0, _utils.isSelected)(((_node$sku = node.sku) === null || _node$sku === void 0 ? void 0 : _node$sku.id) ?? null));
  const dispatch = (0, _reactRedux.useDispatch)();
  const onSelectSkuId = (skuId, selected) => {
    if (!skuId) return;
    dispatch({
      type: 'SELECT_SKU_ID',
      skuId,
      selected
    });
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "select-button-hover-parent select-button-click-parent price-bar-section",
    "data-type": node.collectionType,
    "data-selected": selected
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "price-bar-section__button",
    type: "button",
    onClick: () => {
      var _node$sku2;
      return onSelectSkuId((_node$sku2 = node.sku) === null || _node$sku2 === void 0 ? void 0 : _node$sku2.id, !selected);
    }
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "price-bar-section__name"
  }, /*#__PURE__*/_react.default.createElement("span", null, node.name, " ", node.collectionType), "\xA0", /*#__PURE__*/_react.default.createElement("span", {
    className: "price-bar-section__name__styles"
  }, "(", node.totalStyles, " styles)")), /*#__PURE__*/_react.default.createElement("div", {
    className: "price-bar-section__price"
  }, !selected && /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "price-bar-section__price__full"
  }, /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: node.totalStylesPrice,
    format: _Price.PriceFormat.Truncated
  })), "\xA0", /*#__PURE__*/_react.default.createElement("div", null, node.sku && node.sku.priceWithLicenseOptions && /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: node.sku.priceWithLicenseOptions,
    format: _Price.PriceFormat.Truncated
  })))), node.sku && /*#__PURE__*/_react.default.createElement("div", {
    className: "price-bar-section__select-button"
  }, /*#__PURE__*/_react.default.createElement(_SelectButton.default, {
    sku: node.sku,
    productId: node.sku.id
  })));
}
var _default = PriceBarSection;
exports.default = _default;
