// fontdue-js/dist/components/NewsletterSignup/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NewsletterSignupPreloadedQueryRenderer = NewsletterSignupPreloadedQueryRenderer;
exports.default = NewsletterSignup;
var _NewsletterSignupQuery2 = _interopRequireDefault(require("../../__generated__/NewsletterSignupQuery.graphql"));
var _NewsletterSignupUpdateCustomerMutation2 = _interopRequireDefault(require("../../__generated__/NewsletterSignupUpdateCustomerMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactGoogleRecaptcha = _interopRequireDefault(require("react-google-recaptcha"));
var _TextField = _interopRequireDefault(require("../TextField"));
var _Check = _interopRequireDefault(require("../Icons/Check"));
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const updateCustomerMutation = (_NewsletterSignupUpdateCustomerMutation2.default.hash && _NewsletterSignupUpdateCustomerMutation2.default.hash !== "769087891b6f263122bbb630b3f2ca6c" && console.error("The definition of 'NewsletterSignupUpdateCustomerMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _NewsletterSignupUpdateCustomerMutation2.default);
const query = (_NewsletterSignupQuery2.default.hash && _NewsletterSignupQuery2.default.hash !== "24b303198a6038318723fc0124548862" && console.error("The definition of 'NewsletterSignupQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _NewsletterSignupQuery2.default);
function NewsletterSignupComponent(_ref) {
  var _data$viewer, _data$viewer$settings, _data$viewer2, _data$viewer2$setting, _data$viewer3, _data$viewer3$setting, _data$viewer4, _data$viewer4$setting;
  let {
    optInLabel: optInLabelProp,
    successLabel: successLabelProp,
    buttonLabel = 'Subscribe',
    optInCheckboxChecked = false,
    title,
    intro,
    data
  } = _ref;
  const [name, setName] = (0, _react.useState)('');
  const [email, setEmail] = (0, _react.useState)('');
  const [newsletterOptIn, setNewsletterOptIn] = (0, _react.useState)(optInCheckboxChecked);
  const [error, setError] = (0, _react.useState)(null);
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const [submitted, setSubmitted] = (0, _react.useState)(false);
  const [recaptchaToken, setRecaptchaToken] = (0, _react.useState)(null);
  const [pendingSubmit, setPendingSubmit] = (0, _react.useState)(false);
  const recaptchaRef = (0, _react.useRef)(null);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const recaptchaEnabled = ((_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : (_data$viewer$settings = _data$viewer.settings) === null || _data$viewer$settings === void 0 ? void 0 : _data$viewer$settings.recaptchaEnabled) ?? false;
  const recaptchaSiteKey = (_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : (_data$viewer2$setting = _data$viewer2.settings) === null || _data$viewer2$setting === void 0 ? void 0 : _data$viewer2$setting.recaptchaSiteKey;
  const submitForm = (0, _react.useCallback)(token => {
    setSubmitting(true);
    setPendingSubmit(false);
    setError(null);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: updateCustomerMutation,
      variables: {
        input: {
          name,
          email,
          newsletterOptIn,
          recaptchaToken: token ?? undefined
        }
      },
      onCompleted: (_res, errors) => {
        if (errors && errors.length > 0) {
          var _recaptchaRef$current;
          setError(errors[0].message);
          setSubmitting(false);
          // Reset reCAPTCHA on error so user can retry
          (_recaptchaRef$current = recaptchaRef.current) === null || _recaptchaRef$current === void 0 ? void 0 : _recaptchaRef$current.reset();
          setRecaptchaToken(null);
        } else {
          setSubmitted(true);
          setSubmitting(false);
          setError(null);
        }
      },
      onError: _ref2 => {
        var _recaptchaRef$current2;
        let {
          message
        } = _ref2;
        setError(message);
        setSubmitting(false);
        // Reset reCAPTCHA on error so user can retry
        (_recaptchaRef$current2 = recaptchaRef.current) === null || _recaptchaRef$current2 === void 0 ? void 0 : _recaptchaRef$current2.reset();
        setRecaptchaToken(null);
      }
    });
  }, [environment, name, email, newsletterOptIn]);
  const handleRecaptchaChange = (0, _react.useCallback)(token => {
    setRecaptchaToken(token);
    // If we were waiting for a token to submit, do it now
    if (token && pendingSubmit) {
      submitForm(token);
    }
  }, [pendingSubmit, submitForm]);
  const handleRecaptchaExpired = (0, _react.useCallback)(() => {
    setRecaptchaToken(null);
  }, []);
  const handleSubmit = e => {
    e.preventDefault();

    // If reCAPTCHA is enabled but no token, execute it and wait for callback
    if (recaptchaEnabled && recaptchaSiteKey && !recaptchaToken) {
      var _recaptchaRef$current3;
      setPendingSubmit(true);
      setError(null);
      (_recaptchaRef$current3 = recaptchaRef.current) === null || _recaptchaRef$current3 === void 0 ? void 0 : _recaptchaRef$current3.execute();
      return;
    }
    submitForm(recaptchaToken);
  };

  // Button is disabled only while actively submitting or waiting for reCAPTCHA
  const disabled = !newsletterOptIn || !email || !name;
  const optInLabel = optInLabelProp || ((_data$viewer3 = data.viewer) === null || _data$viewer3 === void 0 ? void 0 : (_data$viewer3$setting = _data$viewer3.settings) === null || _data$viewer3$setting === void 0 ? void 0 : _data$viewer3$setting.newsletterOptInLabel);
  const successLabel = (successLabelProp || ((_data$viewer4 = data.viewer) === null || _data$viewer4 === void 0 ? void 0 : (_data$viewer4$setting = _data$viewer4.settings) === null || _data$viewer4$setting === void 0 ? void 0 : _data$viewer4$setting.newsletterSuccessLabel)) ?? 'Success!';
  if (submitted) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "newsletter-signup"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "newsletter-signup__success",
      dangerouslySetInnerHTML: {
        __html: successLabel
      }
    }));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "newsletter-signup"
  }, /*#__PURE__*/_react.default.createElement("h2", {
    className: "newsletter-signup__title"
  }, title), /*#__PURE__*/_react.default.createElement("p", {
    className: "newsletter-signup__intro"
  }, intro), error && /*#__PURE__*/_react.default.createElement("p", {
    className: "newsletter-signup__error"
  }, error), /*#__PURE__*/_react.default.createElement("form", {
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "newsletter-signup__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Name *",
    placeholder: "Your name",
    autocomplete: "name",
    value: name,
    onChange: value => setName(value)
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "newsletter-signup__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Email *",
    placeholder: "Your email",
    type: "email",
    autocomplete: "email",
    value: email,
    onChange: value => setEmail(value)
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "newsletter-signup__section newsletter-signup__checkbox-field"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox"
  }, /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    id: "customer-newsletter-opt-in",
    checked: newsletterOptIn,
    onChange: e => setNewsletterOptIn(e.target.checked)
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox__icon"
  }, /*#__PURE__*/_react.default.createElement(_Check.default, null))), /*#__PURE__*/_react.default.createElement("label", {
    htmlFor: "customer-newsletter-opt-in",
    dangerouslySetInnerHTML: {
      __html: optInLabel ?? ''
    }
  })), recaptchaEnabled && recaptchaSiteKey && /*#__PURE__*/_react.default.createElement(_reactGoogleRecaptcha.default, {
    ref: recaptchaRef,
    sitekey: recaptchaSiteKey,
    size: "invisible",
    onChange: handleRecaptchaChange,
    onExpired: handleRecaptchaExpired
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "newsletter-signup__section"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "submit-button",
    disabled: submitting || pendingSubmit || disabled
  }, submitting || pendingSubmit ? 'Submitting...' : buttonLabel, ' ', /*#__PURE__*/_react.default.createElement("span", {
    className: "submit-button__arrow"
  }, ' →')))));
}
function NewsletterSignupPreloadedQueryRenderer(_ref3) {
  let {
    preloadedQuery,
    ...props
  } = _ref3;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(query, queryRef);
  return /*#__PURE__*/_react.default.createElement(NewsletterSignupComponent, _extends({
    data: data
  }, props));
}
function NewsletterSignup(props) {
  const data = (0, _reactRelay.useLazyLoadQuery)(query, {});
  return /*#__PURE__*/_react.default.createElement(NewsletterSignupComponent, _extends({}, props, {
    data: data
  }));
}
