// fontdue-js/dist/components/CartButton/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CartButtonPreloadedQueryRenderer = CartButtonPreloadedQueryRenderer;
exports.default = CartButton;
var _CartButtonQuery2 = _interopRequireDefault(require("../../__generated__/CartButtonQuery.graphql"));
var _CartButton_order2 = _interopRequireDefault(require("../../__generated__/CartButton_order.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _reactRelay = require("react-relay");
var _ComponentsContext = _interopRequireDefault(require("../ComponentsContext"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _Icons = require("../Icons");
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
var _Price = _interopRequireWildcard(require("../Price"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function getOrderCount(order) {
  return order.orderItems ? order.orderItems.length : 0;
}
const cartButtonCount = order => {
  if (!order) return 0;
  return getOrderCount(order);
};
const openCart = () => ({
  type: 'OPEN_CART'
});
function CartButtonComponent(_ref) {
  let {
    label,
    buttonStyle,
    order: orderKey,
    children,
    suffix
  } = _ref;
  const orderData = (0, _reactRelay.useFragment)((_CartButton_order2.default.hash && _CartButton_order2.default.hash !== "26a092d7efb579c98adda18413349b3b" && console.error("The definition of 'CartButton_order' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartButton_order2.default), orderKey);
  const [count, setCount] = (0, _react.useState)(0);
  const dispatch = (0, _reactRedux.useDispatch)();
  const {
    storeModal: storeModalConfig
  } = (0, _react.useContext)(_ConfigContext.default);
  const formatSuffix = (0, _react.useCallback)(() => {
    if (!suffix) return null;

    // Check if the suffix contains any substitutions
    const hasSubstitutions = /\{(count|subtotal)\}/.test(suffix);

    // Hide if there are substitutions and cart is empty
    if (hasSubstitutions && count === 0) return null;

    // Split into parts: ['text', '{placeholder}', 'text', '{placeholder}', 'text']
    const parts = suffix.split(/(\{[^}]+\})/);

    // Transform each part based on whether it's a placeholder or text
    const nodes = parts.map((part, index) => {
      const match = part.match(/^\{([^}]+)\}$/);
      if (!match) return part || null; // Regular text

      // Handle placeholder
      const placeholder = match[1];
      switch (placeholder) {
        case 'count':
          return count > 0 ? count.toString() : null;
        case 'subtotal':
          return orderData !== null && orderData !== void 0 && orderData.subtotal ? /*#__PURE__*/_react.default.createElement(_Price.default, {
            key: `price-${index}`,
            price: orderData.subtotal,
            format: _Price.PriceFormat.Truncated
          }) : null;
        default:
          return part;
        // Keep unknown placeholders as-is
      }
    }).filter(Boolean); // Remove null/empty parts

    return nodes.length > 0 ? nodes : null;
  }, [suffix, count, orderData === null || orderData === void 0 ? void 0 : orderData.subtotal]);
  const handleClick = (0, _react.useCallback)(() => {
    if (count > 0 || storeModalConfig.hideAllCollections) {
      dispatch({
        type: 'STORE_MODAL_REPLACE',
        route: 'cart'
      });
    }
    dispatch(openCart());
  }, [count, dispatch, storeModalConfig.hideAllCollections]);
  (0, _react.useEffect)(() => {
    // set count only on the client, because on the server it's always gonna be zero
    setCount(cartButtonCount(orderData));
  }, [orderData, setCount]);
  if (!orderData) return null;
  if (orderData.stripeCharge && orderData.stripeCharge.paid) return null;
  const suffixNodes = formatSuffix();
  return /*#__PURE__*/_react.default.createElement(_ComponentsContext.default.Consumer, null, components => components.CartButton ? /*#__PURE__*/_react.default.createElement(components.CartButton, {
    count: count,
    innerProps: {
      onClick: handleClick
    }
  }) : /*#__PURE__*/_react.default.createElement("button", {
    className: "cart-button",
    onClick: handleClick,
    "data-count": count,
    "data-active": count >= 1,
    "data-button-style": buttonStyle
  }, children ? children : buttonStyle === 'icon' ? /*#__PURE__*/_react.default.createElement(_Icons.CartNew, null) : /*#__PURE__*/_react.default.createElement("span", {
    className: "cart-button__label"
  }, label || 'Cart', suffixNodes && /*#__PURE__*/_react.default.createElement("span", {
    className: "cart-button__suffix"
  }, suffixNodes))));
}
const query = (_CartButtonQuery2.default.hash && _CartButtonQuery2.default.hash !== "5c0da5849bf1215a9d1f264831830f71" && console.error("The definition of 'CartButtonQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartButtonQuery2.default);
function CartButtonPreloadedQueryRenderer(_ref2) {
  var _data$viewer;
  let {
    preloadedQuery,
    ...rest
  } = _ref2;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(query, queryRef);
  return /*#__PURE__*/_react.default.createElement(CartButtonComponent, _extends({}, rest, {
    order: ((_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : _data$viewer.currentOrder) ?? null
  }));
}
function CartButton(props) {
  var _data$viewer2;
  const data = (0, _reactRelay.useLazyLoadQuery)(query, {});
  return /*#__PURE__*/_react.default.createElement(CartButtonComponent, _extends({}, props, {
    order: ((_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : _data$viewer2.currentOrder) ?? null
  }));
}
