// fontdue-js/dist/components/Cart/AddressFields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _AddressFields_viewer2 = _interopRequireDefault(require("../../__generated__/AddressFields_viewer.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _TextField = _interopRequireDefault(require("../TextField"));
var _AddressFieldsRefetchQuery = _interopRequireDefault(require("../../__generated__/AddressFieldsRefetchQuery.graphql"));
var _utils = require("../../utils");
var _reactStripeJs = require("@stripe/react-stripe-js");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const MAPS_API_KEY = 'AIzaSyB9rSjGVPZYcRkPl1PDezIuAp3XCUs2Crk';
const AddressFields = _ref => {
  var _viewer$countryData, _value$street, _viewer$settings;
  let {
    viewer: viewerKey,
    value,
    onChange,
    identityType,
    errorsObject,
    allowedCountries
  } = _ref;
  const [refetching, setRefetching] = (0, _react.useState)(false);
  const [viewer, refetch] = (0, _reactRelay.useRefetchableFragment)((_AddressFields_viewer2.default.hash && _AddressFields_viewer2.default.hash !== "2369fc95f554723701a8294b4522137a" && console.error("The definition of 'AddressFields_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _AddressFields_viewer2.default), viewerKey);
  const countryData = (_viewer$countryData = viewer.countryData) === null || _viewer$countryData === void 0 ? void 0 : _viewer$countryData.response;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  (0, _react.useEffect)(() => {
    setRefetching(true);

    // refetch after using fetchQuery to avoid triggering Suspense
    const subscription = (0, _reactRelay.fetchQuery)(environment, _AddressFieldsRefetchQuery.default, {
      countryCode: value.country,
      includeCountryCode: true
    }).subscribe({
      complete: () => {
        setRefetching(false);
        refetch({
          countryCode: value.country,
          includeCountryCode: true
        }, {
          fetchPolicy: 'store-only'
        });
      },
      error: () => {
        setRefetching(false);
      }
    });
    return () => subscription.unsubscribe();
  }, [value.country]);
  const getErrors = field => {
    var _errorsObject$field;
    return errorsObject === null || errorsObject === void 0 ? void 0 : (_errorsObject$field = errorsObject[field]) === null || _errorsObject$field === void 0 ? void 0 : _errorsObject$field.filter(_utils.notEmpty);
  };
  const lines = (_value$street = value.street) === null || _value$street === void 0 ? void 0 : _value$street.split('\n');

  // Use consolidated phone settings
  const phoneCollection = (_viewer$settings = viewer.settings) === null || _viewer$settings === void 0 ? void 0 : _viewer$settings.checkoutPhoneNumberCollection;
  const phoneFieldConfig = phoneCollection === 'never' ? {
    phone: 'never'
  } : {
    phone: 'always'
  };
  const phoneValidationConfig = phoneCollection === 'required' ? {
    phone: {
      required: 'always'
    }
  } : phoneCollection === 'never' ? {} : {
    phone: {
      required: 'never'
    }
  };

  // Only include phone in defaultValues if phone field is shown
  const defaultValues = {
    name: value.name,
    ...(phoneCollection !== 'never' && {
      phone: value.phone
    }),
    address: {
      country: value.country,
      line1: lines === null || lines === void 0 ? void 0 : lines[0],
      line2: lines === null || lines === void 0 ? void 0 : lines[1],
      postal_code: value.zip,
      city: value.locality,
      state: value.administrativeArea
    }
  };
  const handleAddressElementChange = (0, _react.useCallback)(e => {
    // Always update country immediately, regardless of form completeness
    const country = e.value.address.country;
    const value = {
      name: e.value.name,
      phone: e.value.phone ?? '',
      country,
      street: e.value.address.line1 + (e.value.address.line2 ? `\n${e.value.address.line2}` : ''),
      locality: e.value.address.city,
      sublocality: '',
      administrativeArea: e.value.address.state,
      zip: e.value.address.postal_code
    };
    onChange(value, e.complete);
  }, [onChange]);
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Organization",
    value: value.organization || '',
    onChange: v => onChange({
      organization: v
    })
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: identityType === 'licensee' ? 'Contact email' : 'Email',
    autocomplete: "email",
    type: "email",
    value: value.email || '',
    onChange: v => onChange({
      email: v
    }),
    errors: getErrors('email')
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_reactStripeJs.AddressElement, {
    key: (allowedCountries === null || allowedCountries === void 0 ? void 0 : allowedCountries.join(',')) || 'all',
    options: {
      mode: identityType === 'billing' ? 'billing' : 'shipping',
      autocomplete: {
        mode: 'google_maps_api',
        apiKey: MAPS_API_KEY
      },
      fields: phoneFieldConfig,
      validation: phoneValidationConfig,
      display: {
        name: 'full'
      },
      ...(allowedCountries && allowedCountries.length > 0 ? {
        allowedCountries
      } : {}),
      defaultValues
    },
    onChange: handleAddressElementChange
  })), identityType === 'billing' ? countryData && !refetching ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, (countryData.isEuCountry || value.country === 'GB') && /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "VAT number",
    autocomplete: "vat",
    placeholder: "12345678",
    value: value.vatNumber || '',
    onChange: v => onChange({
      vatNumber: v
    }),
    errors: getErrors('vatNumber'),
    prefix: `${countryData.vatCountryCode} `
  }))) : /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__field-placeholder"
  })) : null);
};
var _default = AddressFields;
exports.default = _default;
