// fontdue-js/dist/components/Cart/CartOrder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _CartOrder_viewer2 = _interopRequireDefault(require("../../__generated__/CartOrder_viewer.graphql"));
var _CartOrder_order2 = _interopRequireDefault(require("../../__generated__/CartOrder_order.graphql"));
var _CartOrderRemoveDiscountMutation2 = _interopRequireDefault(require("../../__generated__/CartOrderRemoveDiscountMutation.graphql"));
var _CartOrderUpdateMutation2 = _interopRequireDefault(require("../../__generated__/CartOrderUpdateMutation.graphql"));
var _CartOrder_UpdateErrors2 = _interopRequireDefault(require("../../__generated__/CartOrder_UpdateErrors.graphql"));
var _CartOrderCompleteOrderMutation2 = _interopRequireDefault(require("../../__generated__/CartOrderCompleteOrderMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Checkout = _interopRequireDefault(require("./Checkout"));
var _CheckoutSteps = _interopRequireDefault(require("./CheckoutSteps"));
var _CartItem = _interopRequireDefault(require("./CartItem"));
var _Icons = require("../Icons");
var _Download = _interopRequireDefault(require("./Download"));
var _CouponCodeInput = _interopRequireDefault(require("./CouponCodeInput"));
var _EmptyCart = _interopRequireDefault(require("./EmptyCart"));
var _IdentityBox = _interopRequireDefault(require("./IdentityBox"));
var _CartTotals = _interopRequireDefault(require("./CartTotals"));
var _utils = require("./utils");
var _reactStripeJs = require("@stripe/react-stripe-js");
var _reactRedux = require("react-redux");
var _CartState = _interopRequireDefault(require("./CartState"));
var _orderTracking = require("./orderTracking");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// @ts-ignore

// @ts-ignore

// note here we're only updating the order (by id).
// careful not to update the current customer's reference
// to its order, since at this time it will turn to `null`
// (this is because the 'order' field on customer represents
//   their current 'cart', and when we pay for an order, we
//   want to reset the cart back to empty).
// but here we want to keep displaying the order that they
// just paid for
const submitCompleteMutation = (_CartOrderCompleteOrderMutation2.default.hash && _CartOrderCompleteOrderMutation2.default.hash !== "86f8fe7e3a867ae234cb381df186c107" && console.error("The definition of 'CartOrderCompleteOrderMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrderCompleteOrderMutation2.default);
_CartOrder_UpdateErrors2.default.hash && _CartOrder_UpdateErrors2.default.hash !== "0bc5cd4fa568238d362a545367f9c306" && console.error("The definition of 'CartOrder_UpdateErrors' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrder_UpdateErrors2.default;
const orderChangeMutation = (_CartOrderUpdateMutation2.default.hash && _CartOrderUpdateMutation2.default.hash !== "9f208ec36dbea5d90eeae4574b48bd15" && console.error("The definition of 'CartOrderUpdateMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrderUpdateMutation2.default);
const removeDiscountMutation = (_CartOrderRemoveDiscountMutation2.default.hash && _CartOrderRemoveDiscountMutation2.default.hash !== "8a84a9135f1406fa9c4cc16dc2e11d30" && console.error("The definition of 'CartOrderRemoveDiscountMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrderRemoveDiscountMutation2.default);
const CartOrder = _ref => {
  var _order$stripePaymentM, _viewer$settings;
  let {
    order: orderKey,
    viewer: viewerKey,
    open
  } = _ref;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const stripe = (0, _reactStripeJs.useStripe)();
  const dispatch = (0, _reactRedux.useDispatch)();

  // Capture analytics consent + attribution on the order for the purchase
  // event the server emits on completion (from a Stripe webhook, where no
  // browser context exists).
  (0, _react.useEffect)(() => {
    if (open) (0, _orderTracking.sendOrderTracking)(environment);
  }, [open, environment]);
  const order = (0, _reactRelay.useFragment)((_CartOrder_order2.default.hash && _CartOrder_order2.default.hash !== "25ef000c40000c4f76f973ac6ac7f593" && console.error("The definition of 'CartOrder_order' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrder_order2.default), orderKey);
  const viewer = (0, _reactRelay.useFragment)((_CartOrder_viewer2.default.hash && _CartOrder_viewer2.default.hash !== "8cf5ab5a33a474483a6c231cc44b0df1" && console.error("The definition of 'CartOrder_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartOrder_viewer2.default), viewerKey);

  // errors is from the payload, typically user errors,
  // whereas 'error' refers to schema/network/catastrophic errors
  const [error, setError] = (0, _react.useState)(null);
  const [errors, setErrors] = (0, _react.useState)(null);
  const [checkingOut, setCheckingOut] = (0, _react.useState)(false);
  const [completing, setCompleting] = (0, _react.useState)(false);
  const [paymentError, setPaymentError] = (0, _react.useState)(null);
  const [checkoutStep, setCheckoutStep] = (0, _react.useState)(0);
  const [eulaAgreed, setEulaAgreed] = (0, _react.useState)(false);
  const modal = (0, _react.useRef)(null);
  const submitOrderChange = orderItems => {
    if (!orderItems) return;
    (0, _reactRelay.commitMutation)(environment, {
      mutation: orderChangeMutation,
      variables: {
        input: {
          orderItems
        }
      },
      onCompleted: (result, errors) => {
        var _result$updateOrder;
        if (errors !== null && errors !== void 0 && errors[0]) {
          setError(errors[0].message);
        } else if ((_result$updateOrder = result.updateOrder) !== null && _result$updateOrder !== void 0 && _result$updateOrder.errors) {
          setErrors(result.updateOrder.errors);
        }
      }
    });
  };
  const openCheckout = () => setCheckingOut(true);
  const closeCheckout = () => setCheckingOut(false);
  const openCheckoutStep = checkoutStep => {
    setCheckingOut(true);
    setCheckoutStep(checkoutStep);
  };
  const complete = () => {
    setCompleting(true);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: submitCompleteMutation,
      variables: {},
      onCompleted: (response, errors) => {
        const requiresAction = response.completeOrder && response.completeOrder.requiresAction;
        if (requiresAction === 'use_stripe_sdk') {
          var _response$completeOrd, _response$completeOrd2, _response$completeOrd3;
          const secret = (_response$completeOrd = response.completeOrder) === null || _response$completeOrd === void 0 ? void 0 : (_response$completeOrd2 = _response$completeOrd.order) === null || _response$completeOrd2 === void 0 ? void 0 : (_response$completeOrd3 = _response$completeOrd2.stripePaymentIntent) === null || _response$completeOrd3 === void 0 ? void 0 : _response$completeOrd3.clientSecret;
          if (!secret) throw new Error('Stripe payment secret not found');
          stripe === null || stripe === void 0 ? void 0 : stripe.handleCardAction(secret).then(result => {
            var _result$error;
            if ((_result$error = result.error) !== null && _result$error !== void 0 && _result$error.message) {
              setPaymentError(result.error.message);
              setCompleting(false);
              setCheckingOut(true);
              setCheckoutStep(_CheckoutSteps.default.Payment);
            } else {
              // re-confirm
              complete();
            }
          });
          return;
        }
        if (requiresAction === 'enter_new_payment_method') {
          setPaymentError('We were unable to complete your payment, please re-enter your card details and try again');
          setCompleting(false);
          setCheckingOut(true);
          setCheckoutStep(_CheckoutSteps.default.Payment);
          return;
        }
        if (errors && errors.length > 0) {
          setPaymentError(errors[0].message);
          setCompleting(false);
          setCheckingOut(true);
        }
      }
    });
  };
  const removeDiscount = () => {
    (0, _reactRelay.commitMutation)(environment, {
      mutation: removeDiscountMutation,
      variables: {},
      onCompleted: (_result, errors) => {
        if (errors !== null && errors !== void 0 && errors[0]) {
          setError(errors[0].message);
        }
      }
    });
  };
  const requiresCheckout = () => !order.canCompleteWithCoupon;
  const clearError = () => {
    submitOrderChange([]);
    setError(null);
  };
  const validateCanSubmit = () => {
    return eulaAgreed && (0, _utils.textVariablesAllHaveText)(order);
  };
  const handleScrollToTop = () => {
    if (modal.current) {
      modal.current.scrollTop = 0;
    }
  };
  if (!order) return null;
  const {
    orderItems
  } = order;
  if (!orderItems || orderItems.length < 1) {
    return /*#__PURE__*/_react.default.createElement(_EmptyCart.default, {
      open: open
    });
  }
  const reviewing = Boolean(order.customer && !order.customer.anonymous && order.billingIdentity && (order.licenseeIdentity || order.licenseeIsBillingIdentity) && (order.stripePaymentMethod || order.canCompleteWithCoupon));
  if (order.completedWithCoupon || order.stripePaymentIntent && order.stripePaymentIntent.status === 'succeeded') {
    return /*#__PURE__*/_react.default.createElement(_Download.default, {
      order: order
    });
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cart",
    "data-open": open,
    "data-eula-agreed": eulaAgreed
  }, open && /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__scroller"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__top"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__nav"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "cart__nav__back",
    type: "button",
    onClick: () => dispatch({
      type: 'CLOSE_CART'
    })
  }, /*#__PURE__*/_react.default.createElement(_Icons.X, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__nav__title"
  }, "Your cart")), (order.customer || order.billingIdentity || order.licenseeIdentity || order.stripePaymentMethod) && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review"
  }, order.billingIdentity && /*#__PURE__*/_react.default.createElement(_IdentityBox.default, {
    identity: order.billingIdentity,
    title: "Billing information",
    onClick: () => openCheckoutStep(_CheckoutSteps.default.Billing)
  }, {
    icon: /*#__PURE__*/_react.default.createElement(_Icons.Receipt, null)
  }), order.licenseeIdentity && /*#__PURE__*/_react.default.createElement(_IdentityBox.default, {
    identity: order.licenseeIdentity,
    title: "Licensee information",
    onClick: () => openCheckoutStep(_CheckoutSteps.default.Licensee)
  }, {
    icon: /*#__PURE__*/_react.default.createElement(_Icons.License, null)
  }), order.customer && !order.customer.anonymous && /*#__PURE__*/_react.default.createElement("button", {
    className: "cart__review__button cart__review__user-info",
    type: "button",
    onClick: () => openCheckoutStep(_CheckoutSteps.default.Customer)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__label"
  }, /*#__PURE__*/_react.default.createElement(_Icons.User, null), " Contact information", ' ', /*#__PURE__*/_react.default.createElement("span", {
    className: "cart__review__label__text"
  }, "Edit")), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, order.customer.name), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, order.customer.email)), order.stripePaymentMethod && /*#__PURE__*/_react.default.createElement("button", {
    className: "cart__review__button cart__review__credit-info",
    type: "button",
    onClick: () => openCheckoutStep(_CheckoutSteps.default.Payment)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__label"
  }, /*#__PURE__*/_react.default.createElement(_Icons.CreditCard, null), " Payment", ' ', /*#__PURE__*/_react.default.createElement("span", {
    className: "cart__review__label__text"
  }, "Edit")), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, "\u2013\u2013\u2013\u2013\u2013\u2013 \u2013\u2013\u2013\u2013\u2013\u2013 ", (_order$stripePaymentM = order.stripePaymentMethod.card) === null || _order$stripePaymentM === void 0 ? void 0 : _order$stripePaymentM.last4)))), (error || errors) && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__error"
  }, "There was a problem with your order.", ' ', /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    onClick: clearError,
    className: "cart__error__button"
  }, "Click here"), ' ', "to empty your cart and continue shopping."), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__middle"
  }, /*#__PURE__*/_react.default.createElement(_CartState.default, {
    order: order
  }, _ref2 => {
    let {
      onUpdateOrderItem
    } = _ref2;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, orderItems.map((item, i) => /*#__PURE__*/_react.default.createElement(_CartItem.default, {
      key: i,
      node: item,
      onChange: onUpdateOrderItem
    })));
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__bottom"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__bottom__wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-blocks-and-agreement"
  }, /*#__PURE__*/_react.default.createElement(_CartTotals.default, {
    order: order,
    onRemoveDiscount: removeDiscount
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-agreement",
    "data-highlighted": !eulaAgreed && reviewing
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox"
  }, /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    id: "eula-agreement",
    checked: eulaAgreed,
    onChange: e => setEulaAgreed(e.target.checked)
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox__icon"
  }, /*#__PURE__*/_react.default.createElement(_Icons.Check, null))), /*#__PURE__*/_react.default.createElement("label", {
    htmlFor: "eula-agreement",
    dangerouslySetInnerHTML: {
      __html: ((_viewer$settings = viewer.settings) === null || _viewer$settings === void 0 ? void 0 : _viewer$settings.eulaAgreementLabel) ?? ''
    }
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__additions"
  }, !order.discount && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__coupon-input"
  }, /*#__PURE__*/_react.default.createElement(_CouponCodeInput.default, null))), reviewing ? /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__checkout-button"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "submit-button",
    onClick: complete,
    disabled: completing || !validateCanSubmit(),
    "data-label": "submit",
    type: "button",
    "data-highlighted": true
  }, completing ? 'Submitting...' : /*#__PURE__*/_react.default.createElement(_react.Fragment, null, order.canCompleteWithCoupon ? 'Complete order using coupon' : 'Submit payment', /*#__PURE__*/_react.default.createElement("span", {
    className: "submit-button__arrow"
  }, ' →')))) : /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__checkout-button"
  }, /*#__PURE__*/_react.default.createElement("button", {
    className: "submit-button",
    onClick: openCheckout,
    type: "button"
  }, "Proceed", requiresCheckout() && ' to checkout', /*#__PURE__*/_react.default.createElement("span", {
    className: "submit-button__arrow"
  }, ' →'))))), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: checkingOut ? 'initial' : 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__click-out",
    onClick: closeCheckout
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__modal",
    ref: modal
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__form"
  }, /*#__PURE__*/_react.default.createElement(_Checkout.default, {
    viewer: viewer,
    order: order,
    error: paymentError,
    onCancel: closeCheckout,
    onComplete: closeCheckout,
    setCheckoutStep: setCheckoutStep,
    checkoutStep: checkoutStep,
    onScrollToTop: handleScrollToTop
  }))))))));
};
var _default = CartOrder;
exports.default = _default;
