// fontdue-js/dist/components/Cart/CartTotals.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartTotals;
var _CartTotals_order2 = _interopRequireDefault(require("../../__generated__/CartTotals_order.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
var _relayRuntime = require("relay-runtime");
var _Price = _interopRequireDefault(require("../Price"));
var _CouponText = _interopRequireDefault(require("../CouponText"));
var _Icons = require("../Icons");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function percentage(decimalValue) {
  if (!decimalValue) return null;
  return `${parseFloat(decimalValue) * 100}%`;
}
function PriceBlock(_ref) {
  let {
    label,
    belowPrice,
    price,
    invert,
    ...rest
  } = _ref;
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    className: "cart__price-block",
    "data-label": label
  }, rest), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__label"
  }, label), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__price"
  }, /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: price,
    invert: invert
  }), belowPrice));
}
function CartTotals(_ref2) {
  let {
    order: orderKey,
    onRemoveDiscount,
    allowDiscountRemoval = true,
    billingIdentity,
    taxSystem
  } = _ref2;
  const order = (0, _reactRelay.useFragment)((_CartTotals_order2.default.hash && _CartTotals_order2.default.hash !== "aee2e882d4d0d063fd4eaf6cb43485db" && console.error("The definition of 'CartTotals_order' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartTotals_order2.default), orderKey);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-blocks",
    "data-has-discount": Boolean(order.discount)
  }, /*#__PURE__*/_react.default.createElement(PriceBlock, {
    label: "Subtotal",
    price: order.subtotal,
    "data-line": "subtotal"
  }), order.discount && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(PriceBlock, {
    label: "Discount applied",
    price: order.discount.amount,
    invert: true,
    belowPrice: allowDiscountRemoval && /*#__PURE__*/_react.default.createElement("button", {
      title: "Remove discount",
      type: "button",
      onClick: onRemoveDiscount,
      className: "cart__remove-discount-button"
    }, /*#__PURE__*/_react.default.createElement(_Icons.X, null)),
    "data-line": "discount"
  }), order.discount.coupon && /*#__PURE__*/_react.default.createElement("p", {
    className: "cart__coupon-text"
  }, '(', /*#__PURE__*/_react.default.createElement(_CouponText.default, {
    coupon: order.discount.coupon,
    subtract: order.discount.amount ? order.discount.amount.amount : 0
  }), ')'), /*#__PURE__*/_react.default.createElement(PriceBlock, {
    label: "Discounted subtotal",
    price: order.discountedSubtotal,
    "data-line": "discounted-subtotal"
  })), taxSystem !== 'none' && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block",
    "data-line": "taxes"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__label"
  }, order.taxNotApplicable ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, order.taxName ?? 'VAT', " not applicable", order.taxNotApplicableReason && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("span", {
    className: "cart__price-block__placeholder"
  }, order.taxNotApplicableReason))) : order.taxes && order.taxName ? `${order.taxName} (${percentage(order.taxRate)})` : 'Tax'), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__price"
  }, order.taxes ? /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: order.taxes
  }) : taxSystem === 'stripe' || taxSystem === 'fontdue' && !billingIdentity ? /*#__PURE__*/_react.default.createElement("span", {
    className: "cart__price-block__placeholder"
  }, "Calculated at checkout") : '-')), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block",
    "data-line": "total"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__label"
  }, "Total", order.gstIncluded ? ' including GST' : ''), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__price-block__price"
  }, /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: order.total
  }))));
}
