// fontdue-js/dist/components/Cart/CartState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartState;
var _CartState_order2 = _interopRequireDefault(require("../../__generated__/CartState_order.graphql"));
var _CartStateRemoveDiscountMutation2 = _interopRequireDefault(require("../../__generated__/CartStateRemoveDiscountMutation.graphql"));
var _CartStateUpdateMutation2 = _interopRequireDefault(require("../../__generated__/CartStateUpdateMutation.graphql"));
var _react = require("react");
var _reactRelay = require("react-relay");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const orderChangeMutation = (_CartStateUpdateMutation2.default.hash && _CartStateUpdateMutation2.default.hash !== "07a639cd3bfbfa0d797f6b8380603859" && console.error("The definition of 'CartStateUpdateMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartStateUpdateMutation2.default);
const removeDiscountMutation = (_CartStateRemoveDiscountMutation2.default.hash && _CartStateRemoveDiscountMutation2.default.hash !== "bdcfcd6638660e83a61555dcd1786886" && console.error("The definition of 'CartStateRemoveDiscountMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartStateRemoveDiscountMutation2.default);
function CartState(_ref) {
  let {
    order: orderKey,
    children
  } = _ref;
  const order = (0, _reactRelay.useFragment)((_CartState_order2.default.hash && _CartState_order2.default.hash !== "43fd3d7af7ffe0b50a192674fde8286c" && console.error("The definition of 'CartState_order' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartState_order2.default), orderKey);
  const [error, setError] = (0, _react.useState)(null);
  const [errors, setErrors] = (0, _react.useState)(null);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const handleUpdateOrder = (0, _react.useCallback)(input => {
    (0, _reactRelay.commitMutation)(environment, {
      mutation: orderChangeMutation,
      variables: {
        input
      },
      onCompleted: (result, error) => {
        if (error) {
          setError(error);
        } else if (result.updateOrder && result.updateOrder.errors) {
          setErrors(result.updateOrder.errors);
        } else {
          setError(null);
          setErrors(null);
        }
      }
    });
  }, [environment, order]);
  const handleRemoveDiscount = (0, _react.useCallback)(() => {
    (0, _reactRelay.commitMutation)(environment, {
      mutation: removeDiscountMutation,
      variables: {},
      onCompleted: (_result, error) => {
        if (error) {
          setError(error);
        }
      }
    });
  }, [environment]);
  const handleUpdateOrderItem = (0, _react.useCallback)(item => {
    // Handle CartItemChangeInput to maintain backward compatibility
    const newItem = item;
    const orderItems = (order.orderItems || []).reduce((res, item) => {
      if (item.id === newItem.id) return res.concat([newItem]);
      if (!item.licenseSelections) return res;
      return res.concat([{
        id: item.id,
        licenseSelections: item.licenseSelections.map(_ref2 => {
          let {
            license,
            licenseVariable,
            licenseOption,
            variableText
          } = _ref2;
          if (!license) return null;
          return {
            licenseId: license.id,
            licenseVariableId: licenseVariable && licenseVariable.id,
            licenseOptionId: licenseOption && licenseOption.id,
            variableText
          };
        })
      }]).filter(Boolean);
    }, [])
    /* remove order items if there are no licenses */.filter(item => item.licenseSelections.length > 0);
    handleUpdateOrder({
      orderItems
    });
  }, [order.orderItems]);
  const handleUpdateOrderVariableSelections = (0, _react.useCallback)(variables => {
    handleUpdateOrder({
      orderVariableSelections: variables
    });
  }, [handleUpdateOrder]);
  return children({
    onUpdateOrderVariableSelections: handleUpdateOrderVariableSelections,
    onRemoveDiscount: handleRemoveDiscount,
    onUpdateOrderItem: handleUpdateOrderItem,
    errors,
    error
  });
}
