// fontdue-js/dist/components/Cart/CouponCodeInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _CouponCodeInputApplyCouponMutation2 = _interopRequireDefault(require("../../__generated__/CouponCodeInputApplyCouponMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Icons = require("../Icons");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const applyCouponMutation = (_CouponCodeInputApplyCouponMutation2.default.hash && _CouponCodeInputApplyCouponMutation2.default.hash !== "c53b64e200d60a461b3f47f15578e34e" && console.error("The definition of 'CouponCodeInputApplyCouponMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CouponCodeInputApplyCouponMutation2.default);
function CouponCodeInput() {
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const [expanded, setExpanded] = (0, _react.useState)(false);
  const [value, setValue] = (0, _react.useState)('');
  const [error, setError] = (0, _react.useState)(null);
  const inputRef = (0, _react.useRef)(null);
  const applyCoupon = () => {
    if (submitting) return;
    setSubmitting(true);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: applyCouponMutation,
      variables: {
        input: {
          code: value
        }
      },
      onCompleted: (_result, errors) => {
        setSubmitting(false);
        if (errors !== null && errors !== void 0 && errors[0]) {
          setError(errors === null || errors === void 0 ? void 0 : errors[0].message);
        }
      }
    });
  };
  const handleSubmit = e => {
    e.preventDefault();
    applyCoupon();
  };
  const handleIconClick = () => {
    if (error) {
      var _inputRef$current;
      setValue('');
      setError(null);
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  };
  if (!expanded) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "cart-additions"
    }, /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      onClick: () => setExpanded(true),
      className: "cart-additions__expand-button"
    }, "Add a coupon code"));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-additions",
    "data-status": error ? 'error' : 'idle'
  }, /*#__PURE__*/_react.default.createElement("form", {
    onSubmit: handleSubmit,
    className: "cart-additions__form"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-additions__field-wrapper"
  }, /*#__PURE__*/_react.default.createElement("input", {
    ref: inputRef,
    type: "text",
    value: value,
    onChange: e => {
      setValue(e.target.value);
      setError(null);
    },
    placeholder: "Enter code",
    size: 21,
    className: "cart-additions__field"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-additions__state-icon",
    onClick: handleIconClick
  }, error ? /*#__PURE__*/_react.default.createElement(_Icons.X, null) : null)), /*#__PURE__*/_react.default.createElement("button", {
    type: "submit",
    className: "cart-additions__button"
  }, "Apply")), error && /*#__PURE__*/_react.default.createElement("p", {
    className: "cart-additions__error"
  }, error));
}
var _default = CouponCodeInput;
exports.default = _default;
