// fontdue-js/dist/components/Cart/CustomerFields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _User = _interopRequireDefault(require("../Icons/User"));
var _Check = _interopRequireDefault(require("../Icons/Check"));
var _X = _interopRequireDefault(require("../Icons/X"));
var _TextField = _interopRequireDefault(require("../TextField"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const notEmpty = value => value != null;
const CustomerFields = _ref => {
  let {
    title,
    value,
    onChange,
    optInLabel,
    errors
  } = _ref;
  const {
    form: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const getErrors = field => {
    var _errors$field;
    return errors === null || errors === void 0 ? void 0 : (_errors$field = errors[field]) === null || _errors$field === void 0 ? void 0 : _errors$field.filter(notEmpty);
  };
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, title ? /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__header"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__customer-info-message"
  }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_User.default, null), " ", title)))) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Name *",
    placeholder: "Your name",
    autocomplete: "name",
    value: value.name || '',
    onChange: v => onChange({
      name: v
    }),
    errors: getErrors('name')
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Email *",
    placeholder: "Your email",
    type: "email",
    autocomplete: "email",
    value: value.email || '',
    onChange: v => onChange({
      email: v
    }),
    errors: getErrors('email')
  })), optInLabel && /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__section"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkout-modal__checkbox-field"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox"
  }, /*#__PURE__*/_react.default.createElement("input", {
    type: "checkbox",
    id: "customer-newsletter-opt-in",
    checked: value.newsletterOptIn ?? false,
    onChange: e => onChange({
      newsletterOptIn: e.target.checked
    })
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "checkbox__icon"
  }, config.checkboxStyle === 'check' ? /*#__PURE__*/_react.default.createElement(_Check.default, null) : /*#__PURE__*/_react.default.createElement(_X.default, null))), /*#__PURE__*/_react.default.createElement("label", {
    htmlFor: "customer-newsletter-opt-in",
    dangerouslySetInnerHTML: {
      __html: optInLabel
    }
  }))));
};
var _default = CustomerFields;
exports.default = _default;
