// fontdue-js/dist/components/Cart/IdentityBox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = IdentityBox;
var _IdentityBox_identity2 = _interopRequireDefault(require("../../__generated__/IdentityBox_identity.graphql"));
var _reactRelay = require("react-relay");
var _react = _interopRequireDefault(require("react"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function IdentityBox(_ref) {
  let {
    identity: identityKey,
    title,
    onClick,
    children: {
      icon
    }
  } = _ref;
  const identity = (0, _reactRelay.useFragment)((_IdentityBox_identity2.default.hash && _IdentityBox_identity2.default.hash !== "d93d9cb6f2c86580fb57b4de6782ec9d" && console.error("The definition of 'IdentityBox_identity' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _IdentityBox_identity2.default), identityKey);
  return /*#__PURE__*/_react.default.createElement("button", {
    className: "cart__review__button",
    type: "button",
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__label"
  }, icon, title, " ", /*#__PURE__*/_react.default.createElement("span", {
    className: "cart__review__label__text"
  }, "Edit")), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, /*#__PURE__*/_react.default.createElement("pre", null, identity.addressFormatted)), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, identity.email), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, identity.phoneFormatted), identity.vatNumber && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__review__section"
  }, "VAT: ", identity.vatNumber));
}
