// fontdue-js/dist/components/Cart/orderTracking.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sendOrderTracking = sendOrderTracking;
var _orderTrackingUpdateOrderTrackingMutation2 = _interopRequireDefault(require("../../__generated__/orderTrackingUpdateOrderTrackingMutation.graphql"));
var _reactRelay = require("react-relay");
var _consent = require("../ConsentBanner/consent");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function readCookie(name) {
  const match = document.cookie.match(new RegExp('(?:^|;\\s*)' + name + '=([^;]*)'));
  return match ? decodeURIComponent(match[1]) : undefined;
}

/**
 * Stores the buyer's analytics context (cookie consent, Meta browser IDs,
 * checkout page URL) on the current order. The server emits the purchase
 * event from a Stripe webhook — outside the browser — so this is how that
 * event respects the consent banner and carries attribution.
 *
 * Fire-and-forget: tracking must never break checkout.
 */
function sendOrderTracking(environment) {
  try {
    (0, _reactRelay.commitMutation)(environment, {
      mutation: (_orderTrackingUpdateOrderTrackingMutation2.default.hash && _orderTrackingUpdateOrderTrackingMutation2.default.hash !== "d59a127a7f140424f507ae549731bac7" && console.error("The definition of 'orderTrackingUpdateOrderTrackingMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _orderTrackingUpdateOrderTrackingMutation2.default),
      variables: {
        input: {
          analyticsConsent: (0, _consent.hasConsent)('analytics'),
          anonymousId: (0, _consent.getClientAnonymousId)(),
          fbp: readCookie('_fbp'),
          fbc: readCookie('_fbc'),
          url: window.location.href
        }
      },
      onCompleted: () => undefined,
      onError: () => undefined
    });
  } catch {
    // Ignore — see above.
  }
}
