// fontdue-js/dist/components/Cart/ShareCart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ShareCart;
var _ShareCartCreateSnapshotMutation2 = _interopRequireDefault(require("../../__generated__/ShareCartCreateSnapshotMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Icons = require("../Icons");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ShareCart() {
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [isSharing, setIsSharing] = (0, _react.useState)(false);
  const [snapshot, setSnapshot] = (0, _react.useState)(null);
  const [shareError, setShareError] = (0, _react.useState)(null);
  const [copied, setCopied] = (0, _react.useState)(false);
  const inputRef = (0, _react.useRef)(null);
  const handleShareCart = () => {
    setIsSharing(true);
    setShareError(null);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: (_ShareCartCreateSnapshotMutation2.default.hash && _ShareCartCreateSnapshotMutation2.default.hash !== "320cfd27011ff030005b0ca1cf90adf7" && console.error("The definition of 'ShareCartCreateSnapshotMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _ShareCartCreateSnapshotMutation2.default),
      variables: {},
      onCompleted: (response, errors) => {
        var _response$createOrder;
        setIsSharing(false);
        if (errors && errors.length > 0) {
          setShareError(errors[0].message);
        } else if ((_response$createOrder = response.createOrderSnapshot) !== null && _response$createOrder !== void 0 && _response$createOrder.orderSnapshot) {
          setSnapshot(response.createOrderSnapshot.orderSnapshot);
        } else {
          setShareError('Failed to create shareable link');
        }
      },
      onError: error => {
        setIsSharing(false);
        setShareError(error.message);
      }
    });
  };
  (0, _react.useEffect)(() => {
    var _inputRef$current;
    (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.select();
  }, [snapshot]);
  const timeout = (0, _react.useRef)(null);
  const handleCopyToClipboard = () => {
    if (snapshot !== null && snapshot !== void 0 && snapshot.shareUrl) {
      navigator.clipboard.writeText(snapshot.shareUrl);
      setCopied(true);
      if (timeout.current) clearTimeout(timeout.current);
      timeout.current = setTimeout(() => {
        setCopied(false);
      }, 1000);
    }
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, snapshot !== null && snapshot !== void 0 && snapshot.shareUrl && snapshot !== null && snapshot !== void 0 && snapshot.quotePdfUrl ? /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__share"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__share-title"
  }, "Share your cart"), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__share-url-input-container"
  }, /*#__PURE__*/_react.default.createElement("input", {
    className: "cart__share-url-input",
    ref: inputRef,
    value: snapshot.shareUrl ?? '',
    readOnly: true,
    onClick: e => e.currentTarget.select()
  }), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "cart__share__button cart__share__copy-button",
    onClick: handleCopyToClipboard,
    "data-copied": copied
  }, /*#__PURE__*/_react.default.createElement("span", null, "Copy"), /*#__PURE__*/_react.default.createElement(_Icons.Check, null))), /*#__PURE__*/_react.default.createElement("a", {
    href: snapshot.quotePdfUrl,
    className: "cart__share__button cart__share__pdf-button",
    target: "_blank",
    rel: "noopener noreferrer"
  }, "Download PDF quote")) : /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "cart__share__button cart__share__share-button",
    onClick: handleShareCart,
    disabled: isSharing
  }, isSharing ? 'Loading...' : 'Share cart'), shareError && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart__share-error"
  }, "Cannot share your cart right now. Please try again later."));
}
