// fontdue-js/dist/components/TypeTester/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _TypeTester_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTester_fontStyle.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _FontStyle = _interopRequireDefault(require("../FontStyle"));
var _SKUPrice = _interopRequireDefault(require("../SKUPrice"));
var _SelectButton = _interopRequireDefault(require("../SelectButton"));
var _utils = require("../../utils");
var _TypeTesterStyleSelectData = _interopRequireDefault(require("./TypeTesterStyleSelectData"));
var _TypeTesterFeatures = _interopRequireDefault(require("./TypeTesterFeatures"));
var _TypeTesterContent = _interopRequireDefault(require("./TypeTesterContent"));
var _TypeTesterState = _interopRequireDefault(require("./TypeTesterState"));
var _TypeTesterToolbar = _interopRequireDefault(require("./TypeTesterToolbar"));
var _TypeTesterVariableAxes = _interopRequireDefault(require("./TypeTesterVariableAxes"));
var _hooks = require("../../hooks");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const TypeTester = _ref => {
  let {
    id,
    productId,
    fontStyle: fontStyleKey,
    tags,
    onFocus,
    onBlur,
    features,
    axes,
    groupEdit,
    onStyleSelect,
    isLoading
  } = _ref;
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTester_fontStyle2.default.hash && _TypeTester_fontStyle2.default.hash !== "bf89a32c8a0661f2102bbcf8c6bd1aa7" && console.error("The definition of 'TypeTester_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTester_fontStyle2.default), fontStyleKey);
  const contentRef = (0, _react.useRef)(null);
  const ref = (0, _react.useRef)(null);
  const [buyButtonIsHovered, setBuyButtonIsHovered] = (0, _react.useState)(false);
  let {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const axesPosition = config.variableAxesPosition === 'auto' ? ((axes === null || axes === void 0 ? void 0 : axes.length) ?? 0) > 1 ? 'features-panel' : 'inline' : config.variableAxesPosition;
  const props = (0, _TypeTesterState.default)({
    id
  });

  // The instance name (e.g. "Regular" → "Custom") is a flex sibling of the
  // inline axis slider, so updating it mid-drag changes its width and shifts
  // the slider. Debounce the value used purely for the name display — the font
  // preview and sliders keep reading the live `props.variableSettings`.
  // Keyed on the font style id so switching fonts snaps the instance name
  // immediately (in sync with the new style) rather than lagging the switch.
  const nameVariableSettings = (0, _hooks.useDebouncedValue)(props.variableSettings, 200, fontStyle === null || fontStyle === void 0 ? void 0 : fontStyle.id);
  (0, _react.useEffect)(() => {
    if (onFocus && props.focused) onFocus();
    if (onBlur && !props.focused) onBlur();
  }, [props.focused]);
  const selected = (0, _reactRedux.useSelector)(fontStyle && fontStyle.sku ? (0, _utils.isSelected)(fontStyle.sku.id) : () => false);
  const precartOpen = (0, _reactRedux.useSelector)(state => state.precartOpen);
  const sku = fontStyle === null || fontStyle === void 0 ? void 0 : fontStyle.sku;
  const onNeedsRefocus = () => {
    var _contentRef$current;
    (_contentRef$current = contentRef.current) === null || _contentRef$current === void 0 ? void 0 : _contentRef$current.focus();
  };
  if (!fontStyle) return null;
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester select-button-hover-parent",
    ref: ref,
    "data-type-tester-id": id,
    "data-focused": props.focused,
    "data-selected": selected,
    "data-buy-button-hovered": buyButtonIsHovered,
    "data-large-font-size": props.size > 72,
    "data-direction": props.direction,
    "data-autofit": props.autofit,
    "data-shy": config.shy,
    "data-select-button-style": config.selectButtonStyle,
    "data-alignment": props.alignment,
    "data-loading": isLoading,
    "data-tags": tags === null || tags === void 0 ? void 0 : tags.join(' ')
  }, axesPosition === 'above' ? /*#__PURE__*/_react.default.createElement(_TypeTesterVariableAxes.default, {
    id: id,
    axes: axes,
    fontStyle: fontStyle
  }) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__toolbar"
  }, /*#__PURE__*/_react.default.createElement(_TypeTesterStyleSelectData.default, {
    fontStyle: fontStyle,
    variableSettings: nameVariableSettings,
    onSelectFontStyleValue: _ref2 => {
      let {
        fontStyleId,
        variableSettings
      } = _ref2;
      onStyleSelect(fontStyleId);
      props.setVariableSettings(variableSettings ?? null);
      props.setFeatures([]);
      onNeedsRefocus();
    },
    onSelectVariableAxes: () => {
      props.setFeaturesOpen(true);
    },
    config: config,
    includeVariableAxesOption: Boolean(axes === null || axes === void 0 ? void 0 : axes.length) && axesPosition === 'features-panel'
  }), config.priceText && fontStyle.sku && /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__toolbar__price-text",
    onMouseDown: e => e.preventDefault()
  }, /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
    sku: fontStyle.sku
  })), config.toolsPosition === 'inline' && /*#__PURE__*/_react.default.createElement(_TypeTesterToolbar.default, {
    id: id,
    features: features,
    axes: axes,
    axesPosition: axesPosition,
    fontStyle: fontStyle,
    onNeedsRefocus: onNeedsRefocus,
    groupEdit: groupEdit
  }), config.selectButton && sku && productId && /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__select-button-wrapper",
    onMouseDown: e => e.preventDefault()
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__select-button"
  }, /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_SelectButton.default, {
    productId: productId,
    sku: sku,
    onMouseEnter: () => setBuyButtonIsHovered(true),
    onMouseLeave: () => setBuyButtonIsHovered(false),
    className: "type-tester__select-button-element"
  }), precartOpen && !selected && /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__price"
  }, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: null
  }, /*#__PURE__*/_react.default.createElement(_SKUPrice.default, {
    sku: sku
  }))))))), config.toolsPosition === 'inline' && (features || axes) && /*#__PURE__*/_react.default.createElement(_TypeTesterFeatures.default, {
    id: id,
    fontStyle: fontStyle,
    features: features ?? null,
    axes: axesPosition === 'features-panel' ? axes ?? null : null
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__text"
  }, /*#__PURE__*/_react.default.createElement(_FontStyle.default, {
    fontStyle: fontStyle,
    className: "type-tester__text__font-style"
  }, _ref3 => {
    let {
      loaded,
      style,
      verticalMetrics
    } = _ref3;
    if (loaded) {
      return /*#__PURE__*/_react.default.createElement(_TypeTesterContent.default, _extends({}, props, {
        ref: contentRef,
        fontStyles: style,
        verticalMetrics: verticalMetrics,
        truncate: config.truncate,
        direction: props.direction,
        min: config.size.min,
        max: config.size.max,
        autofitOnChange: config.autofitOnChange,
        columns: config.columns
      }));
    }
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "type-tester__text__container",
      style: {
        fontSize: `${props.size}px`,
        textAlign: props.alignment
      }
    }, props.content.getCurrentContent().getPlainText());
  }))));
};
var _default = TypeTester;
exports.default = _default;
