// fontdue-js/dist/components/TypeTester/TypeTesterFamilies.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TypeTesterFamiliesProvider;
exports.typeTesterFamiliesStylesQuery = exports.typeTesterFamiliesQuery = void 0;
exports.useTypeTesterFamilies = useTypeTesterFamilies;
var _TypeTesterFamiliesStylesQuery2 = _interopRequireDefault(require("../../__generated__/TypeTesterFamiliesStylesQuery.graphql"));
var _TypeTesterFamiliesQuery2 = _interopRequireDefault(require("../../__generated__/TypeTesterFamiliesQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _utils = require("../../utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// A single, lightweight family-list entry for the type tester family switcher.
// These are all cheap stored columns (plus the once-per-page language filter),
// so this query never triggers per-style font-binary parsing.

const noop = () => {};
const TypeTesterFamiliesContext = /*#__PURE__*/(0, _react.createContext)({
  families: null,
  loadingFamilies: false,
  loadFamilies: noop,
  getFamilyStyles: () => undefined,
  loadFamilyStyles: noop
});
function useTypeTesterFamilies() {
  return (0, _react.useContext)(TypeTesterFamiliesContext);
}

// The family-switcher dropdown list. Cheap fields only; `featureStyle
// { supportedLanguages }` resolves the language filter once per page (instead
// of once per type tester) — see FD-685.
const typeTesterFamiliesQuery = (_TypeTesterFamiliesQuery2.default.hash && _TypeTesterFamiliesQuery2.default.hash !== "cda53ca94a4951ee5ba29cc57281a002" && console.error("The definition of 'TypeTesterFamiliesQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterFamiliesQuery2.default);

// The styles/instances for a single family, fetched lazily when that family is
// opened or switched to. `variableInstances` font-binary parsing is therefore
// scoped to one family rather than the whole library.
exports.typeTesterFamiliesQuery = typeTesterFamiliesQuery;
const typeTesterFamiliesStylesQuery = (_TypeTesterFamiliesStylesQuery2.default.hash && _TypeTesterFamiliesStylesQuery2.default.hash !== "feee079c690b83b56cfb01cb81f36dce" && console.error("The definition of 'TypeTesterFamiliesStylesQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterFamiliesStylesQuery2.default);
exports.typeTesterFamiliesStylesQuery = typeTesterFamiliesStylesQuery;
function TypeTesterFamiliesProvider(_ref) {
  let {
    children
  } = _ref;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [families, setFamilies] = (0, _react.useState)(null);
  const [loadingFamilies, setLoadingFamilies] = (0, _react.useState)(false);
  const familiesRequested = (0, _react.useRef)(false);
  const [familyStyles, setFamilyStyles] = (0, _react.useState)(() => new Map());
  const familyStylesRequested = (0, _react.useRef)(new Set());
  const loadFamilies = (0, _react.useCallback)(() => {
    if (familiesRequested.current) return;
    familiesRequested.current = true;
    setLoadingFamilies(true);
    (0, _reactRelay.fetchQuery)(environment, typeTesterFamiliesQuery, {}).toPromise().then(data => {
      var _data$viewer, _data$viewer$families;
      const list = ((data === null || data === void 0 ? void 0 : (_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : (_data$viewer$families = _data$viewer.families) === null || _data$viewer$families === void 0 ? void 0 : _data$viewer$families.edges) ?? []).map(edge => edge === null || edge === void 0 ? void 0 : edge.node).filter(_utils.notEmpty).map(node => ({
        id: node.id,
        name: node.name,
        isVariableFont: node.isVariableFont,
        featureStyle: node.featureStyle ? {
          supportedLanguages: node.featureStyle.supportedLanguages ?? null
        } : null
      }));
      setFamilies(list);
    }).catch(() => {
      // allow a retry on the next open if the request failed
      familiesRequested.current = false;
    }).finally(() => setLoadingFamilies(false));
  }, [environment]);
  const setEntry = (0, _react.useCallback)((familyId, entry) => {
    setFamilyStyles(prev => {
      const next = new Map(prev);
      next.set(familyId, entry);
      return next;
    });
  }, []);
  const loadFamilyStyles = (0, _react.useCallback)(familyId => {
    if (!familyId || familyStylesRequested.current.has(familyId)) return;
    familyStylesRequested.current.add(familyId);
    setEntry(familyId, {
      loading: true,
      isVariableFont: false,
      styles: null
    });
    (0, _reactRelay.fetchQuery)(environment, typeTesterFamiliesStylesQuery, {
      familyId
    }).toPromise().then(data => {
      const node = data === null || data === void 0 ? void 0 : data.node;
      const fontStyles = node && 'fontStyles' in node ? node.fontStyles : null;
      const isVariableFont = node && 'isVariableFont' in node ? node.isVariableFont : false;
      setEntry(familyId, {
        loading: false,
        isVariableFont: isVariableFont ?? false,
        styles: (fontStyles ?? []).map(style => ({
          id: style.id,
          name: style.name,
          cssWeight: style.cssWeight ?? null,
          cssStyle: style.cssStyle ?? null,
          cssStretch: style.cssStretch ?? null,
          variableInstances: style.variableInstances ?? null
        }))
      });
    }).catch(() => {
      // allow a retry on the next open if the request failed
      familyStylesRequested.current.delete(familyId);
      setEntry(familyId, {
        loading: false,
        isVariableFont: false,
        styles: null
      });
    });
  }, [environment, setEntry]);
  const getFamilyStyles = (0, _react.useCallback)(familyId => familyStyles.get(familyId), [familyStyles]);
  return /*#__PURE__*/_react.default.createElement(TypeTesterFamiliesContext.Provider, {
    value: {
      families,
      loadingFamilies,
      loadFamilies,
      getFamilyStyles,
      loadFamilyStyles
    }
  }, children);
}
