// fontdue-js/dist/components/TypeTester/TypeTesterStyleSelect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _utils = require("../../utils");
var _Select = _interopRequireDefault(require("../Select"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const VARIABLE_AXES = 'VARIABLE_AXES';
const LOADING = '__loading__';
const TypeTesterStyleSelect = _ref => {
  var _selectedFamily$fontS4;
  let {
    families,
    selectedFamily,
    familyValue,
    selectedFontStyleId,
    selectedVariableSettings,
    loadingFamilies,
    loading,
    onOpen,
    onSelectFontStyleValue,
    onSelectVariableAxes,
    onSelectFamilyId,
    includeVariableAxesOption
  } = _ref;
  const handleSelectFontStyle = selection => {
    if (selection === VARIABLE_AXES) {
      onSelectVariableAxes();
    } else if (selection.includes(':::')) {
      var _selectedFamily$fontS, _selectedFamily$fontS2, _selectedFamily$fontS3;
      const [fontStyleId, instanceName] = selection.split(':::');
      const instance = selectedFamily === null || selectedFamily === void 0 ? void 0 : (_selectedFamily$fontS = selectedFamily.fontStyles) === null || _selectedFamily$fontS === void 0 ? void 0 : (_selectedFamily$fontS2 = _selectedFamily$fontS.find(style => style.id === fontStyleId)) === null || _selectedFamily$fontS2 === void 0 ? void 0 : (_selectedFamily$fontS3 = _selectedFamily$fontS2.variableInstances) === null || _selectedFamily$fontS3 === void 0 ? void 0 : _selectedFamily$fontS3.find(inst => inst.name === instanceName);
      let variableSettings;
      if (instance) variableSettings = (0, _utils.variableInstanceSettings)(instance);
      onSelectFontStyleValue({
        fontStyleId,
        variableSettings
      });
    } else {
      onSelectFontStyleValue({
        fontStyleId: selection
      });
    }
  };
  const familyOptions = families.map(_ref2 => {
    let {
      id,
      name
    } = _ref2;
    return {
      value: id,
      text: name
    };
  });
  if (loadingFamilies) familyOptions.push({
    value: LOADING,
    text: 'Loading…'
  });
  let isVariable = false;
  const styleOptions = (selectedFamily === null || selectedFamily === void 0 ? void 0 : (_selectedFamily$fontS4 = selectedFamily.fontStyles) === null || _selectedFamily$fontS4 === void 0 ? void 0 : _selectedFamily$fontS4.flatMap(_ref3 => {
    let {
      id,
      name,
      variableInstances
    } = _ref3;
    if (variableInstances && variableInstances.length > 0) {
      isVariable = true;
      return variableInstances.map(instance => ({
        value: `${id}:::${instance.name}`,
        text: instance.name
      }));
    }
    return [{
      value: id,
      text: name
    }];
  })) ?? [];
  if (isVariable && includeVariableAxesOption) {
    styleOptions.push({
      value: VARIABLE_AXES,
      text: 'Variable axes...'
    });
  }
  let styleValue = selectedFontStyleId;
  if (selectedVariableSettings) {
    var _selectedFamily$fontS5, _selectedFamily$fontS6, _selectedFamily$fontS7;
    const instance = selectedFamily === null || selectedFamily === void 0 ? void 0 : (_selectedFamily$fontS5 = selectedFamily.fontStyles) === null || _selectedFamily$fontS5 === void 0 ? void 0 : (_selectedFamily$fontS6 = _selectedFamily$fontS5.find(style => style.id === selectedFontStyleId)) === null || _selectedFamily$fontS6 === void 0 ? void 0 : (_selectedFamily$fontS7 = _selectedFamily$fontS6.variableInstances) === null || _selectedFamily$fontS7 === void 0 ? void 0 : _selectedFamily$fontS7.find(inst => (0, _utils.compareVariableSettings)(inst, selectedVariableSettings));
    styleValue = instance ? `${selectedFontStyleId}:::${instance.name}` : null;
  }

  // Ignore the synthetic "Loading…" placeholder if it is somehow selected.
  const ignoreLoading = cb => e => {
    if (e.target.value === LOADING) return;
    cb(e.target.value);
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester-select",
    "data-loading": loading || undefined,
    "aria-busy": loading || undefined,
    style: {
      opacity: loading ? 0.6 : 1,
      transition: 'opacity 120ms ease'
    }
  }, /*#__PURE__*/_react.default.createElement(_Select.default, {
    options: familyOptions,
    value: familyValue,
    onChange: ignoreLoading(onSelectFamilyId),
    onMouseDown: () => onOpen === null || onOpen === void 0 ? void 0 : onOpen(),
    onFocus: () => onOpen === null || onOpen === void 0 ? void 0 : onOpen(),
    onClick: e => e.stopPropagation()
  }), styleOptions.length > 0 && /*#__PURE__*/_react.default.createElement(_Select.default, {
    options: styleOptions,
    value: styleValue,
    nullText: "Custom",
    onChange: ignoreLoading(handleSelectFontStyle),
    onMouseDown: () => onOpen === null || onOpen === void 0 ? void 0 : onOpen(),
    onFocus: () => onOpen === null || onOpen === void 0 ? void 0 : onOpen(),
    onClick: e => e.stopPropagation()
  }));
};
var _default = TypeTesterStyleSelect;
exports.default = _default;
