// fontdue-js/dist/components/TypeTester/TypeTesterStandaloneElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TypeTesterStandaloneElement = void 0;
var _react = _interopRequireDefault(require("react"));
var _TypeTesterStandalone = _interopRequireDefault(require("./TypeTesterStandalone"));
var _utils = require("../../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const isAlignment = str => {
  return str === 'left' || str === 'center' || str === 'right';
};
const isDirection = str => {
  return str === 'ltr' || str === 'rtl';
};
const parseBool = input => input === 'true' ? true : false;

// parse a string like "wdth 1000, ital 0.5" into
// [{ axis: "wdth", value: 1000 }, { axis: "ital", value: 0.5 }]
function parseVariableSettings(input) {
  if (!input) return null;
  return input.split(/\s*,\s*/).map(settingString => {
    const m = settingString.match(/^([A-Za-z0-9]{4}) (\d+(?:\.\d+)?)$/);
    if (m && m.length > 2) {
      return {
        axis: m[1],
        value: parseFloat(m[2])
      };
    } else {
      console.warn(`fontdue-type-tester failed to parse variable-settings value: "${settingString}", ignoring`);
    }
  }).filter(_utils.notEmpty);
}
const TypeTesterStandaloneElement = _ref => {
  let {
    lineHeight,
    letterSpacing,
    fontSize,
    axes,
    features,
    featuresSelected,
    variableSettings,
    alignment,
    direction,
    autofit,
    ...rest
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_TypeTesterStandalone.default, _extends({
    autofit: parseBool(autofit),
    lineHeight: lineHeight ? parseFloat(lineHeight) : undefined,
    letterSpacing: letterSpacing ? parseFloat(letterSpacing) : undefined,
    fontSize: fontSize ? parseFloat(fontSize) : undefined,
    axes: (axes === null || axes === void 0 ? void 0 : axes.split(',')) ?? [],
    features: (features === null || features === void 0 ? void 0 : features.split(',')) ?? [],
    featureSettings: featuresSelected === null || featuresSelected === void 0 ? void 0 : featuresSelected.split(',').map(feature => ({
      feature,
      value: '1'
    })),
    alignment: isAlignment(alignment) ? alignment : undefined,
    direction: isDirection(direction) ? direction : undefined,
    variableSettings: parseVariableSettings(variableSettings)
  }, rest));
};
exports.TypeTesterStandaloneElement = TypeTesterStandaloneElement;
var _default = TypeTesterStandaloneElement;
exports.default = _default;
