// fontdue-js/dist/components/TypeTester/TypeTesterStandalone.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeTesterStandalonePreloadedQueryRenderer = TypeTesterStandalonePreloadedQueryRenderer;
exports.default = TypeTesterStandalone;
var _TypeTesterStandaloneQuery2 = _interopRequireDefault(require("../../__generated__/TypeTesterStandaloneQuery.graphql"));
var _TypeTesterStandaloneChangedStylesQuery2 = _interopRequireDefault(require("../../__generated__/TypeTesterStandaloneChangedStylesQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _index = _interopRequireDefault(require("./index"));
var _TypeTesterContext = _interopRequireDefault(require("./TypeTesterContext"));
var _types = require("./types");
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _useLicenseAndOrderVariables = require("../../hooks/useLicenseAndOrderVariables");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const changedStylesQuery = (_TypeTesterStandaloneChangedStylesQuery2.default.hash && _TypeTesterStandaloneChangedStylesQuery2.default.hash !== "98374227c6b2c3eceab5b48138e1a662" && console.error("The definition of 'TypeTesterStandaloneChangedStylesQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterStandaloneChangedStylesQuery2.default);
function TypeTesterStandaloneComponent(_ref) {
  var _data$viewer, _data$viewer2, _data$viewer2$fontSty, _data$viewer2$fontSty2, _data$viewer2$fontSty3, _data$viewer3, _data$viewer3$fontSty, _data$viewer3$fontSty2;
  let {
    content,
    direction,
    axes,
    features,
    featuresSelected,
    featureSettings = null,
    variableSettings = null,
    lineHeight,
    letterSpacing,
    alignment,
    fontSize,
    autofit,
    onFocus,
    onBlur,
    data
  } = _ref;
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  let defaultAlignment = config.initialAlignment;
  if (direction === 'rtl') defaultAlignment = 'right';
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [changedStyle, setChangedStyle] = (0, _react.useState)('');
  const [changedStylesFetchKey, setChangedStylesFetchKey] = (0, _react.useState)(0);
  const {
    licenseOptionsSpecs,
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  const handleStyleSelect = (0, _react.useCallback)(fontStyleId => {
    (0, _reactRelay.fetchQuery)(environment, changedStylesQuery, {
      id: fontStyleId,
      licenseOptions: licenseOptionsSpecs,
      orderVariables: orderVariableSelections
    }).subscribe({
      complete: () => {
        setChangedStyle(fontStyleId);
        setChangedStylesFetchKey(prev => prev + 1);
      }
    });
  }, [environment, changedStyle, licenseOptionsSpecs, orderVariableSelections]);
  const changedStylesData = (0, _reactRelay.useLazyLoadQuery)(changedStylesQuery, {
    id: changedStyle,
    licenseOptions: licenseOptionsSpecs,
    orderVariables: orderVariableSelections
  }, {
    fetchKey: changedStylesFetchKey,
    fetchPolicy: 'store-only'
  });
  return /*#__PURE__*/_react.default.createElement(_TypeTesterContext.default, {
    testers: {
      standalone: {
        content: content ?? 'The quick brown fox...',
        autofit: autofit ?? false,
        size: fontSize ?? 72,
        lineHeight: lineHeight ?? 1,
        letterSpacing: letterSpacing ?? 0,
        alignment: alignment ?? defaultAlignment,
        features: (featureSettings === null || featureSettings === void 0 ? void 0 : featureSettings.filter(setting => setting.value === '1').map(setting => setting.feature)) ?? featuresSelected ?? null,
        variableSettings: (variableSettings === null || variableSettings === void 0 ? void 0 : variableSettings.reduce((acc, _ref2) => {
          let {
            axis,
            value
          } = _ref2;
          acc[axis] = value;
          return acc;
        }, {})) ?? null,
        direction: direction ?? 'ltr'
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_index.default, {
    id: "standalone",
    fontStyle: changedStyle ? changedStylesData.node : (_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : _data$viewer.fontStyle,
    onStyleSelect: handleStyleSelect,
    axes: axes,
    features: (0, _types.normalizeFeaturesProp)(features),
    productId: ((_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : (_data$viewer2$fontSty = _data$viewer2.fontStyle) === null || _data$viewer2$fontSty === void 0 ? void 0 : (_data$viewer2$fontSty2 = _data$viewer2$fontSty.family) === null || _data$viewer2$fontSty2 === void 0 ? void 0 : (_data$viewer2$fontSty3 = _data$viewer2$fontSty2.parent) === null || _data$viewer2$fontSty3 === void 0 ? void 0 : _data$viewer2$fontSty3.id) ?? ((_data$viewer3 = data.viewer) === null || _data$viewer3 === void 0 ? void 0 : (_data$viewer3$fontSty = _data$viewer3.fontStyle) === null || _data$viewer3$fontSty === void 0 ? void 0 : (_data$viewer3$fontSty2 = _data$viewer3$fontSty.family) === null || _data$viewer3$fontSty2 === void 0 ? void 0 : _data$viewer3$fontSty2.id),
    groupEdit: false,
    onFocus: onFocus,
    onBlur: onBlur
  }));
}
const query = (_TypeTesterStandaloneQuery2.default.hash && _TypeTesterStandaloneQuery2.default.hash !== "951214eacb2370ea3ca0b19bebe7fbe7" && console.error("The definition of 'TypeTesterStandaloneQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterStandaloneQuery2.default);
function TypeTesterStandalonePreloadedQueryRenderer(_ref3) {
  let {
    preloadedQuery,
    ...rest
  } = _ref3;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(query, queryRef);
  return /*#__PURE__*/_react.default.createElement(TypeTesterStandaloneComponent, _extends({
    data: data
  }, rest));
}
function TypeTesterStandalone(_ref4) {
  var _data$viewer4;
  let {
    familyName,
    styleName,
    ...rest
  } = _ref4;
  const data = (0, _reactRelay.useLazyLoadQuery)(query, {
    familyName,
    styleName
  });
  const fontStyle = (_data$viewer4 = data.viewer) === null || _data$viewer4 === void 0 ? void 0 : _data$viewer4.fontStyle;
  if (!fontStyle) {
    console.log(`[TypeTester]: No FontStyle found for {familyName: "${familyName}", styleName: "${styleName}"}`);
  }
  return /*#__PURE__*/_react.default.createElement(TypeTesterStandaloneComponent, _extends({
    data: data
  }, rest));
}
