// fontdue-js/dist/components/TypeTester/TypeTesterInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TypeTesterInput = void 0;
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const TypeTesterInput = _ref => {
  let {
    value,
    onChange,
    min,
    max,
    setFocused
  } = _ref;
  const [localValue, setLocalValue] = (0, _react.useState)(value);
  const [isError, setIsError] = (0, _react.useState)(false);
  (0, _react.useEffect)(() => {
    setLocalValue(Math.round(value));
  }, [value]);
  const propagateValue = (0, _react.useCallback)(() => {
    if (localValue && localValue >= min && localValue <= max) {
      setIsError(false);
      onChange(localValue);
    } else {
      setIsError(true);
    }
  }, [localValue, onChange, setIsError]);
  const handleChange = (0, _react.useCallback)(e => {
    if (e.target.value === '') {
      return setLocalValue(null);
    }
    const parsed = parseFloat(e.target.value);
    if (!Number.isNaN(parsed)) {
      setLocalValue(parsed);
    }
  }, []);
  const handleKeyDown = (0, _react.useCallback)(e => {
    if (e.key === 'Enter') propagateValue();
    if (localValue) {
      const distance = e.shiftKey ? 10 : 1;
      if (e.key === 'ArrowUp') {
        e.preventDefault();
        onChange(Math.min(max, localValue + distance));
      }
      if (e.key === 'ArrowDown') {
        e.preventDefault();
        onChange(Math.max(min, localValue - distance));
      }
    }
  }, [propagateValue, localValue]);
  return /*#__PURE__*/_react.default.createElement("label", {
    className: "type-tester__toolbar__label",
    "data-label": "size-input",
    "data-is-error": isError
  }, /*#__PURE__*/_react.default.createElement("input", {
    className: "type-tester__toolbar__input",
    type: "text",
    style: {
      width: `${(localValue === null || localValue === void 0 ? void 0 : localValue.toString().length) ?? 2}ch`
    },
    value: localValue || '',
    onChange: handleChange,
    onKeyDown: handleKeyDown,
    onFocus: () => setFocused(true),
    onBlur: () => setFocused(false)
  }), /*#__PURE__*/_react.default.createElement("span", null, "px"));
};
exports.TypeTesterInput = TypeTesterInput;
var _default = TypeTesterInput;
exports.default = _default;
