// fontdue-js/dist/components/TypeTester/TypeTesterSlider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactRanger = require("../../react-ranger");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const logInterpolator = {
  getPercentageForValue: (val, min, max) => {
    const minSign = Math.sign(min);
    const maxSign = Math.sign(max);
    if (minSign !== maxSign) {
      throw new Error('Error: logarithmic interpolation does not support ranges that cross 0.');
    }
    let percent = 100 / (Math.log10(Math.abs(max)) - Math.log10(Math.abs(min))) * (Math.log10(Math.abs(val)) - Math.log10(Math.abs(min)));
    if (minSign < 0) {
      // negative range, means we need to invert our percent because of the Math.abs above
      return 100 - percent;
    }
    return percent;
  },
  getValueForClientX: (clientX, trackDims, min, max) => {
    const {
      left,
      width
    } = trackDims; // FIXME
    let value = clientX - left;
    value *= Math.log10(max) - Math.log10(min);
    value /= width;
    value = Math.pow(10, Math.log10(min) + value);
    return value;
  }
};
const TypeTesterSlider = _ref => {
  let {
    min,
    max,
    stepSize,
    value,
    onChange,
    interpolator
  } = _ref;
  const handleChange = (0, _react.useCallback)(values => onChange(values[0]), [onChange]);
  const {
    getTrackProps,
    handles
  } = (0, _reactRanger.useRanger)({
    min,
    max,
    stepSize,
    values: [value],
    onChange: handleChange,
    onDrag: handleChange,
    interpolator: interpolator === 'log10' ? logInterpolator : undefined
  });
  return /*#__PURE__*/_react.default.createElement("div", getTrackProps({
    className: 'type-tester__slider__track'
  }), handles.map((_ref2, index) => {
    let {
      getHandleProps
    } = _ref2;
    const {
      key,
      ...handleProps
    } = getHandleProps({
      className: 'type-tester__slider__handle'
    });
    return /*#__PURE__*/_react.default.createElement("button", _extends({
      key: key ?? index
    }, handleProps));
  }));
};
var _default = TypeTesterSlider;
exports.default = _default;
