// fontdue-js/dist/components/TypeTester/TypeTesterStyleSelectData.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TypeTesterStyleSelectData;
var _TypeTesterStyleSelectData_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTesterStyleSelectData_fontStyle.graphql"));
var _TypeTesterStyleSelectData_fontStyleData2 = _interopRequireDefault(require("../../__generated__/TypeTesterStyleSelectData_fontStyleData.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _relayRuntime = require("relay-runtime");
var _utils = require("../../utils");
var _TypeTesterStyleSelect = _interopRequireDefault(require("./TypeTesterStyleSelect"));
var _TypeTesterFamilies = require("./TypeTesterFamilies");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
_TypeTesterStyleSelectData_fontStyleData2.default.hash && _TypeTesterStyleSelectData_fontStyleData2.default.hash !== "fb09207ccab08c7e73677486469a3c95" && console.error("The definition of 'TypeTesterStyleSelectData_fontStyleData' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterStyleSelectData_fontStyleData2.default;
// get a score for similarity between two styles
// 0 < n <= 1
// where 1 = exact match
const computeSimilarity = (left, right) => {
  let weightDiff = null;
  if (left.cssWeight && right.cssWeight) {
    weightDiff = Math.abs(parseInt(left.cssWeight, 10) - parseInt(right.cssWeight, 10));
  }

  // since weight class values go from 1 to 1000
  const weightSimilarity = weightDiff === null ? 0 : (1000 - weightDiff) / 1000;

  // use 0.9 for non matching styles, since we prefer weight but if there's a matching style
  // it shouldn't take preference over a closer weight
  const styleSimilarity = left.cssStyle === right.cssStyle ? 1 : 0.9;

  // it's possible to make this better but we'll just use the same algorithm as style for now
  const stretchSimilarity = left.cssStretch === right.cssStretch ? 1 : 0.9;
  return weightSimilarity * styleSimilarity * stretchSimilarity;
};

// Build the `selectedFamily` shape the dropdown expects. Uses the lazily
// fetched styles when available, otherwise stubs the current family with just
// the current style so the collapsed selector still shows something meaningful.
function buildSelectedFamily(selectedFamilyId, entry, fontStyle, families) {
  var _fontStyle$family, _families$find, _fontStyle$family2;
  const currentFamilyId = ((_fontStyle$family = fontStyle.family) === null || _fontStyle$family === void 0 ? void 0 : _fontStyle$family.id) ?? '';
  const name = (families === null || families === void 0 ? void 0 : (_families$find = families.find(f => f.id === selectedFamilyId)) === null || _families$find === void 0 ? void 0 : _families$find.name) ?? (selectedFamilyId === currentFamilyId ? (_fontStyle$family2 = fontStyle.family) === null || _fontStyle$family2 === void 0 ? void 0 : _fontStyle$family2.name : undefined) ?? '';
  if (entry !== null && entry !== void 0 && entry.styles) {
    return {
      id: selectedFamilyId,
      name,
      isVariableFont: entry.isVariableFont,
      fontStyles: entry.styles.map(style => ({
        id: style.id,
        name: style.name,
        variableInstances: style.variableInstances
      }))
    };
  }
  if (selectedFamilyId === currentFamilyId) {
    var _fontStyle$variableIn;
    return {
      id: selectedFamilyId,
      name,
      isVariableFont: (((_fontStyle$variableIn = fontStyle.variableInstances) === null || _fontStyle$variableIn === void 0 ? void 0 : _fontStyle$variableIn.length) ?? 0) > 0,
      fontStyles: [{
        id: fontStyle.id,
        name: fontStyle.name,
        variableInstances: fontStyle.variableInstances ?? null
      }]
    };
  }

  // Switched to a family whose styles haven't arrived yet.
  return {
    id: selectedFamilyId,
    name,
    isVariableFont: false,
    fontStyles: null
  };
}
function TypeTesterStyleSelectSelectable(_ref) {
  var _fontStyle$family3;
  let {
    fontStyle,
    variableSettings,
    onSelectFontStyleValue,
    onSelectVariableAxes,
    includeVariableAxesOption,
    fallbackName
  } = _ref;
  const {
    families,
    loadingFamilies,
    loadFamilies,
    getFamilyStyles,
    loadFamilyStyles
  } = (0, _TypeTesterFamilies.useTypeTesterFamilies)();
  const currentFamilyId = ((_fontStyle$family3 = fontStyle.family) === null || _fontStyle$family3 === void 0 ? void 0 : _fontStyle$family3.id) ?? '';

  // The family the user has switched to but whose styles are still loading.
  // We keep the style selector + font on the *current* family until this
  // family's data is ready, then commit family + style + font together — so the
  // toolbar never flashes a blank style selector mid-switch.
  const [pendingFamilyId, setPendingFamilyId] = (0, _react.useState)(null);

  // Tracks the family we've already submitted a style pick for, so the pick
  // effect fires exactly once per switch (it stays mounted until the font
  // commits, after which `fontStyle` would otherwise retrigger it).
  const submittedFamilyRef = (0, _react.useRef)(null);

  // While a switch is in flight the pick sets the live variableSettings ahead
  // of the font committing; freezing to the last settled value keeps the style
  // selector showing the active font's instance until they realign.
  const frozenVariableSettingsRef = (0, _react.useRef)(variableSettings);
  const handleOpen = (0, _react.useCallback)(() => {
    loadFamilies();
    loadFamilyStyles(currentFamilyId);
  }, [loadFamilies, loadFamilyStyles, currentFamilyId]);
  const handleSelectFamilyId = (0, _react.useCallback)(id => {
    if (id === currentFamilyId) {
      // Re-selecting the active family cancels any in-flight switch.
      setPendingFamilyId(null);
      submittedFamilyRef.current = null;
      return;
    }
    loadFamilyStyles(id);
    setPendingFamilyId(id);
  }, [loadFamilyStyles, currentFamilyId]);
  const pendingEntry = pendingFamilyId ? getFamilyStyles(pendingFamilyId) : undefined;

  // A switch is visibly in flight from the moment the user picks a new family
  // until the font actually commits to it (currentFamilyId catches up).
  const switchingToPending = !!pendingFamilyId && currentFamilyId !== pendingFamilyId;

  // Once the picked family's styles arrive, choose a matching style and report
  // it upward. We deliberately keep `pendingFamilyId` set (and the busy cue on)
  // until the font actually switches — clearing it here would briefly revert the
  // family dropdown during the changed-styles round-trip.
  (0, _react.useEffect)(() => {
    if (!pendingFamilyId || !pendingEntry || pendingEntry.loading || !pendingEntry.styles) {
      return;
    }
    if (submittedFamilyRef.current === pendingFamilyId) return;
    submittedFamilyRef.current = pendingFamilyId;
    const styles = pendingEntry.styles;
    let fontStyleId = fontStyle.id;
    let newVariableSettings;
    if (pendingEntry.isVariableFont) {
      const newFontStyle = styles[0];
      if (newFontStyle) {
        var _newFontStyle$variabl;
        fontStyleId = newFontStyle.id;
        const instance = (_newFontStyle$variabl = newFontStyle.variableInstances) === null || _newFontStyle$variabl === void 0 ? void 0 : _newFontStyle$variabl[0];
        if (instance) newVariableSettings = (0, _utils.variableInstanceSettings)(instance);
      }
    } else if (styles.length > 0) {
      fontStyleId = styles.map(right => ({
        id: right.id,
        similarity: computeSimilarity(fontStyle, right)
      })).reduce((closest, current) => current.similarity > closest.similarity ? current : closest).id;
    }
    onSelectFontStyleValue({
      fontStyleId,
      variableSettings: newVariableSettings
    });
  }, [pendingFamilyId, pendingEntry, fontStyle, onSelectFontStyleValue]);

  // The font has actually switched to the picked family — drop the pending
  // state, which lifts the busy cue and settles the family dropdown on it.
  (0, _react.useEffect)(() => {
    if (pendingFamilyId && currentFamilyId === pendingFamilyId) {
      setPendingFamilyId(null);
      submittedFamilyRef.current = null;
    }
  }, [currentFamilyId, pendingFamilyId]);

  // Capture the variableSettings while settled so it can be held frozen for the
  // duration of a switch (see `frozenVariableSettingsRef`).
  (0, _react.useEffect)(() => {
    if (!switchingToPending) frozenVariableSettingsRef.current = variableSettings;
  }, [switchingToPending, variableSettings]);

  // Family options, filtered so they intersect the current style's languages
  // (e.g. an Arabic-only style only offers families that also support Arabic).
  const familyOptions = (0, _react.useMemo)(() => {
    if (!families) {
      return fontStyle.family ? [{
        id: fontStyle.family.id,
        name: fontStyle.family.name
      }] : [];
    }
    return families.filter(family => {
      var _family$featureStyle, _family$featureStyle$;
      return (_family$featureStyle = family.featureStyle) === null || _family$featureStyle === void 0 ? void 0 : (_family$featureStyle$ = _family$featureStyle.supportedLanguages) === null || _family$featureStyle$ === void 0 ? void 0 : _family$featureStyle$.some(lang => {
        var _fontStyle$supportedL;
        return (_fontStyle$supportedL = fontStyle.supportedLanguages) === null || _fontStyle$supportedL === void 0 ? void 0 : _fontStyle$supportedL.includes(lang);
      });
    }).map(family => ({
      id: family.id,
      name: family.name
    }));
  }, [families, fontStyle.family, fontStyle.supportedLanguages]);
  if (!fontStyle.family) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, fallbackName);
  }

  // Once the list has loaded, a zero-language font (no family shares the current
  // style's languages) falls back to the plain name, matching the
  // non-selectable rendering. See FD-628.
  if (families && familyOptions.length === 0) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, fallbackName);
  }

  // Style options always reflect the *active* font's family. When a pending
  // switch commits, `fontStyle` updates and this moves to the new family — so
  // the style selector + font transition together with no blank intermediate.
  const selectedFamily = buildSelectedFamily(currentFamilyId, getFamilyStyles(currentFamilyId), fontStyle, families);

  // Hold the displayed instance on the active font's value during a switch; once
  // it commits, use the live (now-synced) value so axis drags etc. flow through.
  const effectiveVariableSettings = switchingToPending ? frozenVariableSettingsRef.current : variableSettings;
  return /*#__PURE__*/_react.default.createElement(_TypeTesterStyleSelect.default, {
    families: familyOptions,
    selectedFamily: selectedFamily
    // Hold the displayed family on the active font's family until the switch
    // commits, so both selectors + the font flip together. The busy cue is
    // the only immediate feedback while the picked family's styles load.
    ,
    familyValue: currentFamilyId,
    selectedFontStyleId: fontStyle.id,
    selectedVariableSettings: effectiveVariableSettings,
    loadingFamilies: loadingFamilies && !families,
    loading: switchingToPending,
    onOpen: handleOpen,
    onSelectFamilyId: handleSelectFamilyId,
    onSelectFontStyleValue: onSelectFontStyleValue,
    onSelectVariableAxes: onSelectVariableAxes,
    includeVariableAxesOption: includeVariableAxesOption
  });
}
function TypeTesterStyleSelectData(_ref2) {
  var _fontStyle$variableIn2;
  let {
    fontStyle: fontStyleKey,
    onSelectFontStyleValue,
    onSelectVariableAxes,
    config,
    variableSettings,
    includeVariableAxesOption
  } = _ref2;
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTesterStyleSelectData_fontStyle2.default.hash && _TypeTesterStyleSelectData_fontStyle2.default.hash !== "368fd433966e05fc6ebecd2b60c71e3f" && console.error("The definition of 'TypeTesterStyleSelectData_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterStyleSelectData_fontStyle2.default), fontStyleKey);
  const instance = variableSettings ? (_fontStyle$variableIn2 = fontStyle.variableInstances) === null || _fontStyle$variableIn2 === void 0 ? void 0 : _fontStyle$variableIn2.find(inst => (0, _utils.compareVariableSettings)(inst, variableSettings)) : null;
  const name = fontStyle.family ? /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__name"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "type-tester__name__family"
  }, fontStyle.family.name), ' ', /*#__PURE__*/_react.default.createElement("span", {
    className: "type-tester__name__style"
  }, variableSettings ? (instance === null || instance === void 0 ? void 0 : instance.name) ?? 'Custom' : fontStyle.name.replace(/ /g, '\xa0'))) : null;
  if (!config.selectable) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, name);
  }
  return /*#__PURE__*/_react.default.createElement(TypeTesterStyleSelectSelectable, {
    fontStyle: fontStyle,
    variableSettings: variableSettings,
    onSelectFontStyleValue: onSelectFontStyleValue,
    onSelectVariableAxes: onSelectVariableAxes,
    includeVariableAxesOption: includeVariableAxesOption,
    fallbackName: name
  });
}
