// fontdue-js/dist/components/TypeTester/TypeTesterVariableAxes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _TypeTesterVariableAxes_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTesterVariableAxes_fontStyle.graphql"));
var _react = _interopRequireDefault(require("react"));
var _TypeTesterSlider = _interopRequireDefault(require("./TypeTesterSlider"));
var _TypeTesterState = _interopRequireDefault(require("./TypeTesterState"));
var _reactRelay = require("react-relay");
var _utils = require("../../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const TypeTesterVariableAxes = _ref => {
  let {
    id,
    fontStyle: fontStyleKey,
    axes
  } = _ref;
  const {
    variableSettings,
    setVariableSettings
  } = (0, _TypeTesterState.default)({
    id
  });
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTesterVariableAxes_fontStyle2.default.hash && _TypeTesterVariableAxes_fontStyle2.default.hash !== "a73e45dbb7a154107456088125a01f88" && console.error("The definition of 'TypeTesterVariableAxes_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterVariableAxes_fontStyle2.default), fontStyleKey);
  if (!variableSettings) return null;
  if (!axes) return null;
  const handleChange = (axis, value) => {
    setVariableSettings({
      ...variableSettings,
      [axis]: value
    });
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__variable-axes"
  }, axes.filter(showAxis => showAxis in variableSettings) // only show the slider if the value is set
  .map(showAxis => {
    var _fontStyle$variableAx;
    return (_fontStyle$variableAx = fontStyle.variableAxes) === null || _fontStyle$variableAx === void 0 ? void 0 : _fontStyle$variableAx.find(_ref2 => {
      let {
        axis
      } = _ref2;
      return axis === showAxis;
    });
  }).filter(_utils.notEmpty).map(_ref3 => {
    let {
      axis,
      name,
      minValue,
      maxValue
    } = _ref3;
    const isSmall = Math.abs(maxValue - minValue) <= 1;
    const value = variableSettings[axis];
    const format = v => isSmall ? v.toFixed(2) : v.toFixed(0);

    // Reserve enough room for the widest value the axis can show so the
    // number changing (e.g. 99 → 100, or a decimal flipping sign) never
    // reflows the slider beside it. The widest string is always at one of
    // the endpoints — most digits plus any minus sign. With tabular
    // figures (tnum, set in CSS) each character is a stable 1ch wide.
    const valueWidth = Math.max(format(minValue).length, format(maxValue).length);
    return /*#__PURE__*/_react.default.createElement("div", {
      key: axis,
      className: "type-tester__variable-axes__axis"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "type-tester__variable-axes__name"
    }, name), /*#__PURE__*/_react.default.createElement("span", {
      className: "type-tester__variable-axes__value",
      style: {
        width: `${valueWidth}ch`
      }
    }, format(value)), /*#__PURE__*/_react.default.createElement("div", {
      className: "type-tester__variable-axes__slider"
    }, /*#__PURE__*/_react.default.createElement(_TypeTesterSlider.default, {
      value: value,
      min: minValue,
      max: maxValue,
      onChange: value => handleChange(axis, value),
      stepSize: isSmall ? 0.01 : 1,
      interpolator: "linear"
    })));
  }));
};
var _default = TypeTesterVariableAxes;
exports.default = _default;
