// fontdue-js/dist/components/TypeTester/TypeTesterContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _useTypeTesterStyler = _interopRequireDefault(require("./useTypeTesterStyler"));
var _draftJs = require("draft-js");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const TypeTesterContent = /*#__PURE__*/_react.default.forwardRef((_ref, ref) => {
  let {
    content,
    setContent,
    setFocused,
    direction,
    alignment,
    ...props
  } = _ref;
  const editorRef = (0, _react.useRef)(null);
  (0, _react.useImperativeHandle)(ref, () => ({
    focus: () => {
      var _editorRef$current;
      (_editorRef$current = editorRef.current) === null || _editorRef$current === void 0 ? void 0 : _editorRef$current.focus();
    }
  }));
  const style = (0, _useTypeTesterStyler.default)({
    ...props,
    content,
    alignment
  });
  return /*#__PURE__*/_react.default.createElement("div", _extends({}, style, {
    className: "type-tester__text__container"
  }), /*#__PURE__*/_react.default.createElement(_draftJs.Editor, {
    ref: editorRef,
    editorState: content,
    onChange: value => setContent(value),
    onBlur: () => setFocused(false),
    onFocus: () => setFocused(true),
    spellCheck: false,
    stripPastedStyles: true,
    textDirectionality: direction === 'rtl' ? 'RTL' : 'LTR',
    textAlignment: alignment
  }));
});
var _default = TypeTesterContent;
exports.default = _default;
