// fontdue-js/dist/components/TypeTester/TypeTesterFeatures.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TypeTesterFeatures;
var _TypeTesterFeatures_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTesterFeatures_fontStyle.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _TypeTesterBullet = _interopRequireDefault(require("../TypeTester/TypeTesterBullet"));
var _TypeTesterState = _interopRequireDefault(require("./TypeTesterState"));
var _Checkbox = _interopRequireDefault(require("../Checkbox"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _useFeaturesData = _interopRequireDefault(require("./useFeaturesData"));
var _TypeTesterVariableAxes = _interopRequireDefault(require("./TypeTesterVariableAxes"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function useHeightAnimation() {
  let isOpen = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  let duration = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 350;
  const contentRef = (0, _react.useRef)(null);
  const [height, setHeight] = (0, _react.useState)(isOpen ? 'auto' : 0);
  const changed = (0, _react.useRef)(false);
  (0, _react.useEffect)(() => {
    var _contentRef$current;
    const contentRect = (_contentRef$current = contentRef.current) === null || _contentRef$current === void 0 ? void 0 : _contentRef$current.getBoundingClientRect();
    if (!contentRect) return;
    if (changed.current === false) {
      // don't run initially. couldn't think of a better solution
      changed.current = true;
      return;
    }
    setHeight(`${contentRect.height}px`);
    const first = setTimeout(() => setHeight(isOpen ? `${contentRect.height}px` : 0));
    const second = setTimeout(() => {
      if (isOpen) {
        setHeight('auto');
      }
    }, duration);
    return () => {
      clearTimeout(first);
      clearTimeout(second);
    };
  }, [isOpen]);
  return {
    height,
    contentRef
  };
}
const shouldIncludeFeature = (supportedFeatures, feature) => {
  if (!supportedFeatures) return false;
  return supportedFeatures.indexOf(feature) >= 0;
};
function TypeTesterFeatures(_ref) {
  var _config$openTypeFeatu, _config$openTypeFeatu2, _config$openTypeFeatu3, _config$openTypeFeatu4;
  let {
    id,
    fontStyle: fontStyleKey,
    features: showFeatures,
    axes
  } = _ref;
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTesterFeatures_fontStyle2.default.hash && _TypeTesterFeatures_fontStyle2.default.hash !== "d0c693dadaa9ae68f5295f3fc69a7578" && console.error("The definition of 'TypeTesterFeatures_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterFeatures_fontStyle2.default), fontStyleKey);
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const {
    features: selectedFeatures,
    toggleFeature,
    featuresOpen
  } = (0, _TypeTesterState.default)({
    id
  });
  const selectionStyle = config.openTypeFeatures.selectionStyle;
  const {
    featureNames,
    fontFeatures
  } = (0, _useFeaturesData.default)({
    fontStyle
  });
  const renderFeature = feature => {
    const featureCode = typeof feature === 'string' ? feature : feature.code;
    const featureName = typeof feature === 'string' ? featureNames[featureCode] : feature.name;
    const checked = selectedFeatures.indexOf(featureCode) >= 0;
    return /*#__PURE__*/_react.default.createElement("button", {
      key: featureCode,
      className: "type-tester__features__button",
      type: "button",
      onClick: () => toggleFeature(featureCode)
    }, selectionStyle === 'bullet' && /*#__PURE__*/_react.default.createElement(_TypeTesterBullet.default, {
      checked: checked,
      config: config
    }), selectionStyle === 'checkbox' && /*#__PURE__*/_react.default.createElement(_Checkbox.default, {
      checked: checked,
      readOnly: true
    }), /*#__PURE__*/_react.default.createElement("span", {
      className: "type-tester__features__name"
    }, featureName));
  };
  let featuresColumns;
  if ((_config$openTypeFeatu = config.openTypeFeatures) !== null && _config$openTypeFeatu !== void 0 && (_config$openTypeFeatu2 = _config$openTypeFeatu.columns) !== null && _config$openTypeFeatu2 !== void 0 && _config$openTypeFeatu2.length) {
    featuresColumns = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, config.openTypeFeatures.columns.map((columnConfig, i) => /*#__PURE__*/_react.default.createElement("div", {
      key: i,
      "data-range": columnConfig.range,
      className: "type-tester__features__column",
      style: {
        '--grid-range': columnConfig.range
      }
    }, columnConfig.features.filter(feature => shouldIncludeFeature(fontFeatures === null || fontFeatures === void 0 ? void 0 : fontFeatures.supportedFeatures, typeof feature === 'string' ? feature : feature.code)).map(renderFeature))));
  } else {
    const expanded = showFeatures !== null && showFeatures !== void 0 && showFeatures.includes('*') ? (fontFeatures === null || fontFeatures === void 0 ? void 0 : fontFeatures.supportedFeatures) ?? [] : (showFeatures ?? []).filter(feature => shouldIncludeFeature(fontFeatures === null || fontFeatures === void 0 ? void 0 : fontFeatures.supportedFeatures, feature));
    featuresColumns = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, expanded.map(renderFeature));
  }
  const {
    height: featuresHeight,
    contentRef: featuresContentRef
  } = useHeightAnimation(featuresOpen, 300);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__features-container",
    onMouseDown: e => e.preventDefault(),
    "data-open": featuresOpen,
    style: {
      height: featuresHeight,
      overflow: featuresHeight === 'auto' ? 'visible' : 'hidden',
      transition: `height 300ms ease-out`
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: featuresContentRef
  }, axes ? /*#__PURE__*/_react.default.createElement(_TypeTesterVariableAxes.default, {
    id: id,
    axes: axes,
    fontStyle: fontStyle
  }) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__features",
    "data-selection-style": selectionStyle,
    "data-columns": (_config$openTypeFeatu3 = config.openTypeFeatures) !== null && _config$openTypeFeatu3 !== void 0 && (_config$openTypeFeatu4 = _config$openTypeFeatu3.columns) !== null && _config$openTypeFeatu4 !== void 0 && _config$openTypeFeatu4.length ? 'set' : 'auto'
  }, featuresColumns)));
}
