// fontdue-js/dist/components/TypeTester/TypeTesterContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TypeTesterContext = void 0;
var _draftJs = require("draft-js");
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const TypeTesterContext = /*#__PURE__*/(0, _react.createContext)({});
exports.TypeTesterContext = TypeTesterContext;
const updateTester = (state, id, newState, propagate) => {
  const testers = state.testers;
  if (state.mode === 'local' || propagate === false) {
    return {
      ...state,
      testers: {
        ...testers,
        [id]: {
          ...testers[id],
          ...newState
        }
      }
    };
  }
  if (state.mode === 'group') {
    // affect all testers
    const newTesters = Object.keys(testers).reduce((acc, tid) => {
      let content = testers[tid].content;
      if (newState.content) {
        if (id === tid) {
          // for the currently edited tester, we need to use the new content
          // as-is to maintain the selection state
          content = newState.content;
        } else {
          var _testers$id$content;
          if (((_testers$id$content = testers[id].content) === null || _testers$id$content === void 0 ? void 0 : _testers$id$content.getCurrentContent()) !== newState.content.getCurrentContent()) {
            // but we can't use the `content` state as-is for other testers, because
            // the selection state seems to be globally tied to the browser selection,
            // so the cursor jumps to the other tester. we have to make a new
            // EditorState with a copy of the content
            content = _draftJs.EditorState.createWithContent(newState.content.getCurrentContent());
          }
        }
      }
      acc[tid] = {
        ...testers[tid],
        ...newState,
        content
      };
      return acc;
    }, {});
    return {
      ...state,
      testers: newTesters
    };
  }
  return state;
};
const groupReducer = (state, action) => {
  const {
    testers
  } = state;
  switch (action.type) {
    case 'TOGGLE_MODE':
      return {
        ...state,
        mode: state.mode === 'group' ? 'local' : 'group'
      };
    case 'TOGGLE_GLOBAL_FEATURES_OPEN':
      return {
        ...state,
        globalFeaturesOpen: !state.globalFeaturesOpen
      };
    case 'SET_CONTENT':
      return updateTester(state, action.id, {
        content: action.value,
        contentEdited: true
      });
    case 'SET_SIZE':
      return updateTester(state, action.id, {
        size: action.value
      }, action.propagate);
    case 'SET_LETTER_SPACING':
      return updateTester(state, action.id, {
        letterSpacing: action.value
      });
    case 'SET_FEATURES':
      return updateTester(state, action.id, {
        features: action.value
      });
    case 'SET_LINE_HEIGHT':
      return updateTester(state, action.id, {
        lineHeight: action.value
      });
    case 'SET_ALIGNMENT':
      return updateTester(state, action.id, {
        alignment: action.value
      });
    case 'SET_VARIABLE_SETTINGS':
      return updateTester(state, action.id, {
        variableSettings: action.value
      });
    case 'SET_FOCUSED':
      return {
        ...state,
        testers: {
          ...testers,
          [action.id]: {
            ...testers[action.id],
            focused: action.value
          }
        }
      };
    case 'SET_FEATURES_OPEN':
      return {
        ...state,
        testers: {
          ...testers,
          [action.id]: {
            ...testers[action.id],
            featuresOpen: action.value
          }
        }
      };
  }
};
const TypeTesterContextComponent = _ref => {
  let {
    children,
    testers: initialTestersState,
    defaultMode = 'group'
  } = _ref;
  const testers = Object.keys(initialTestersState).reduce((acc, id) => {
    const initialState = initialTestersState[id];
    const content = _draftJs.EditorState.createWithContent(_draftJs.ContentState.createFromText(initialState.content ?? ''));
    acc[id] = {
      content,
      contentEdited: false,
      size: initialState.size ?? 72,
      letterSpacing: initialState.letterSpacing ?? 0,
      lineHeight: initialState.lineHeight ?? 1,
      autofit: initialState.autofit,
      features: initialState.features ?? [],
      featuresOpen: false,
      focused: false,
      alignment: initialState.alignment,
      variableSettings: initialState.variableSettings,
      direction: initialState.direction
    };
    return acc;
  }, {});
  const [value, dispatch] = (0, _react.useReducer)(groupReducer, {
    mode: defaultMode,
    globalFeaturesOpen: false,
    testers
  });
  return /*#__PURE__*/_react.default.createElement(TypeTesterContext.Provider, {
    value: {
      ...value,
      dispatch
    }
  }, children);
};
var _default = TypeTesterContextComponent;
exports.default = _default;
