// fontdue-js/dist/components/TypeTester/TypeTesterFloatingToolbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TypeTesterFloatingToolbar;
var _TypeTesterFloatingToolbar_testers2 = _interopRequireDefault(require("../../__generated__/TypeTesterFloatingToolbar_testers.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _TypeTesterContext = require("./TypeTesterContext");
var _TypeTesterFeatures = _interopRequireDefault(require("./TypeTesterFeatures"));
var _TypeTesterToolbar = _interopRequireDefault(require("./TypeTesterToolbar"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function TypeTesterFloatingToolbar(_ref) {
  var _testersData$find;
  let {
    testers: testersKey,
    features,
    axes,
    onToolbarOpenClose
  } = _ref;
  const testersData = (0, _reactRelay.useFragment)((_TypeTesterFloatingToolbar_testers2.default.hash && _TypeTesterFloatingToolbar_testers2.default.hash !== "5f037fca9e02cba8c202f17e08ff8db6" && console.error("The definition of 'TypeTesterFloatingToolbar_testers' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterFloatingToolbar_testers2.default), testersKey);
  const {
    testers
  } = (0, _react.useContext)(_TypeTesterContext.TypeTesterContext);
  const id = Object.keys(testers).find(id => testers[id].focused);
  const [open, setOpen] = (0, _react.useState)(false);

  // when focus is lost, we lose the latest id, but we don't want to immediately
  // un-render the toolbar for that id because there's a transition to hide the
  // toolbar. so this keeps the tools rendered with the last non-null id
  const [activeId, setActiveId] = (0, _react.useState)(id);
  (0, _react.useEffect)(() => {
    if (id) {
      setActiveId(id);
      setOpen(true); // always open if we get a new id
    }
  }, [id]);
  (0, _react.useEffect)(() => {
    let observer = new IntersectionObserver(entries => {
      const entry = entries[0];
      if (!entry.isIntersecting) {
        var _document$activeEleme;
        setOpen(false);
        (_document$activeEleme = document.activeElement) === null || _document$activeEleme === void 0 ? void 0 : _document$activeEleme.blur();
      }
    });
    const typeTesterElement = document.querySelector(`[data-type-tester-id="${activeId}"]`);
    if (!typeTesterElement) return;
    observer.observe(typeTesterElement);
    return () => {
      observer.disconnect();
    };
  }, [activeId]);
  (0, _react.useEffect)(() => {
    onToolbarOpenClose === null || onToolbarOpenClose === void 0 ? void 0 : onToolbarOpenClose(open);
  }, [onToolbarOpenClose, open]);
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const fontStyle = (_testersData$find = testersData.find(tester => tester.id === activeId)) === null || _testersData$find === void 0 ? void 0 : _testersData$find.fontStyle;
  const axesPosition = config.variableAxesPosition === 'auto' ? ((axes === null || axes === void 0 ? void 0 : axes.length) ?? 0) > 1 ? 'features-panel' : 'inline' : config.variableAxesPosition;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__floating-toolbar",
    onMouseDown: e => e.preventDefault(),
    "data-open": activeId && open
  }, activeId && fontStyle && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_TypeTesterToolbar.default, {
    id: activeId,
    features: features,
    axes: axes,
    axesPosition: axesPosition,
    fontStyle: fontStyle
  }), features && /*#__PURE__*/_react.default.createElement(_TypeTesterFeatures.default, {
    id: activeId,
    features: features,
    axes: axes,
    fontStyle: fontStyle
  })));
}
