// fontdue-js/dist/components/TypeTester/TypeTesterFeaturesButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _TypeTesterFeaturesButton_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTesterFeaturesButton_fontStyle.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _CarrotDown = _interopRequireDefault(require("../Icons/CarrotDown"));
var _CarrotUp = _interopRequireDefault(require("../Icons/CarrotUp"));
var _Select = _interopRequireDefault(require("../Select"));
var _TypeTesterState = _interopRequireDefault(require("./TypeTesterState"));
var _useFeaturesData = _interopRequireDefault(require("./useFeaturesData"));
var _reactRelay = require("react-relay");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const TypeTesterFeaturesButton = _ref => {
  let {
    id,
    fontStyle: fontStyleKey,
    onNeedsRefocus,
    features: showFeatures
  } = _ref;
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const {
    featuresOpen,
    toggleFeaturesOpen,
    features,
    setFeatures
  } = (0, _TypeTesterState.default)({
    id
  });
  const selectedFeature = features[0];
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTesterFeaturesButton_fontStyle2.default.hash && _TypeTesterFeaturesButton_fontStyle2.default.hash !== "bfda99c867db0b614314fc774911c4b0" && console.error("The definition of 'TypeTesterFeaturesButton_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterFeaturesButton_fontStyle2.default), fontStyleKey);
  const {
    featureNames,
    fontFeatures
  } = (0, _useFeaturesData.default)({
    fontStyle
  });
  const effectiveFeatures = showFeatures !== null && showFeatures !== void 0 && showFeatures.includes('*') ? (fontFeatures === null || fontFeatures === void 0 ? void 0 : fontFeatures.supportedFeatures) ?? [] : showFeatures;
  if (config.openTypeFeatures.interactionStyle === 'panel') {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "type-tester__features-button",
      "data-features-open": featuresOpen,
      "data-features-selected": features.length || 0
    }, /*#__PURE__*/_react.default.createElement("button", {
      className: "type-tester__features-button__button",
      type: "button",
      onClick: () => toggleFeaturesOpen()
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "type-tester__features-button__text"
    }, config.openTypeFeatures.buttonLabel), ' ', featuresOpen ? /*#__PURE__*/_react.default.createElement(_CarrotUp.default, null) : /*#__PURE__*/_react.default.createElement(_CarrotDown.default, null)));
  }
  if (config.openTypeFeatures.interactionStyle === 'select') {
    if (!effectiveFeatures || effectiveFeatures.length === 0) return null;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "type-tester__features-select"
    }, /*#__PURE__*/_react.default.createElement(_Select.default, {
      value: selectedFeature ?? '',
      options: [{
        value: '',
        text: 'OT Features'
      }].concat(effectiveFeatures.map(feature => ({
        value: feature,
        text: featureNames[feature]
      }))),
      onChange: e => {
        setFeatures(e.target.value === '' ? [] : [e.target.value]);
        onNeedsRefocus === null || onNeedsRefocus === void 0 ? void 0 : onNeedsRefocus();
      },
      onClick: e => e.stopPropagation(),
      onMouseDown: e => e.stopPropagation()
    }));
  }
  return null;
};
var _default = TypeTesterFeaturesButton;
exports.default = _default;
