// fontdue-js/dist/components/TypeTester/TypeTesterState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _TypeTesterContext = require("./TypeTesterContext");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const useTypeTesterState = _ref => {
  let {
    id
  } = _ref;
  const config = (0, _react.useContext)(_ConfigContext.default);
  const {
    testers,
    globalFeaturesOpen,
    dispatch
  } = (0, _react.useContext)(_TypeTesterContext.TypeTesterContext);
  const toolsPosition = config.typeTester.toolsPosition;
  const {
    content,
    contentEdited,
    size,
    features,
    focused,
    featuresOpen,
    autofit,
    lineHeight,
    letterSpacing,
    alignment,
    variableSettings,
    direction
  } = testers[id];
  return {
    autofit,
    content,
    setContent: value => {
      dispatch({
        type: 'SET_CONTENT',
        id,
        value
      });
    },
    contentEdited,
    size,
    setSize: (value, options) => {
      dispatch({
        type: 'SET_SIZE',
        value,
        id,
        propagate: (options === null || options === void 0 ? void 0 : options.auto) === false
      });
    },
    lineHeight,
    setLineHeight: value => dispatch({
      type: 'SET_LINE_HEIGHT',
      value,
      id
    }),
    letterSpacing,
    setLetterSpacing: value => dispatch({
      type: 'SET_LETTER_SPACING',
      value,
      id
    }),
    alignment,
    setAlignment: value => dispatch({
      type: 'SET_ALIGNMENT',
      value,
      id
    }),
    focused,
    setFocused: value => {
      dispatch({
        type: 'SET_FOCUSED',
        value,
        id
      });
    },
    features,
    setFeatures: value => {
      dispatch({
        type: 'SET_FEATURES',
        value,
        id
      });
    },
    toggleFeature: feature => {
      let value;
      if (features.indexOf(feature) === -1) {
        value = features.concat(feature);
      } else {
        value = features.filter(f => f !== feature);
      }
      dispatch({
        type: 'SET_FEATURES',
        value,
        id
      });
    },
    featuresOpen: toolsPosition === 'floating' ? globalFeaturesOpen : featuresOpen,
    toggleFeaturesOpen: value => {
      if (toolsPosition === 'floating') {
        dispatch({
          type: 'TOGGLE_GLOBAL_FEATURES_OPEN'
        });
      } else {
        dispatch({
          type: 'SET_FEATURES_OPEN',
          value: value ?? !featuresOpen,
          id
        });
      }
    },
    setFeaturesOpen: value => {
      dispatch({
        type: 'SET_FEATURES_OPEN',
        value,
        id
      });
    },
    variableSettings,
    setVariableSettings: value => dispatch({
      type: 'SET_VARIABLE_SETTINGS',
      value,
      id
    }),
    direction
  };
};
var _default = useTypeTesterState;
exports.default = _default;
