// fontdue-js/dist/components/TypeTester/TypeTesterToolbar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _TypeTesterToolbar_fontStyle2 = _interopRequireDefault(require("../../__generated__/TypeTesterToolbar_fontStyle.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _TypeTesterFeaturesButton = _interopRequireDefault(require("./TypeTesterFeaturesButton"));
var _TypeTesterEditAll = _interopRequireDefault(require("./TypeTesterEditAll"));
var _TypeTesterSlider = _interopRequireDefault(require("./TypeTesterSlider"));
var _TypeTesterInput = _interopRequireDefault(require("./TypeTesterInput"));
var _TypeTesterState = _interopRequireDefault(require("./TypeTesterState"));
var _TypeTesterAlignButtons = _interopRequireDefault(require("./TypeTesterAlignButtons"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _TypeTesterVariableAxes = _interopRequireDefault(require("./TypeTesterVariableAxes"));
var _reactRelay = require("react-relay");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const TypeTesterToolbar = _ref => {
  let {
    id,
    features,
    axes,
    axesPosition,
    fontStyle: fontStyleKey,
    onNeedsRefocus,
    groupEdit
  } = _ref;
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const props = (0, _TypeTesterState.default)({
    id
  });
  const fontStyle = (0, _reactRelay.useFragment)((_TypeTesterToolbar_fontStyle2.default.hash && _TypeTesterToolbar_fontStyle2.default.hash !== "1a0f844cbe51f96ed0145d8009f8bff0" && console.error("The definition of 'TypeTesterToolbar_fontStyle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesterToolbar_fontStyle2.default), fontStyleKey);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__toolbar__tools",
    onMouseDown: e => {
      if (e.target instanceof HTMLInputElement) return;
      e.preventDefault();
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__slider"
  }, config.size.label && /*#__PURE__*/_react.default.createElement("span", {
    className: "type-tester__slider__size-label"
  }, config.size.label), config.textInput && /*#__PURE__*/_react.default.createElement(_TypeTesterInput.default, {
    min: config.size.min,
    max: config.size.max,
    value: props.size,
    onChange: props.setSize,
    setFocused: props.setFocused
  }), /*#__PURE__*/_react.default.createElement(_TypeTesterSlider.default, {
    min: config.size.min,
    max: config.size.max,
    value: props.size,
    onChange: props.setSize,
    stepSize: 1,
    interpolator: "log10"
  })), config.letterSpacing && /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__slider__letter-spacing"
  }, config.letterSpacing.label && /*#__PURE__*/_react.default.createElement("span", {
    className: "type-tester__slider__line-height-label"
  }, config.letterSpacing.label), /*#__PURE__*/_react.default.createElement(_TypeTesterSlider.default, {
    min: config.letterSpacing.min,
    max: config.letterSpacing.max,
    value: props.letterSpacing,
    onChange: props.setLetterSpacing,
    stepSize: 0.01,
    interpolator: "linear"
  })), config.lineHeight && /*#__PURE__*/_react.default.createElement("div", {
    className: "type-tester__slider__line-height"
  }, config.lineHeight.label && /*#__PURE__*/_react.default.createElement("span", {
    className: "type-tester__slider__line-height-label"
  }, config.lineHeight.label), /*#__PURE__*/_react.default.createElement(_TypeTesterSlider.default, {
    min: config.lineHeight.min,
    max: config.lineHeight.max,
    value: props.lineHeight,
    onChange: props.setLineHeight,
    stepSize: 0.01,
    interpolator: "log10"
  })), axesPosition === 'inline' && /*#__PURE__*/_react.default.createElement(_TypeTesterVariableAxes.default, {
    id: id,
    fontStyle: fontStyle,
    axes: axes
  }), (Boolean(features === null || features === void 0 ? void 0 : features.length) || Boolean(axes === null || axes === void 0 ? void 0 : axes.length) && axesPosition === 'features-panel') && /*#__PURE__*/_react.default.createElement(_TypeTesterFeaturesButton.default, {
    id: id,
    fontStyle: fontStyle,
    onNeedsRefocus: onNeedsRefocus,
    features: features
  }), config.alignmentButtons && /*#__PURE__*/_react.default.createElement(_TypeTesterAlignButtons.default, props), (groupEdit ?? config.groupEdit) && /*#__PURE__*/_react.default.createElement(_TypeTesterEditAll.default, props));
};
var _default = TypeTesterToolbar;
exports.default = _default;
