// fontdue-js/dist/components/TypeTester/useTypeTesterStyler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _useAutofit = _interopRequireDefault(require("../../hooks/useAutofit"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const useTypeTesterStyler = _ref => {
  let {
    size,
    autofit,
    autofitOnChange,
    features,
    setSize,
    min,
    max,
    truncate,
    focused,
    lineHeight,
    letterSpacing,
    fontStyles,
    verticalMetrics,
    content,
    contentEdited,
    alignment,
    variableSettings,
    columns
  } = _ref;
  const fontVariationSettings = variableSettings ? Object.keys(variableSettings).map(axis => `'${axis}' ${variableSettings[axis]}`).join(',') : undefined;
  const fontFeatureSettings = features.map(feature => `"${feature}"`).join(', ');
  const sideBuffer = autofit ? 10 : 0;
  const text = content.getCurrentContent().getPlainText();
  const {
    ref,
    fontSize: autofitSize
  } = (0, _useAutofit.default)({
    text,
    fontFamily: fontStyles.fontFamily,
    fontSize: size,
    min,
    max,
    letterSpacing,
    fontWeight: fontStyles.fontWeight,
    fontStyle: fontStyles.fontStyle,
    fontVariationSettings,
    enabled: autofit && (!contentEdited || autofitOnChange),
    padding: sideBuffer
  });

  // When autofit calculates a new size, push it into the shared state
  // before the browser paints to avoid a flash of wrapped text.
  (0, _react.useLayoutEffect)(() => {
    if (autofit && autofitSize !== size) {
      setSize(autofitSize, {
        auto: true
      });
    }
  }, [autofitSize]);

  // `truncate` is `true` (1 line), a positive number (that many lines), or
  // falsy (off). Editing always shows the full paragraph, so only truncate
  // when the tester is not focused.
  const truncateLines = truncate === true ? 1 : typeof truncate === 'number' ? Math.max(0, Math.floor(truncate)) : 0;
  const shouldTruncate = truncateLines > 0 && !focused;

  // How far a line's ink spills past its line-box, top and bottom, as a fraction
  // of the font-size (em). `fontLoader` feeds these exact metrics into the
  // @font-face as ascent/descent-override, so this matches the browser's line
  // geometry precisely. It is 0 once the line-height is at least the font's
  // natural content height ((ascender − descender) / unitsPerEm) — i.e. for any
  // normal reading line-height — and only grows for tight, overlapping setting.
  // Kept in em (not px) so it scales with the *rendered* size, including the
  // `--type-tester--adjustment` that shrinks the preview on mobile.
  const lineOverflowEm = shouldTruncate && verticalMetrics && verticalMetrics.unitsPerEm > 0 ? Math.max(0, ((verticalMetrics.ascender - verticalMetrics.descender) / verticalMetrics.unitsPerEm - lineHeight) / 2) : 0;

  // Columns are what make truncation overflow flow sideways — into clipped,
  // off-screen columns — instead of stacking below the last visible line. That
  // is what lets the metric padding reveal descenders/ascenders cleanly, and it
  // means a multi-column tester (e.g. a 3-column specimen) truncates to
  // `truncateLines` lines *per column*. When truncating with columns disabled we
  // still establish a single column so the clipped overflow behaves the same way.
  const columnStyle = columns ? {
    columnCount: columns.count,
    columnWidth: columns.width,
    columnGap: columns.gap
  } : undefined;

  // A glyph's side bearings — ink that sits outside its advance width, like a
  // leading lowercase 'j' or an italic overhang — would be shorn by the inline
  // clip. Per-glyph bearings aren't in the font metrics, so reserve a flat 0.1em
  // on each inline edge and cancel it with a negative margin so
  // wrapping and text position stay put.
  const sideBearing = shouldTruncate ? '0.1em' : '0px';
  const style = {
    fontSize: autofit ? `${size}px` : `calc(var(--type-tester--adjustment, 1) * ${size}px)`,
    lineHeight: `${lineHeight}`,
    ...(shouldTruncate ? {
      // Cap each column at `truncateLines` lines. `max-height` (not `height`)
      // lets shorter paragraphs collapse to their content instead of padding
      // out to N lines of whitespace. Columns send the clipped overflow
      // sideways (into off-screen columns) rather than below, so the last
      // visible line of every column is never partially covered by the next
      // one. That lets the symmetric `lineOverflowEm` padding fully reveal the
      // first line's ascenders/accents and the last line's descenders at any
      // line-height; the negative margins cancel that padding in layout so a
      // collapsed paragraph still occupies exactly its line count.
      //
      // Everything is in `em` so it tracks the rendered font-size (including
      // the mobile `--type-tester--adjustment`); the `+ 2px` is slack that
      // keeps sub-pixel rounding from dropping the Nth line (the extra never
      // reveals line N+1 because that line is clipped out to the side).
      boxSizing: 'border-box',
      maxHeight: `calc(${truncateLines * lineHeight + 2 * lineOverflowEm}em + 2px)`,
      paddingTop: `${lineOverflowEm}em`,
      paddingBottom: `${lineOverflowEm}em`,
      paddingLeft: sideBearing,
      paddingRight: sideBearing,
      marginTop: `${-lineOverflowEm}em`,
      marginBottom: `${-lineOverflowEm}em`,
      overflow: 'hidden',
      // Fill each column to its full N lines before spilling to the next, and
      // allow a final column with a single line — otherwise the default
      // `widows: 2` / `column-fill: balance` rebalance a 4-line paragraph into
      // two columns of 2, showing N−1 lines instead of N.
      columnFill: 'auto',
      widows: 1,
      orphans: 1,
      ...(columnStyle ?? {
        columnCount: 1
      })
    } : {
      height: 'auto',
      overflow: 'visible',
      ...columnStyle
    }),
    fontFeatureSettings,
    marginRight: `calc(${alignment !== 'right' ? -sideBuffer : 0}px - ${sideBearing})`,
    marginLeft: `calc(${alignment === 'right' ? -sideBuffer : 0}px - ${sideBearing})`,
    letterSpacing: `${letterSpacing}em`,
    fontVariationSettings,
    textAlign: alignment
  };
  return {
    ref,
    style
  };
};
var _default = useTypeTesterStyler;
exports.default = _default;
