// fontdue-js/dist/components/ConsentBanner/consent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.activateScripts = activateScripts;
exports.getClientAnonymousId = getClientAnonymousId;
exports.hasConsent = hasConsent;
exports.onConsent = onConsent;
exports.setClientAnonymousId = setClientAnonymousId;
exports.setConsent = setConsent;
const CONSENT_COOKIE = '_fontdue_cc';
const CONSENT_EVENT = 'fontdue:consent';
const MAX_AGE = 60 * 60 * 24 * 365; // 1 year

/** Store the client-generated anonymous_id so it can be sent with consent events. */
let clientAnonymousId;
function setClientAnonymousId(id) {
  clientAnonymousId = id;
}
function getClientAnonymousId() {
  return clientAnonymousId;
}

/** Parse the _fontdue_cc cookie and check if it contains the given category. */
function hasConsent(category) {
  const match = document.cookie.match(new RegExp('(?:^|;\\s*)' + CONSENT_COOKIE + '=([^;]*)'));
  if (!match) return false;
  return match[1].split(',').includes(category);
}

/** Set the _fontdue_cc cookie with the given categories. */
function setConsent(categories) {
  document.cookie = CONSENT_COOKIE + '=' + categories.join(',') + ';path=/;max-age=' + MAX_AGE + ';SameSite=Lax;Secure';
  window.dispatchEvent(new CustomEvent(CONSENT_EVENT, {
    detail: {
      categories
    }
  }));
}

/**
 * Call callback when the given consent category is granted.
 * If already granted, calls immediately. Otherwise polls (500ms) and
 * listens for the fontdue:consent event. Returns a cleanup function.
 */
function onConsent(category, callback) {
  if (hasConsent(category)) {
    callback();
    return () => {};
  }
  let called = false;
  const check = () => {
    if (!called && hasConsent(category)) {
      called = true;
      callback();
      cleanup();
    }
  };
  const interval = setInterval(check, 500);
  window.addEventListener(CONSENT_EVENT, check);
  const cleanup = () => {
    clearInterval(interval);
    window.removeEventListener(CONSENT_EVENT, check);
  };
  return cleanup;
}

/**
 * Activate blocked scripts for the given consent category.
 * Finds all <script type="text/plain" data-consent-category="<category>">
 * elements and replaces them with executable copies.
 */
function activateScripts(category) {
  const scripts = document.querySelectorAll(`script[type="text/plain"][data-consent-category="${category}"]`);
  scripts.forEach(blocked => {
    var _blocked$parentNode;
    const script = document.createElement('script');

    // Copy all attributes except type and data-consent-category
    for (const attr of Array.from(blocked.attributes)) {
      if (attr.name !== 'type' && attr.name !== 'data-consent-category') {
        script.setAttribute(attr.name, attr.value);
      }
    }

    // Copy inline content
    if (blocked.textContent) {
      script.textContent = blocked.textContent;
    }
    (_blocked$parentNode = blocked.parentNode) === null || _blocked$parentNode === void 0 ? void 0 : _blocked$parentNode.replaceChild(script, blocked);
  });
}
