// fontdue-js/dist/components/ConsentBanner/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _UrlContext = require("../UrlContext");
var _consent = require("./consent");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const DEFAULT_MESSAGE = 'We use cookies to analyze site usage and improve your\u00A0experience.';
const ConsentBanner = () => {
  const config = (0, _react.useContext)(_ConfigContext.default);
  const url = (0, _UrlContext.useFontdueUrl)();

  // Start dismissed to avoid hydration mismatch (server has no document.cookie).
  // Check on the client in useEffect.
  const [dismissed, setDismissed] = (0, _react.useState)(true);
  (0, _react.useEffect)(() => {
    if (!config.tracking.consentRequired) return;
    if (!document.cookie.includes('_fontdue_cc=')) {
      setDismissed(false);
    }
  }, [config.tracking.consentRequired]);
  if (dismissed) return null;
  const message = config.tracking.consentMessage || DEFAULT_MESSAGE;
  const sendConsentEvent = categories => {
    // Fire a tracking event so the server sets the _fontdue_state cookie
    // on this response (rather than waiting for the next page load).
    // Include the client anonymous_id so it gets synced to the cookie.
    const body = {
      event: 'Consent Granted',
      properties: {
        categories: categories.join(',')
      }
    };
    const anonId = (0, _consent.getClientAnonymousId)();
    if (anonId) body.anonymous_id = anonId;
    fetch(`${url}/api/track`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      credentials: 'include',
      body: JSON.stringify(body)
    }).catch(() => {});
  };
  const handleAcceptAll = () => {
    (0, _consent.setConsent)(['necessary', 'analytics']);
    (0, _consent.activateScripts)('analytics');
    if (config.tracking.enabled) {
      sendConsentEvent(['necessary', 'analytics']);
    }
    setDismissed(true);
  };
  const handleReject = () => {
    (0, _consent.setConsent)(['necessary']);
    setDismissed(true);
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "fontdue-consent-banner",
    role: "region",
    "aria-label": "Cookie consent"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "fontdue-consent-banner__container"
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: "fontdue-consent-banner__message"
  }, message), /*#__PURE__*/_react.default.createElement("div", {
    className: "fontdue-consent-banner__actions"
  }, /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "fontdue-consent-banner__button fontdue-consent-banner__button--secondary",
    onClick: handleReject
  }, "Necessary only"), /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "fontdue-consent-banner__button fontdue-consent-banner__button--primary",
    onClick: handleAcceptAll
  }, "Accept all"))));
};
var _default = ConsentBanner;
exports.default = _default;
