// fontdue-js/dist/corsError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.handlePossibleCorsError = handlePossibleCorsError;
exports.setCorsModalEnabled = setCorsModalEnabled;
var _retryImport = _interopRequireDefault(require("./retryImport"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
let detected = false;
let navigating = false;
let modalEnabled = true;
function setCorsModalEnabled(enabled) {
  modalEnabled = enabled;
}
if (typeof window !== 'undefined') {
  window.addEventListener('beforeunload', () => {
    navigating = true;
  });
}
function isCrossOrigin(fetchUrl) {
  try {
    const url = new URL(fetchUrl, window.location.href);
    return url.origin !== window.location.origin;
  } catch {
    return false;
  }
}
function startPolling(fetchUrl) {
  const interval = setInterval(async () => {
    try {
      await fetch(fetchUrl, {
        method: 'POST',
        credentials: 'include'
      });
      clearInterval(interval);
      window.location.reload();
    } catch {
      // still blocked, keep polling
    }
  }, 3000);
}
async function isCorsBlocked(fetchUrl) {
  // A no-cors fetch bypasses CORS entirely — if it also fails,
  // the server is unreachable (network error), not CORS-blocked.
  try {
    await fetch(fetchUrl, {
      method: 'HEAD',
      mode: 'no-cors'
    });
    return true; // server reachable, so the original error was CORS
  } catch {
    return false; // server unreachable, network error
  }
}
function showCorsError(origin, fetchUrl) {
  console.error(`[Fontdue] Cross-origin request to ${fetchUrl} was blocked.\n\n` + `Your website (${origin}) is not listed as an allowed origin ` + `in your Fontdue CORS settings.\n\n` + `To fix this:\n` + `1. Log in to your Fontdue dashboard\n` + `2. Go to Settings \u2192 Security\n` + `3. Add "${origin}" to the "Cross-origin API access" field\n` + `4. Save \u2014 this page will reload automatically`);
  if (modalEnabled) {
    (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./components/CorsErrorModal')))).then(_ref => {
      let {
        renderCorsErrorModal
      } = _ref;
      renderCorsErrorModal(origin, fetchUrl);
    }).catch(() => {
      // Chunk failed to load — the console.error above already
      // told the developer what's wrong.
    });
  }
  startPolling(fetchUrl);
}
function handlePossibleCorsError(error, fetchUrl) {
  if (typeof window === 'undefined') return false;
  if (!(error instanceof TypeError)) return false;
  if (navigating) return false;
  if (!isCrossOrigin(fetchUrl)) return false;
  if (detected) return true;
  detected = true;
  const origin = window.location.origin;

  // Verify this is actually a CORS error and not a network failure.
  // We do this async — handlePossibleCorsError still returns true
  // synchronously so the caller can return a stub response instead
  // of throwing, but we only show the modal if CORS is confirmed.
  isCorsBlocked(fetchUrl).then(blocked => {
    if (blocked) {
      showCorsError(origin, fetchUrl);
    } else {
      // Network error, not CORS — reset so a real CORS error
      // can still be detected if connectivity recovers.
      detected = false;
      console.warn(`[Fontdue] Request to ${fetchUrl} failed — this looks like a network issue, not a CORS configuration problem.`);
    }
  });
  return true;
}
