// fontdue-js/dist/fontLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isFontLoaded = isFontLoaded;
exports.loadFont = loadFont;
exports.selectFontSource = selectFontSource;
const cache = new Map();
const loaded = new Set();
function isFontLoaded(fontFamily) {
  return loaded.has(fontFamily);
}
async function loadFont(fontFamily, sources, verticalMetrics) {
  // Return cached promise if loading is in progress or complete
  const existing = cache.get(fontFamily);
  if (existing) return existing;
  const promise = doLoadFont(fontFamily, sources, verticalMetrics);
  cache.set(fontFamily, promise);
  promise.catch(() => {
    cache.delete(fontFamily);
    loaded.delete(fontFamily);
  });
  return promise;
}

/**
 * Pick the best `FontSource` for the current browser from a list emitted by
 * the backend's `webfontSources` field.
 *
 * The backend already returns sources in CSS `src:` priority order
 * (dedicated COLRv1 → COLRv0 → SVG → sbix → bundled → outline-fallback,
 * within format). The job here is to drop entries whose color tech the
 * browser can't render — without this, fetching e.g. an SVG-in-OpenType
 * file in Chrome pins a face Chrome can only render as outline.
 *
 * Sources with no `tech` tag are kept unconditionally (outline / universal).
 * If every tech-tagged source is filtered out, falls back to the unfiltered
 * list — the browser will at least try to render something rather than the
 * fallback font.
 *
 * `techSupportCheck` is injectable for tests; defaults to
 * `CSS.supports("font-tech(color-<tech>)")`.
 */
function selectFontSource(sources) {
  let techSupportCheck = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultTechSupport;
  const valid = sources.filter(s => s != null && s.url != null);
  const supported = valid.filter(s => !s.tech || techSupportCheck(s.tech));
  const candidates = supported.length > 0 ? supported : valid;
  return candidates.find(s => s.format === 'woff2') ?? candidates[0] ?? null;
}
function defaultTechSupport(tech) {
  // SSR or very old browsers without CSS.supports: be optimistic, don't drop
  // tech-tagged sources. Filtering exists to *prevent* loading a file the
  // browser can't render, not to forbid sources we're unsure about.
  if (typeof CSS === 'undefined' || typeof CSS.supports !== 'function') {
    return true;
  }
  return CSS.supports(`font-tech(color-${tech})`);
}
async function doLoadFont(fontFamily, sources, verticalMetrics) {
  const source = selectFontSource(sources);
  if (!source || !source.url) throw new Error(`No font sources for ${fontFamily}`);
  const response = await fetch(source.url);
  if (!response.ok) throw new Error(`Failed to fetch font: ${response.status}`);
  const buffer = await response.arrayBuffer();
  const descriptors = {
    weight: '400',
    style: 'normal',
    display: 'block'
  };
  if (verticalMetrics && verticalMetrics.unitsPerEm > 0) {
    const {
      unitsPerEm,
      ascender,
      descender,
      lineGap
    } = verticalMetrics;
    descriptors.ascentOverride = `${ascender / unitsPerEm * 100}%`;
    descriptors.descentOverride = `${descender / unitsPerEm * -100}%`;
    if (lineGap != null) {
      descriptors.lineGapOverride = `${lineGap / unitsPerEm * 100}%`;
    }
  }
  const fontFace = new FontFace(fontFamily, buffer, descriptors);
  await fontFace.load();
  document.fonts.add(fontFace);
  loaded.add(fontFamily);
  return fontFace;
}
