// fontdue-js/dist/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collectionSkuIdWithDiscount = void 0;
Object.defineProperty(exports, "collectionSkuIdsDifferences", {
  enumerable: true,
  get: function () {
    return _utils.collectionSkuIdsDifferences;
  }
});
exports.createDefaultStore = createDefaultStore;
exports.default = void 0;
Object.defineProperty(exports, "isSelected", {
  enumerable: true,
  get: function () {
    return _utils.isSelected;
  }
});
var _productState = require("./components/Root/productState");
var _redux = require("redux");
var _reactRedux = require("react-redux");
var _utils = require("./utils");
var _ConfigContext = require("./components/ConfigContext");
const unselectLicense = (state, action) => {
  if (!(action.type === 'PRECART_UNSELECT_LICENSE' || action.type === 'PRECART_TOGGLE_LICENSE')) return state;
  return {
    ...state,
    licenseOptions: state.licenseOptions.filter(option => option.licenseId !== action.licenseId)
  };
};
const selectLicense = (state, action) => {
  if (!(action.type === 'PRECART_SELECT_LICENSE' || action.type === 'PRECART_TOGGLE_LICENSE')) return state;
  const existingLicenseOption = state.licenseOptions.findIndex(option => option.licenseId === action.licenseOptionSpec.licenseId);
  if (existingLicenseOption >= 0) return state;
  const excludedIds = action.excludedLicenseIds ?? [];
  const filtered = excludedIds.length > 0 ? state.licenseOptions.filter(option => !excludedIds.includes(option.licenseId)) : state.licenseOptions;
  return {
    ...state,
    licenseOptions: filtered.concat(action.licenseOptionSpec)
  };
};
const toggleLicense = (state, action) => {
  if (action.type !== 'PRECART_TOGGLE_LICENSE') return state;
  const {
    licenseOptions
  } = state;
  const existingLicenseOption = licenseOptions.findIndex(option => option.licenseId === action.licenseOptionSpec.licenseId);
  if (existingLicenseOption >= 0 && !action.checked) {
    return unselectLicense(state, action);
  } else if (existingLicenseOption === -1 && action.checked) {
    return selectLicense(state, action);
  }
  return state;
};
const selectLicenseOption = (state, action) => {
  if (action.type !== 'PRECART_SELECT_LICENSE_OPTION') return state;
  const {
    licenseId,
    variableId,
    optionId
  } = action;
  const {
    licenseOptions
  } = state;
  return {
    ...state,
    licenseOptions: licenseOptions.map(option => {
      if (option.licenseId === licenseId) {
        return {
          ...option,
          licenseOptions: {
            ...option.licenseOptions,
            [variableId]: optionId
          }
        };
      }
      return option;
    })
  };
};
const collectionSkuIdWithDiscount = (state, skuId) => {
  const differences = (0, _utils.collectionSkuIdsDifferences)(state, skuId);
  const collectionId = Object.keys(differences).filter(collSkuId => differences[collSkuId] <= state.skuPrices[skuId]).reduce((smallest, key) => {
    if (!smallest || differences[key] < differences[smallest]) return key;
    if (differences[key] === differences[smallest]) {
      // When differences are equal, prefer a collection that is not the
      // SKU being selected — this ensures collection bundles are preferred
      // over selecting the family itself at the same price
      if (smallest === skuId && key !== skuId) return key;
      // Among non-self candidates, prefer the one covering more unique
      // font styles — this ensures the most comprehensive collection wins
      if (smallest !== skuId && key !== skuId) {
        var _state$collectionStyl, _state$collectionStyl2;
        const smallestStyles = new Set((_state$collectionStyl = state.collectionStyleSkus[smallest]) === null || _state$collectionStyl === void 0 ? void 0 : _state$collectionStyl.fontStyleSkuIds).size;
        const keyStyles = new Set((_state$collectionStyl2 = state.collectionStyleSkus[key]) === null || _state$collectionStyl2 === void 0 ? void 0 : _state$collectionStyl2.fontStyleSkuIds).size;
        if (keyStyles > smallestStyles) return key;
      }
    }
    return smallest;
  }, null);
  if (collectionId) return [collectionId, differences[collectionId]];
  return [null, null];
};
exports.collectionSkuIdWithDiscount = collectionSkuIdWithDiscount;
const selectedCollectionSkuIdContainingSkuId = (state, skuId) => Object.keys(state.selectedSkuIds).filter(id => state.selectedSkuIds[id] === true).find(selectedSkuId => (0, _utils.collTransitivelyContainsSkuId)(state.collectionStyleSkus, selectedSkuId, skuId));
const setKeysFalse = (acc, key) => {
  acc[key] = false;
  return acc;
};
const unselectedMembersSkuIds = (state, action) => {
  if (action.type !== 'SELECT_SKU_ID') throw new Error('Wrong action ${action.type}');
  const {
    skuId,
    selected
  } = action;
  if (!selected) return {}; // only do this in case we are selecting something

  // Recursively collect all descendant SKU IDs so that transitive
  // selections (e.g., collection bundle → family → bundle) are all
  // marked as false
  const result = {};
  const collectMembers = collSkuId => {
    const members = state.collectionStyleSkus[collSkuId];
    if (!members) return;
    for (const id of members.fontStyleSkuIds) {
      result[id] = false;
    }
    for (const id of members.childrenSkuIds) {
      result[id] = false;
      collectMembers(id);
    }
  };
  collectMembers(skuId);
  return result;
};
const selectSkuId = (state, action) => {
  if (action.type !== 'SELECT_SKU_ID') return state;
  const {
    skuId,
    selected
  } = action;
  let actualSkuId = skuId;
  if (selected) {
    // check if selecting any collections instead of this sku
    // would reduce the selection price
    const [collectionSkuId, _amount] = collectionSkuIdWithDiscount(state, skuId);
    actualSkuId = collectionSkuId || actualSkuId;
  } else if (state.selectedSkuIds[skuId] === false) {
    // we're clicking "deselect" on an item that's not actually selected,
    // it's a member of a collection which is selected. in this case
    // we'll find the collection and deselect that collection instead
    const parentId = selectedCollectionSkuIdContainingSkuId(state, skuId);
    if (parentId) actualSkuId = parentId;
  }
  return {
    ...state,
    precartOpen: true,
    selectedSkuIds: {
      ...state.selectedSkuIds,
      [actualSkuId]: selected,
      ...unselectedMembersSkuIds(state, {
        ...action,
        skuId: actualSkuId,
        selected
      })
    }
  };
};
const unselectSkus = state => ({
  ...state,
  selectedSkuIds: {}
});
const storeModalNavigate = (state, action) => {
  if (action.type !== 'STORE_MODAL_NAVIGATE') return state;
  if (state.storeModalRoute.name === action.route && (0, _reactRedux.shallowEqual)(state.storeModalRoute.variables, action.variables)) {
    return state;
  }
  return {
    ...state,
    storeModalRoute: {
      name: action.route,
      variables: action.variables
    },
    storeModalHistory: state.storeModalHistory.concat(state.storeModalRoute)
  };
};
const storeModalReplace = (state, action) => {
  if (action.type !== 'STORE_MODAL_REPLACE') return state;
  return {
    ...state,
    storeModalRoute: {
      name: action.route,
      variables: action.variables
    }
  };
};
const storeModalBack = state => {
  const history = state.storeModalHistory;
  if (history.length < 1) return state;
  const previousRoute = history[history.length - 1];
  return {
    ...state,
    storeModalRoute: previousRoute,
    storeModalHistory: history.slice(0, history.length - 1)
  };
};
const updateOrderVariableSelections = (state, action) => {
  if (action.type !== 'UPDATE_ORDER_VARIABLE_SELECTIONS') return state;
  return {
    ...state,
    orderVariableSelections: action.selections
  };
};
const reducer = (state, action) => {
  if (!state) throw 'Reducer called without state';
  switch (action.type) {
    case 'POPULATE_PRODUCT_STATE':
      return (0, _productState.populateProductStateReducer)(state, action);
    case 'PRECART_TOGGLE_LICENSE':
      return toggleLicense(state, action);
    case 'PRECART_UNSELECT_LICENSE':
      return unselectLicense(state, action);
    case 'PRECART_SELECT_LICENSE':
      return selectLicense(state, action);
    case 'PRECART_SELECT_LICENSE_OPTION':
      return selectLicenseOption(state, action);
    case 'SELECT_SKU_ID':
      return selectSkuId(state, action);
    case 'UNSELECT_SKUS':
      return unselectSkus(state);
    case 'OPEN_PRECART':
      return {
        ...state,
        precartOpen: true
      };
    case 'CLOSE_PRECART':
      return {
        ...state,
        precartOpen: false
      };
    case 'OPEN_CART':
      return {
        ...state,
        cartOpen: true
      };
    case 'CLOSE_CART':
      return {
        ...state,
        cartOpen: false
      };
    case 'STORE_MODAL_NAVIGATE':
      return storeModalNavigate(state, action);
    case 'STORE_MODAL_REPLACE':
      return storeModalReplace(state, action);
    case 'STORE_MODAL_BACK':
      return storeModalBack(state);
    case 'UPDATE_ORDER_VARIABLE_SELECTIONS':
      return updateOrderVariableSelections(state, action);
    case 'UPDATE_LICENSEE_IS_BILLING_IDENTITY':
      return {
        ...state,
        licenseeIsBillingIdentity: action.value
      };
    default:
      return state;
  }
};
function createDefaultStore(config) {
  const configValue = (0, _ConfigContext.makeConfig)(config);
  const initialRoute = configValue.storeModal.hideAllCollections ? {
    name: 'cart'
  } : {
    name: 'index',
    variables: {
      excludeTags: configValue.storeModal.indexExcludeTags
    }
  };
  return (0, _redux.createStore)(reducer, {
    selectedSkuIds: {},
    // { [skuId]: Boolean }
    licenseOptions: [],
    cartOpen: false,
    precartOpen: false,
    collectionStyleSkus: {},
    skuPrices: {},
    collectionSkus: {},
    fetchedCollectionIds: [],
    storeModalRoute: initialRoute,
    storeModalHistory: [],
    orderVariableSelections: [],
    licenseeIsBillingIdentity: null
  });
}
var _default = reducer;
exports.default = _default;
