// fontdue-js/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compareVariableSettings = exports.collectionSkuIdsDifferences = exports.collTransitivelyContainsSkuId = exports.collContainsSkuId = void 0;
exports.getFeatureGlyphs = getFeatureGlyphs;
exports.isSelected = exports.inspect = void 0;
exports.kebabToCamel = kebabToCamel;
exports.notEmpty = notEmpty;
exports.variableSettingsCSS = exports.variableInstanceSettings = exports.variableInstanceCSS = exports.pluralize = exports.parseCustomProperties = void 0;
const pluralize = (count, singular, plural) => {
  if (count === 1) return `1 ${singular}`;
  if (!plural) return `${count} ${singular}s`;
  return `${count}\xa0${plural}`;
};
exports.pluralize = pluralize;
function notEmpty(value) {
  return value !== null && value !== undefined;
}
function kebabToCamel(str) {
  return str.replace(/-([a-z])/g, function (g) {
    return g[1].toUpperCase();
  });
}

// where coll is a member of `collectionStyleSkus`;
const collContainsSkuId = (coll, skuIdOrStyleId) => {
  if (!coll) return false;
  return coll.childrenSkuIds.indexOf(skuIdOrStyleId) >= 0 || coll.fontStyleSkuIds.indexOf(skuIdOrStyleId) >= 0 || coll.fontStyleIds.indexOf(skuIdOrStyleId) >= 0;
};

// Transitive containment check — walks the collection hierarchy to find
// if a SKU is contained at any depth (e.g., collection bundle → family → bundle)
exports.collContainsSkuId = collContainsSkuId;
const collTransitivelyContainsSkuId = (collectionStyleSkus, collSkuId, skuIdOrStyleId) => {
  const coll = collectionStyleSkus[collSkuId];
  if (!coll) return false;
  if (collContainsSkuId(coll, skuIdOrStyleId)) return true;
  return coll.childrenSkuIds.some(childSkuId => collTransitivelyContainsSkuId(collectionStyleSkus, childSkuId, skuIdOrStyleId));
};
exports.collTransitivelyContainsSkuId = collTransitivelyContainsSkuId;
const isSelected = skuIdOrStyleId => state => {
  if (!skuIdOrStyleId) return false;
  if (state.selectedSkuIds[skuIdOrStyleId] === true) return true;
  // get the collections that contain this skuId as a fontStyle or child
  const parentCollectionSkuIds = Object.keys(state.collectionStyleSkus).filter(collSkuId => collContainsSkuId(state.collectionStyleSkus[collSkuId], skuIdOrStyleId));
  // Recursively check ancestors — handles transitive selection
  // (e.g., collection bundle → family → bundle)
  return parentCollectionSkuIds.some(parentSkuId => isSelected(parentSkuId)(state));
};

// find any collection ancestors that contain this skuId,
// then for each collection, sum the price of selected skus,
// and return the delta between the collection price and the sum.
exports.isSelected = isSelected;
const collectionSkuIdsDifferences = (state, skuId) => Object.keys(state.collectionStyleSkus).filter(collectionSkuId => {
  return skuId === collectionSkuId || collTransitivelyContainsSkuId(state.collectionStyleSkus, collectionSkuId, skuId);
}).reduce((acc, collectionSkuId) => {
  const selectedSkuIdsInColl = Object.keys(state.selectedSkuIds).filter(selectedSkuId => state.selectedSkuIds[selectedSkuId] === true && collTransitivelyContainsSkuId(state.collectionStyleSkus, collectionSkuId, selectedSkuId));
  const sumOfSelectedSkuIdsInColl = selectedSkuIdsInColl.map(selectedSkuId => state.skuPrices[selectedSkuId] ?? 0).reduce((acc, price) => acc + price, 0);
  acc[collectionSkuId] = (state.skuPrices[collectionSkuId] ?? 0) - sumOfSelectedSkuIdsInColl;
  return acc;
}, {});
exports.collectionSkuIdsDifferences = collectionSkuIdsDifferences;
const variableInstanceSettings = instance => instance === null || instance === void 0 ? void 0 : instance.coordinates.reduce((acc, _ref) => {
  let {
    axis,
    value
  } = _ref;
  acc[axis] = value;
  return acc;
}, {});
exports.variableInstanceSettings = variableInstanceSettings;
const compareVariableSettings = (instance, settings) => instance.coordinates.every(coord => settings[coord.axis] === coord.value);
exports.compareVariableSettings = compareVariableSettings;
const variableSettingsCSS = variableSettings => ({
  fontVariationSettings: Object.keys(variableSettings).map(axis => `'${axis}' ${variableSettings[axis]}`).join(',')
});
exports.variableSettingsCSS = variableSettingsCSS;
const variableInstanceCSS = variableInstance => ({
  fontVariationSettings: variableInstance.coordinates.map(_ref2 => {
    let {
      axis,
      value
    } = _ref2;
    return `'${axis}' ${value}`;
  }).join(',')
});
exports.variableInstanceCSS = variableInstanceCSS;
const parseCustomProperties = props => {
  if (!props) return {};
  return props.reduce((acc, prop) => {
    const [key, val] = prop.split(':');
    acc[key] = val;
    return acc;
  }, {});
};
exports.parseCustomProperties = parseCustomProperties;
const inspect = value => {
  console.log(value);
  return value;
};
exports.inspect = inspect;
function getFeatureGlyphs(supportedLanguages) {
  if (supportedLanguages !== null && supportedLanguages !== void 0 && supportedLanguages.includes('en')) return 'Aa';
  if (supportedLanguages !== null && supportedLanguages !== void 0 && supportedLanguages.includes('ar')) return 'ت';
  if (supportedLanguages !== null && supportedLanguages !== void 0 && supportedLanguages.includes('gr')) return 'Αα';
  if (supportedLanguages !== null && supportedLanguages !== void 0 && supportedLanguages.includes('ru')) return 'Аа';
  return null;
}
