// fontdue-js/dist/components/Root/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactErrorBoundary = require("react-error-boundary");
var _reactDom = _interopRequireDefault(require("react-dom"));
var _uuid = require("uuid");
var _reducer = require("../../reducer");
var _retryImport = _interopRequireDefault(require("../../retryImport"));
var _utils = require("../../utils");
var _FontdueProvider = _interopRequireDefault(require("../FontdueProvider"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const customElementMap = {
  // @ts-ignore
  'fontdue-add-to-cart-banner': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../AddToCartBanner'))))),
  // @ts-ignore
  'fontdue-buying-options': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../BuyingOptions'))))),
  'fontdue-buy-button': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../BuyButton'))))),
  'fontdue-cart': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../Cart'))))),
  'fontdue-cart-button': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../CartButton'))))),
  'fontdue-character-viewer': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../CharacterViewer'))))),
  // @ts-ignore
  'fontdue-collection-aa': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../CollectionAa'))))),
  // @ts-ignore
  'fontdue-cookie-notification': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../CookieNotification'))))),
  'fontdue-font-families': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../FontFamilies'))))),
  'fontdue-customer-login-form': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../CustomerLoginForm'))))),
  'fontdue-newsletter-signup': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../NewsletterSignup/NewsletterSignupElement'))))),
  'fontdue-node-password-form': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../NodePasswordForm'))))),
  // @ts-ignore
  'fontdue-precart': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../Precart'))))),
  // @ts-ignore
  'fontdue-specimen-link': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../SpecimenLink'))))),
  // @ts-ignore
  'fontdue-sticky-nav': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../StickyNav'))))),
  'fontdue-store-modal': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../StoreModal'))))),
  'fontdue-test-fonts-form': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../TestFontsForm/TestFontsFormElement'))))),
  'fontdue-type-tester': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../TypeTester/TypeTesterStandaloneElement'))))),
  'fontdue-type-testers': /*#__PURE__*/(0, _react.lazy)(() => (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../TypeTesters/TypeTestersElement')))))
};
Object.keys(customElementMap).forEach(elementName => {
  // this might be more sophisticated in the future with a shadow DOM, etc,
  // but for now we just want the API to use custom tags
  class FontdueElement extends HTMLElement {}
  customElements.get(elementName) || customElements.define(elementName, FontdueElement);
});
const renderCustomElement = (element, key) => {
  const Component = customElementMap[element.nodeName.toLowerCase()];
  const attributes = element.getAttributeNames();
  const props = attributes.reduce((acc, attribute) => {
    acc[(0, _utils.kebabToCamel)(attribute)] = element.getAttribute(attribute) ?? '';
    return acc;
  }, {});
  return /*#__PURE__*/_reactDom.default.createPortal( /*#__PURE__*/_react.default.createElement(_reactErrorBoundary.ErrorBoundary, {
    fallback: null
  }, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: null
  }, /*#__PURE__*/_react.default.createElement(Component, _extends({}, props, {
    key: key
  })))), element);
};
const customElementTags = Object.keys(customElementMap);
const SINGLETONS = ['fontdue-store-modal'];
const allFontdueElements = element => customElementTags.flatMap(elementName => {
  if (SINGLETONS.includes(elementName)) {
    return [element.querySelector(elementName)];
  } else {
    return Array.from(element.querySelectorAll(elementName));
  }
});
const getInitialElements = () => {
  const map = new Map();
  allFontdueElements(document.body).filter(_utils.notEmpty).forEach(element => map.set(element, (0, _uuid.v4)()));
  return map;
};
const Root = props => {
  // use a Map here instead of an array in case we register the same element
  // more than once (otherwise we run into issues rendering the component
  // twice into the same element). this also gives us a way to set unique keys.
  const [elements, setElements] = (0, _react.useState)(getInitialElements());
  const store = (0, _react.useRef)((0, _reducer.createDefaultStore)(props.config));
  (0, _react.useEffect)(() => {
    // watch for any new or removed fontdue elements

    const mutationObserver = new MutationObserver(mutations => {
      mutations.forEach(mutation => {
        mutation.addedNodes.forEach(node => {
          if (!(node instanceof HTMLElement)) return;
          let newElements = [];
          if (customElementTags.includes(node.nodeName.toLowerCase())) {
            newElements.push(node);
          }
          newElements = newElements.concat(allFontdueElements(node).filter(_utils.notEmpty));
          if (newElements.length > 0) {
            setElements(elements => {
              const newMap = new Map(elements);
              newElements.forEach(el => newMap.set(el, (0, _uuid.v4)()));
              return newMap;
            });
          }
        });
        mutation.removedNodes.forEach(node => {
          if (!(node instanceof HTMLElement)) return;
          let removedElements = [];
          if (customElementTags.includes(node.nodeName.toLowerCase())) {
            removedElements.push(node);
          }
          removedElements = removedElements.concat(allFontdueElements(node).filter(_utils.notEmpty));
          if (removedElements.length > 0) {
            setElements(elements => {
              const newMap = new Map(elements);
              removedElements.forEach(el => newMap.delete(el));
              return newMap;
            });
          }
        });
      });
    });
    mutationObserver.observe(document.body, {
      subtree: true,
      childList: true
    });
    function handleRoute(url) {
      const routes = {
        '^index$': handleIndex,
        '^cart/?$': handleCart,
        '^product/([a-zA-Z0-9+/=]+)/?$': handleProduct
      };
      for (const [path, handler] of Object.entries(routes)) {
        const match = url.match(new RegExp(path));
        if (match) {
          const params = match.slice(1);
          handler(...params);
          return;
        }
      }
    }
    function handleIndex() {
      store.current.dispatch({
        type: 'STORE_MODAL_REPLACE',
        route: 'index'
      });
    }
    function handleCart() {
      if (store.current.getState().storeModalHistory) store.current.dispatch({
        type: 'STORE_MODAL_REPLACE',
        route: 'cart'
      });
    }
    function handleProduct(productId) {
      store.current.dispatch({
        type: 'STORE_MODAL_REPLACE',
        route: 'product',
        variables: {
          productId
        }
      });
    }
    function fontdueClickHandler(target, _event) {
      if (target.getAttribute('fontdue-click') === 'open-store-modal') {
        store.current.dispatch({
          type: 'OPEN_CART'
        });
        const route = target.getAttribute('fontdue-store-route');
        if (route) handleRoute(route);
        const selectSkuId = target.getAttribute('fontdue-store-select');
        if (selectSkuId) {
          store.current.dispatch({
            type: 'SELECT_SKU_ID',
            skuId: selectSkuId,
            selected: true
          });
        }
      }
    }
    function handleClick(event) {
      // search all parents of target for the click attribute
      var target = event.target;
      while (target && target instanceof Element) {
        if (target.hasAttribute('fontdue-click')) {
          fontdueClickHandler(target, event);
          return;
        }
        target = target.parentNode;
      }
    }
    document.body.addEventListener('click', handleClick);
    return () => {
      mutationObserver.disconnect();
      document.body.removeEventListener('click', handleClick);
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_FontdueProvider.default, {
    url: props.url,
    stripeIntegration: props.stripeIntegration,
    config: props.config,
    components: props.components,
    store: store.current
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, Array.from(elements).map(_ref => {
    let [el, key] = _ref;
    return renderCustomElement(el, key);
  })));
};
var _default = Root;
exports.default = _default;
