// fontdue-js/dist/components/Root/productState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.populateProductStateReducer = exports.populateProductState = void 0;
var _productState_Query2 = _interopRequireDefault(require("../../__generated__/productState_Query.graphql"));
var _productState_node2 = _interopRequireDefault(require("../../__generated__/productState_node.graphql"));
var _productState_collectionBundle2 = _interopRequireDefault(require("../../__generated__/productState_collectionBundle.graphql"));
var _productState_bundle2 = _interopRequireDefault(require("../../__generated__/productState_bundle.graphql"));
var _productState_collection2 = _interopRequireDefault(require("../../__generated__/productState_collection.graphql"));
var _productState_sku2 = _interopRequireDefault(require("../../__generated__/productState_sku.graphql"));
var _reactRelay = require("react-relay");
var _utils = require("../../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
_productState_sku2.default.hash && _productState_sku2.default.hash !== "2cee6ddddc85699f0c90e2607c0634ef" && console.error("The definition of 'productState_sku' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_sku2.default;
_productState_collection2.default.hash && _productState_collection2.default.hash !== "90e56933f956c6773c0958284de81884" && console.error("The definition of 'productState_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_collection2.default;
_productState_bundle2.default.hash && _productState_bundle2.default.hash !== "fa38bbc9f2062a727da4f5062b362054" && console.error("The definition of 'productState_bundle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_bundle2.default;
_productState_collectionBundle2.default.hash && _productState_collectionBundle2.default.hash !== "6744aa8f57d14b0111a0c49f488d7b03" && console.error("The definition of 'productState_collectionBundle' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_collectionBundle2.default;
_productState_node2.default.hash && _productState_node2.default.hash !== "a1e172db45ca550a254819d877497d12" && console.error("The definition of 'productState_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_node2.default;
const query = (_productState_Query2.default.hash && _productState_Query2.default.hash !== "ac3292bae9a274139d5a525fd0162711" && console.error("The definition of 'productState_Query' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _productState_Query2.default);
const populateProductState = (store, environment, id) => {
  const existingState = store.getState().fetchedCollectionIds;
  if (existingState && existingState.indexOf(id) >= 0) return;
  return (0, _reactRelay.fetchQuery)(environment, query, {
    id
  }).subscribe({
    next: data => store.dispatch({
      type: 'POPULATE_PRODUCT_STATE',
      data: data
    })
  });
};
exports.populateProductState = populateProductState;
const allSkus = function (collection) {
  var _collection$fontStyle;
  let includeCollections = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  let includeFontStyles = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  let bundleSkus = [];
  if ('bundles' in collection) {
    var _collection$bundles;
    bundleSkus = (collection === null || collection === void 0 ? void 0 : (_collection$bundles = collection.bundles) === null || _collection$bundles === void 0 ? void 0 : _collection$bundles.flatMap(bundle => allSkus(bundle, includeCollections, includeFontStyles))) ?? [];
  }
  let collectionBundleSkus = [];
  if ('collectionBundles' in collection && collection.collectionBundles) {
    collectionBundleSkus = collection.collectionBundles.flatMap(collBundle => allSkus(collBundle, includeCollections, includeFontStyles));
  }

  // For collection bundles, get SKUs from member collections
  let memberCollectionSkus = [];
  if ('memberCollections' in collection && collection.memberCollections) {
    memberCollectionSkus = collection.memberCollections.flatMap(member => {
      var _member$fontStyles;
      return [includeCollections ? member.sku : null, ...(includeFontStyles ? ((_member$fontStyles = member.fontStyles) === null || _member$fontStyles === void 0 ? void 0 : _member$fontStyles.map(style => style.sku)) ?? [] : [])];
    }).filter(_utils.notEmpty);
  }
  let childrenSkus = [];
  if ('children' in collection) {
    var _collection$children;
    childrenSkus = ((_collection$children = collection.children) === null || _collection$children === void 0 ? void 0 : _collection$children.flatMap(child => allSkus(child, includeCollections, includeFontStyles))) ?? [];
  }
  return [includeCollections ? collection === null || collection === void 0 ? void 0 : collection.sku : null, ...(includeFontStyles && 'fontStyles' in collection ? (collection === null || collection === void 0 ? void 0 : (_collection$fontStyle = collection.fontStyles) === null || _collection$fontStyle === void 0 ? void 0 : _collection$fontStyle.map(style => style.sku)) ?? [] : []), ...bundleSkus, ...collectionBundleSkus, ...memberCollectionSkus, ...childrenSkus].filter(_utils.notEmpty);
};
const flattenSkuData = collection => {
  let bundles = {};
  if ('bundles' in collection && collection.bundles) {
    bundles = collection.bundles.reduce((res, bundle) => ({
      ...res,
      ...flattenSkuData(bundle)
    }), {});
  }
  let collectionBundles = {};
  if ('collectionBundles' in collection && collection.collectionBundles) {
    collectionBundles = collection.collectionBundles.reduce((res, collBundle) => ({
      ...res,
      ...flattenSkuData(collBundle)
    }), {});
  }
  let children = {};
  if ('children' in collection && collection.children) {
    children = collection.children.reduce((res, child) => ({
      ...res,
      ...flattenSkuData(child)
    }), {});
  }
  let collectionData = {};
  if (collection.sku) {
    var _collection$fontStyle2, _collection$children2;
    // For collection bundles, get font style IDs from member collections
    const fontStyleIds = 'memberCollections' in collection && collection.memberCollections ? collection.memberCollections.flatMap(member => {
      var _member$fontStyles2;
      return ((_member$fontStyles2 = member.fontStyles) === null || _member$fontStyles2 === void 0 ? void 0 : _member$fontStyles2.map(style => style.id)) ?? [];
    }) : 'fontStyles' in collection ? (((_collection$fontStyle2 = collection.fontStyles) === null || _collection$fontStyle2 === void 0 ? void 0 : _collection$fontStyle2.map(style => style.id)) ?? []).concat('children' in collection ? ((_collection$children2 = collection.children) === null || _collection$children2 === void 0 ? void 0 : _collection$children2.flatMap(child => {
      var _child$fontStyles;
      return ((_child$fontStyles = child.fontStyles) === null || _child$fontStyles === void 0 ? void 0 : _child$fontStyles.map(style => style.id)) ?? [];
    })) ?? [] : []) : [];
    collectionData = {
      [collection.sku.id]: {
        fontStyleIds,
        fontStyleSkuIds: allSkus(collection, false, true).map(sku => sku.id),
        childrenSkuIds: allSkus(collection, true, false).map(sku => sku.id).filter(skuId => {
          var _collection$sku;
          return skuId !== ((_collection$sku = collection.sku) === null || _collection$sku === void 0 ? void 0 : _collection$sku.id);
        }),
        // exclude self
        name: collection.name
      }
    };
  }
  return {
    ...collectionData,
    ...children,
    ...bundles,
    ...collectionBundles
  };
};
const flattenSkuPrices = collection => {
  return allSkus(collection).reduce((acc, sku) => {
    acc[sku.id] = sku.price.amount;
    return acc;
  }, {});
};
const populateProductStateReducer = (state, action) => {
  var _data$node, _data$node$sku;
  if (action.type !== 'POPULATE_PRODUCT_STATE') return state;
  const {
    data
  } = action;
  if (((_data$node = data.node) === null || _data$node === void 0 ? void 0 : _data$node.__typename) !== 'FontCollection') return state;
  return {
    ...state,
    // {
    //   [collectionSkuId]: {
    //     childrenSkuIds: [skuId...],
    //     fontStyleSkuIds: [skuId...],
    //   }
    // }
    collectionStyleSkus: {
      ...state.collectionStyleSkus,
      ...flattenSkuData(data.node)
    },
    // {
    //   [skuId]: amount,
    // }
    skuPrices: {
      ...state.skuPrices,
      ...flattenSkuPrices(data.node)
    },
    // {
    //   [collectionId]: [skuId]
    // }
    collectionSkus: {
      ...state.collectionSkus,
      [`${data.node.id}`]: ((_data$node$sku = data.node.sku) === null || _data$node$sku === void 0 ? void 0 : _data$node$sku.id) ?? null
    },
    fetchedCollectionIds: state.fetchedCollectionIds.concat(data.node.id)
  };
};
exports.populateProductStateReducer = populateProductStateReducer;
