// fontdue-js/dist/components/CustomerLoginForm/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _CustomerLoginFormQuery2 = _interopRequireDefault(require("../../__generated__/CustomerLoginFormQuery.graphql"));
var _CustomerLoginFormLoginMutation2 = _interopRequireDefault(require("../../__generated__/CustomerLoginFormLoginMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _TextField = _interopRequireDefault(require("../TextField"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const loginMutation = (_CustomerLoginFormLoginMutation2.default.hash && _CustomerLoginFormLoginMutation2.default.hash !== "975b639fac1e0d88e0f0c9c55acd8b9c" && console.error("The definition of 'CustomerLoginFormLoginMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CustomerLoginFormLoginMutation2.default);
const query = (_CustomerLoginFormQuery2.default.hash && _CustomerLoginFormQuery2.default.hash !== "f32c7b0b485a2b2a02ddf2cbc5b87a3c" && console.error("The definition of 'CustomerLoginFormQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CustomerLoginFormQuery2.default);
const DEFAULT_SUBMITTED_LABEL = '<p>Submitted!</p><p>Please check your email inbox for a link to log in.</p><p>If you don\u2019t receive an email, please contact us to retrieve your order information.</p>';
const CustomerLoginForm = _ref => {
  var _data$viewer, _data$viewer$settings;
  let {
    submitLabel = 'Submit'
  } = _ref;
  const [email, setEmail] = (0, _react.useState)('');
  const [error, setError] = (0, _react.useState)(null);
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const [submitted, setSubmitted] = (0, _react.useState)(false);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const data = (0, _reactRelay.useLazyLoadQuery)(query, {});
  const submittedLabel = ((_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : (_data$viewer$settings = _data$viewer.settings) === null || _data$viewer$settings === void 0 ? void 0 : _data$viewer$settings.customerLoginSubmittedLabel) || DEFAULT_SUBMITTED_LABEL;
  const handleSubmit = e => {
    e.preventDefault();
    setSubmitting(true);
    (0, _reactRelay.commitMutation)(environment, {
      mutation: loginMutation,
      variables: {
        input: {
          email
        }
      },
      onCompleted: (_res, errors) => {
        if (errors && errors.length > 0) {
          setError(errors[0].message);
          setSubmitting(false);
        } else {
          setSubmitted(true);
          setSubmitting(false);
          setError(null);
        }
      },
      onError: _ref2 => {
        let {
          message
        } = _ref2;
        setError(message);
        setSubmitting(false);
      }
    });
  };
  const disabled = !email;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "login-form"
  }, error && /*#__PURE__*/_react.default.createElement("div", {
    className: "login-form__errors"
  }, error), submitted ? /*#__PURE__*/_react.default.createElement("div", {
    className: "login-form__submitted",
    dangerouslySetInnerHTML: {
      __html: submittedLabel
    }
  }) : /*#__PURE__*/_react.default.createElement("form", {
    className: "login-form__form",
    onSubmit: handleSubmit
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "login-form__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Email *",
    type: "email",
    value: email,
    onChange: value => setEmail(value),
    autocomplete: "email",
    placeholder: "Your email"
  })), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("button", {
    className: "submit-button",
    type: "submit",
    disabled: submitting || disabled
  }, submitting ? 'Submitting...' : submitLabel, /*#__PURE__*/_react.default.createElement("span", {
    className: "submit-button__arrow"
  }, " \u2192")))));
};
var _default = CustomerLoginForm;
exports.default = _default;
