// fontdue-js/dist/components/Tracking/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _UrlContext = require("../UrlContext");
var _consent = require("../ConsentBanner/consent");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function getCampaignDataFromUrl() {
  const urlParams = new URLSearchParams(window.location.search);
  const campaign = {};
  const properties = {};
  let hasCampaign = false;
  let hasProperties = false;
  const utmMapping = {
    utm_source: 'source',
    utm_medium: 'medium',
    utm_campaign: 'name',
    utm_term: 'term',
    utm_content: 'content'
  };
  for (const [urlParam, campaignKey] of Object.entries(utmMapping)) {
    const value = urlParams.get(urlParam);
    if (value) {
      campaign[campaignKey] = value;
      hasCampaign = true;
    }
  }
  const clickIds = ['gclid', 'fbclid'];
  for (const param of clickIds) {
    const value = urlParams.get(param);
    if (value) {
      properties[param] = value;
      hasProperties = true;
    }
  }
  return {
    campaign: hasCampaign ? campaign : null,
    properties: hasProperties ? properties : null
  };
}
async function sendPageView(baseUrl, anonymousId) {
  try {
    const {
      campaign,
      properties: clickIdProperties
    } = getCampaignDataFromUrl();
    const properties = {
      url: window.location.href,
      path: window.location.pathname,
      referrer: document.referrer,
      title: document.title,
      search: window.location.search,
      ...clickIdProperties
    };
    const context = {
      userAgent: navigator.userAgent
    };
    if (campaign) {
      context.campaign = campaign;
    }
    const body = {
      properties,
      context
    };
    if (anonymousId) {
      body.anonymous_id = anonymousId;
    }
    const response = await fetch(`${baseUrl}/api/track/page`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      credentials: 'include',
      body: JSON.stringify(body)
    });
    const data = await response.json();
    return data.anonymous_id || null;
  } catch (error) {
    console.debug('Tracking error:', error);
    return null;
  }
}
const Tracking = () => {
  const config = (0, _react.useContext)(_ConfigContext.default);
  const url = (0, _UrlContext.useFontdueUrl)();
  (0, _react.useEffect)(() => {
    var _config$tracking$segm, _config$tracking$segm2;
    if (!config.tracking.enabled) return;

    // Generate client-side anonymous ID for cookieless tracking when consent is required
    const clientAnonymousId = config.tracking.consentRequired ? crypto.randomUUID() : undefined;
    if (clientAnonymousId) {
      (0, _consent.setClientAnonymousId)(clientAnonymousId);
    }

    // Only send the client ID when in cookieless mode (consent required but not yet granted)
    const analyticsConsent = (0, _consent.hasConsent)('analytics');
    const effectiveAnonymousId = clientAnonymousId && config.tracking.consentRequired && !analyticsConsent ? clientAnonymousId : undefined;

    // Send initial page view
    const pageViewPromise = sendPageView(url, effectiveAnonymousId);

    // Segment SDK for device-mode integrations (e.g. Google Tag Manager)
    // The analytics factory is provided by the entry point — it calls AnalyticsBrowser.load()
    // when invoked. We defer calling it until after consent when consentRequired is true.
    const loadAnalytics = (_config$tracking$segm = config.tracking.segment) === null || _config$tracking$segm === void 0 ? void 0 : _config$tracking$segm.analytics;
    const integrations = ((_config$tracking$segm2 = config.tracking.segment) === null || _config$tracking$segm2 === void 0 ? void 0 : _config$tracking$segm2.integrations) || {
      All: false,
      'Google Tag Manager': true
    };
    let consentCleanup;
    if (loadAnalytics) {
      const fireSegmentPage = async () => {
        try {
          const analytics = loadAnalytics();
          const anonymousId = await pageViewPromise;
          if (anonymousId) {
            await analytics.setAnonymousId(anonymousId);
          }
          await analytics.page(undefined, undefined, undefined, {
            integrations
          });
        } catch (error) {
          console.debug('Segment SDK error:', error);
        }
      };
      if (config.tracking.consentRequired) {
        consentCleanup = (0, _consent.onConsent)('analytics', fireSegmentPage);
      } else {
        fireSegmentPage();
      }
    }

    // Track client-side route changes
    const originalPushState = history.pushState.bind(history);
    const originalReplaceState = history.replaceState.bind(history);
    const handleRouteChange = () => {
      // Defer slightly to let document.title update
      setTimeout(() => sendPageView(url), 0);
    };
    history.pushState = function () {
      originalPushState(...arguments);
      handleRouteChange();
    };
    history.replaceState = function () {
      originalReplaceState(...arguments);
      handleRouteChange();
    };
    window.addEventListener('popstate', handleRouteChange);
    return () => {
      var _consentCleanup;
      history.pushState = originalPushState;
      history.replaceState = originalReplaceState;
      window.removeEventListener('popstate', handleRouteChange);
      (_consentCleanup = consentCleanup) === null || _consentCleanup === void 0 ? void 0 : _consentCleanup();
    };
  }, []);
  return null;
};
var _default = Tracking;
exports.default = _default;
