// fontdue-js/dist/relay/environment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createEnvironment = createEnvironment;
exports.responseCache = exports.networkFetch = void 0;
exports.useCurrentEnvironment = useCurrentEnvironment;
var _react = require("react");
var _relayRuntime = require("relay-runtime");
var _corsError = require("../corsError");
var _package = require("../../package.json");
// @ts-ignore - JSON import

const IS_SERVER = typeof window === typeof undefined;
const FONTDUE_URL = typeof process !== 'undefined' ? process.env.NEXT_PUBLIC_FONTDUE_URL : undefined;
if (IS_SERVER && (FONTDUE_URL == null || FONTDUE_URL === '')) {
  throw new Error('fontdue-js requires a Fontdue URL to be configured via environment variables on the server, please set a NEXT_PUBLIC_FONTDUE_URL');
}
const STRIPE_INTEGRATION = typeof process !== 'undefined' ? process.env.NEXT_PUBLIC_FONTDUE_STRIPE_INTEGRATION : undefined;
const CACHE_TTL = 10 * 1000; // 10 seconds, to resolve preloaded results

function createNetworkFetch(options) {
  return async function networkFetch(request, variables) {
    const url = (FONTDUE_URL && `${FONTDUE_URL}/graphql`) ?? ((options === null || options === void 0 ? void 0 : options.url) ?? '') + '/graphql';
    for (let attempt = 0; attempt <= 2; attempt++) {
      try {
        const resp = await fetch(url + `?queryName=${request.name}`, {
          method: 'POST',
          credentials: 'include',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            'fontdue-stripe-integration': STRIPE_INTEGRATION ?? ((options === null || options === void 0 ? void 0 : options.stripeIntegration) || 'dynamic'),
            'fontdue-client-version': _package.version
          },
          body: JSON.stringify({
            query: request.text,
            variables
          }),
          // @ts-ignore
          next: {
            tags: ['graphql', `operation:${request.name}`]
          }
        });
        const json = await resp.json();

        // GraphQL returns exceptions (for example, a missing required variable) in the "errors"
        // property of the response. If any exceptions occurred when processing the request,
        // throw an error to indicate to the developer what went wrong.
        if (Array.isArray(json.errors)) {
          var _json$errors, _error$extensions;
          const error = (_json$errors = json.errors) === null || _json$errors === void 0 ? void 0 : _json$errors[0];
          console.error('GraphQL Error:', {
            message: error.message,
            code: (_error$extensions = error.extensions) === null || _error$extensions === void 0 ? void 0 : _error$extensions.code,
            path: error.path
          });
        }
        return json;
      } catch (error) {
        // Retry on network errors (TypeError) before falling through to CORS detection
        if (attempt < 2 && error instanceof TypeError) {
          await new Promise(resolve => setTimeout(resolve, 1000));
          continue;
        }
        if ((0, _corsError.handlePossibleCorsError)(error, url)) {
          return {
            data: undefined,
            errors: [{
              message: 'Cross-origin request blocked'
            }]
          };
        }
        throw error;
      }
    }

    // Unreachable — final iteration either returns or throws
    throw new Error('Unexpected: exhausted retries');
  };
}
const networkFetch = createNetworkFetch();
exports.networkFetch = networkFetch;
const responseCache = IS_SERVER ? null : new _relayRuntime.QueryResponseCache({
  size: 250,
  ttl: CACHE_TTL
});
exports.responseCache = responseCache;
function createNetwork(options) {
  async function fetchResponse(params, variables, cacheConfig) {
    const isQuery = params.operationKind === 'query';
    // @ts-ignore
    const cacheKey = params.id ?? params.cacheID;
    const forceFetch = cacheConfig && cacheConfig.force;
    if (responseCache != null && isQuery && !forceFetch) {
      const fromCache = responseCache.get(cacheKey, variables);
      if (fromCache != null) {
        return Promise.resolve(fromCache);
      }
    }
    return createNetworkFetch(options)(params, variables);
  }
  const network = _relayRuntime.Network.create(fetchResponse);
  return network;
}
function createEnvironment(options) {
  return new _relayRuntime.Environment({
    network: createNetwork(options),
    store: new _relayRuntime.Store(_relayRuntime.RecordSource.create()),
    isServer: IS_SERVER
  });
}
function useCurrentEnvironment(options) {
  if (IS_SERVER) {
    return createEnvironment(options);
  }
  const environmentRef = (0, _react.useRef)(createEnvironment(options));
  return environmentRef.current;
}
