// fontdue-js/dist/relay/useSerializablePreloadedQuery.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useSerializablePreloadedQuery;
var _react = require("react");
var _reactRelay = require("react-relay");
var _environment = require("./environment");
// Convert preloaded query object (with raw GraphQL Response) into
// Relay's PreloadedQuery.

// This hook convert serializable preloaded query
// into Relay's PreloadedQuery object.
// It is also writes this serializable preloaded query
// into QueryResponseCache, so we the network layer
// can use these cache results when fetching data
// in `usePreloadedQuery`.
function useSerializablePreloadedQuery(
//environment: IEnvironment,
preloadQuery) {
  let fetchPolicy = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "store-or-network";
  (0, _react.useMemo)(() => {
    writePreloadedQueryToCache(preloadQuery);
  }, [preloadQuery]);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  return {
    environment,
    fetchKey: preloadQuery.params.id ?? preloadQuery.params.cacheID,
    fetchPolicy,
    isDisposed: false,
    name: preloadQuery.params.name,
    kind: "PreloadedQuery",
    variables: preloadQuery.variables,
    dispose: () => {
      return;
    }
  };
}
function writePreloadedQueryToCache(preloadedQueryObject) {
  // @ts-ignore
  const cacheKey = preloadedQueryObject.params.id ?? preloadedQueryObject.params.cacheID;
  _environment.responseCache === null || _environment.responseCache === void 0 ? void 0 : _environment.responseCache.set(cacheKey, preloadedQueryObject.variables, preloadedQueryObject.response);
}
