// fontdue-js/dist/components/TypeTesters/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TypeTestersPreloadedIDQueryRenderer = TypeTestersPreloadedIDQueryRenderer;
exports.TypeTestersPreloadedSlugQueryRenderer = TypeTestersPreloadedSlugQueryRenderer;
exports.default = TypeTesters;
var _TypeTestersSlugQuery2 = _interopRequireDefault(require("../../__generated__/TypeTestersSlugQuery.graphql"));
var _TypeTestersIDQuery2 = _interopRequireDefault(require("../../__generated__/TypeTestersIDQuery.graphql"));
var _TypeTesters_collection2 = _interopRequireDefault(require("../../__generated__/TypeTesters_collection.graphql"));
var _TypeTestersChangedStylesQuery2 = _interopRequireDefault(require("../../__generated__/TypeTestersChangedStylesQuery.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _PriceBar = _interopRequireDefault(require("../PriceBar"));
var _utils = require("../../utils");
var _TypeTesterContext = _interopRequireDefault(require("../TypeTester/TypeTesterContext"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _useRefetchOnLicenseChanges = require("../../hooks/useRefetchOnLicenseChanges");
var _TypeTestersRefetchQuery = _interopRequireDefault(require("../../__generated__/TypeTestersRefetchQuery.graphql"));
var _TypeTester = _interopRequireDefault(require("../TypeTester"));
var _TypeTesterFloatingToolbar = _interopRequireDefault(require("../TypeTester/TypeTesterFloatingToolbar"));
var _types = require("../TypeTester/types");
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
var _useLicenseAndOrderVariables = require("../../hooks/useLicenseAndOrderVariables");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const findPriceBarCollection = (collection, familyId) => {
  if (collection.id === familyId) return collection;
  if (!collection.families) return null;
  return collection.families.find(family => family.id === familyId);
};
const getTypeTesterGroups = collection => {
  var _collection$typeTeste;
  const edges = (collection === null || collection === void 0 ? void 0 : (_collection$typeTeste = collection.typeTesters) === null || _collection$typeTeste === void 0 ? void 0 : _collection$typeTeste.edges) ?? [];
  const nodes = edges.map(edge => edge === null || edge === void 0 ? void 0 : edge.node).filter(_utils.notEmpty);

  // group type testers by fontStyle
  return nodes.reduce((acc, typeTester) => {
    if (acc.length > 0) {
      const prevTypeTesterGroup = acc[acc.length - 1].slice(0);
      const prevTypeTester = prevTypeTesterGroup[0];
      const getFamilyId = tt => tt.fontStyle && tt.fontStyle.family && tt.fontStyle.family.id;
      const prevFamilySkuId = getFamilyId(prevTypeTester);
      const currFamilySkuId = getFamilyId(typeTester);
      if (prevFamilySkuId === currFamilySkuId) {
        acc[acc.length - 1] = prevTypeTesterGroup.concat([typeTester]);
        return acc;
      }
    }
    return acc.concat([[typeTester]]);
  }, []);
};
const changedStylesQuery = (_TypeTestersChangedStylesQuery2.default.hash && _TypeTestersChangedStylesQuery2.default.hash !== "8bfe3f55f542c9c1cac867aa05f7be53" && console.error("The definition of 'TypeTestersChangedStylesQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTestersChangedStylesQuery2.default);
function TypeTestersComponent(_ref) {
  var _collection$typeTeste2, _collection$typeTeste3, _collection$typeTeste4, _collection$typeTeste5;
  let {
    collection: collectionKey,
    defaultMode,
    autofit,
    features: featuresOverride,
    onFocus,
    onBlur,
    onToolbarOpenClose,
    tags,
    excludeTags,
    ...rest
  } = _ref;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [changedStyles, setChangedStyles] = (0, _react.useState)({});
  const [changedStylesFetchKey, setChangedStylesFetchKey] = (0, _react.useState)(0);
  const {
    typeTester: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const {
    licenseOptionsSpecs,
    orderVariableSelections
  } = (0, _useLicenseAndOrderVariables.useLicenseAndOrderVariables)();
  const [collection, refetch] = (0, _reactRelay.useRefetchableFragment)((_TypeTesters_collection2.default.hash && _TypeTesters_collection2.default.hash !== "0f8422550dd500368594bfedc5f58bf4" && console.error("The definition of 'TypeTesters_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTesters_collection2.default), collectionKey);
  const handleStyleSelect = (0, _react.useCallback)((testerId, fontStyleId) => {
    const newChangedStyles = {
      ...changedStyles,
      [testerId]: fontStyleId
    };
    (0, _reactRelay.fetchQuery)(environment, _TypeTestersChangedStylesQuery2.default, {
      ids: Object.values(newChangedStyles),
      licenseOptions: licenseOptionsSpecs,
      orderVariables: orderVariableSelections
    }).subscribe({
      complete: () => {
        setChangedStyles(newChangedStyles);
        setChangedStylesFetchKey(prev => prev + 1);
      }
    });
  }, [environment, changedStyles, licenseOptionsSpecs, orderVariableSelections]);
  const precartOpen = (0, _reactRedux.useSelector)(state => state.precartOpen);
  const refetchVariables = (0, _react.useMemo)(() => collection !== null && collection !== void 0 && collection.id ? {
    id: collection.id,
    excludeTags,
    tags
  } : undefined, [collection === null || collection === void 0 ? void 0 : collection.id, excludeTags, tags]);
  (0, _useRefetchOnLicenseChanges.useRefetchOnLicenseChanges)({
    environment,
    refetch,
    query: _TypeTestersRefetchQuery.default,
    variables: refetchVariables
  });
  const changedStylesData = (0, _reactRelay.useLazyLoadQuery)(changedStylesQuery, {
    ids: Object.values(changedStyles).filter(Boolean),
    licenseOptions: licenseOptionsSpecs,
    orderVariables: orderVariableSelections
  }, {
    fetchKey: changedStylesFetchKey,
    fetchPolicy: 'store-only'
  });
  if (!collection) return null;
  const testers = (_collection$typeTeste2 = collection.typeTesters) === null || _collection$typeTeste2 === void 0 ? void 0 : (_collection$typeTeste3 = _collection$typeTeste2.edges) === null || _collection$typeTeste3 === void 0 ? void 0 : _collection$typeTeste3.reduce((acc, edge) => {
    var _node$variableSetting, _node$featureSettings;
    const node = edge.node;
    const direction = node.direction === 'rtl' ? 'rtl' : 'ltr';
    let alignment = config.initialAlignment;
    if (direction === 'rtl') alignment = 'right';
    acc[node.id] = {
      content: node.content,
      size: node.size ? parseFloat(node.size) : 72,
      lineHeight: node.lineHeight ? parseFloat(node.lineHeight) : 1,
      autofit: autofit ?? node.autofit ?? false,
      letterSpacing: node.letterSpacing ? parseFloat(node.letterSpacing) : 0,
      direction,
      alignment,
      variableSettings: ((_node$variableSetting = node.variableSettings) === null || _node$variableSetting === void 0 ? void 0 : _node$variableSetting.reduce((acc, _ref2) => {
        let {
          axis,
          value
        } = _ref2;
        acc[axis] = value;
        return acc;
      }, {})) ?? null,
      features: ((_node$featureSettings = node.featureSettings) === null || _node$featureSettings === void 0 ? void 0 : _node$featureSettings.filter(setting => setting.value === '1').map(setting => setting.feature)) ?? null
    };
    return acc;
  }, {});
  if (!testers) return null;
  const features = (0, _types.normalizeFeaturesProp)(featuresOverride) ?? collection.typeTesterFeatures;
  return /*#__PURE__*/_react.default.createElement(_TypeTesterContext.default, {
    key: collection.id,
    defaultMode: defaultMode ?? (autofit ? 'local' : 'group'),
    testers: testers
  }, config.toolsPosition === 'floating' && /*#__PURE__*/_react.default.createElement(_TypeTesterFloatingToolbar.default, {
    testers: ((_collection$typeTeste4 = collection.typeTesters) === null || _collection$typeTeste4 === void 0 ? void 0 : (_collection$typeTeste5 = _collection$typeTeste4.edges) === null || _collection$typeTeste5 === void 0 ? void 0 : _collection$typeTeste5.map(edge => edge.node)) ?? [],
    features: features,
    axes: collection.typeTesterAxes,
    onToolbarOpenClose: onToolbarOpenClose
  }), getTypeTesterGroups(collection).map((typeTesterGroup, i) => {
    var _anyTypeTester$fontSt, _anyTypeTester$fontSt2, _collection$parent;
    if (typeTesterGroup.length === 0) return null;
    const anyTypeTester = typeTesterGroup[0];
    const currentFamilyId = (_anyTypeTester$fontSt = anyTypeTester.fontStyle) === null || _anyTypeTester$fontSt === void 0 ? void 0 : (_anyTypeTester$fontSt2 = _anyTypeTester$fontSt.family) === null || _anyTypeTester$fontSt2 === void 0 ? void 0 : _anyTypeTester$fontSt2.id;

    // when clicking Select/Buy, open modal in relevant superfamily/family
    const productId = ((_collection$parent = collection.parent) === null || _collection$parent === void 0 ? void 0 : _collection$parent.id) ?? collection.id;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: i
    }, config.priceBar && precartOpen && currentFamilyId && /*#__PURE__*/_react.default.createElement(_PriceBar.default, {
      node: findPriceBarCollection(collection, currentFamilyId) ?? null
    }), typeTesterGroup.map((node, j) => {
      const defaultStyle = node.fontStyle;
      const changedStyleId = changedStyles[node.id];

      // Use changed style from query if it exists
      const fontStyle = changedStyleId && changedStylesData !== null && changedStylesData !== void 0 && changedStylesData.nodes ? changedStylesData.nodes.find(style => (style === null || style === void 0 ? void 0 : style.id) === changedStyleId && 'id' in style) ?? defaultStyle : defaultStyle;
      return /*#__PURE__*/_react.default.createElement(_TypeTester.default, _extends({}, rest, {
        id: node.id,
        fontStyle: fontStyle,
        onStyleSelect: fontStyleId => handleStyleSelect(node.id, fontStyleId),
        key: j,
        productId: productId,
        features: features,
        axes: collection.typeTesterAxes,
        tags: node.tags
      }));
    }));
  }));
}

// TODO: add a new component that will be used to query the changed styles, then the base component can call usePreloadedQuery
// https://relay.dev/docs/api-reference/use-query-loader/

const idQuery = (_TypeTestersIDQuery2.default.hash && _TypeTestersIDQuery2.default.hash !== "d3bbb8623a258269114d6e16169cab6b" && console.error("The definition of 'TypeTestersIDQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTestersIDQuery2.default);
function TypeTestersPreloadedIDQueryRenderer(_ref3) {
  let {
    preloadedQuery,
    ...rest
  } = _ref3;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(idQuery, queryRef);
  return /*#__PURE__*/_react.default.createElement(TypeTestersComponent, _extends({}, data, rest));
}
function TypeTestersIDQueryRenderer(_ref4) {
  let {
    collectionId,
    tags,
    excludeTags,
    ...rest
  } = _ref4;
  const data = (0, _reactRelay.useLazyLoadQuery)(idQuery, {
    collectionId,
    tags,
    excludeTags
  });
  return /*#__PURE__*/_react.default.createElement(TypeTestersComponent, _extends({}, data, rest, {
    tags: tags,
    excludeTags: excludeTags
  }));
}
const slugQuery = (_TypeTestersSlugQuery2.default.hash && _TypeTestersSlugQuery2.default.hash !== "a7965166991a9ccda985c694ac45a093" && console.error("The definition of 'TypeTestersSlugQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _TypeTestersSlugQuery2.default);
function TypeTestersPreloadedSlugQueryRenderer(_ref5) {
  var _data$viewer$slug;
  let {
    preloadedQuery,
    ...rest
  } = _ref5;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(slugQuery, queryRef);
  return /*#__PURE__*/_react.default.createElement(TypeTestersComponent, _extends({
    collection: ((_data$viewer$slug = data.viewer.slug) === null || _data$viewer$slug === void 0 ? void 0 : _data$viewer$slug.collection) ?? null
  }, rest));
}
function TypeTestersSlugQueryRenderer(_ref6) {
  var _data$viewer$slug2;
  let {
    collectionSlug,
    tags,
    excludeTags,
    ...rest
  } = _ref6;
  const data = (0, _reactRelay.useLazyLoadQuery)(slugQuery, {
    collectionSlug,
    tags,
    excludeTags
  });
  return /*#__PURE__*/_react.default.createElement(TypeTestersComponent, _extends({
    collection: ((_data$viewer$slug2 = data.viewer.slug) === null || _data$viewer$slug2 === void 0 ? void 0 : _data$viewer$slug2.collection) ?? null
  }, rest, {
    tags: tags,
    excludeTags: excludeTags
  }));
}
function TypeTesters(_ref7) {
  let {
    collectionId,
    collectionSlug,
    ...rest
  } = _ref7;
  if (collectionId) {
    return /*#__PURE__*/_react.default.createElement(TypeTestersIDQueryRenderer, _extends({
      collectionId: collectionId
    }, rest));
  }
  if (collectionSlug) {
    return /*#__PURE__*/_react.default.createElement(TypeTestersSlugQueryRenderer, _extends({
      collectionSlug: collectionSlug
    }, rest));
  }
  return null;
}
