// fontdue-js/dist/components/TypeTesters/TypeTestersElement.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _index = _interopRequireDefault(require("./index"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const getDefaultMode = (stringValue, autofit) => {
  if (autofit) return 'local';
  if (stringValue === 'group') return 'group';
  return 'local';
};
function TypeTestersElement(_ref) {
  let {
    autofit,
    tags,
    excludeTags,
    defaultMode,
    features,
    ...rest
  } = _ref;
  return /*#__PURE__*/_react.default.createElement(_index.default, _extends({
    tags: tags === null || tags === void 0 ? void 0 : tags.split(','),
    excludeTags: excludeTags === null || excludeTags === void 0 ? void 0 : excludeTags.split(','),
    autofit: autofit ? autofit === 'true' : undefined,
    defaultMode: getDefaultMode(defaultMode, autofit === 'true'),
    features: features === null || features === void 0 ? void 0 : features.split(',')
  }, rest));
}
var _default = TypeTestersElement;
exports.default = _default;
