// fontdue-js/dist/components/Cart/CartItem/CartItemLicense.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartItemLicense;
var _CartItemLicense_selection2 = _interopRequireDefault(require("../../../__generated__/CartItemLicense_selection.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Icons = require("../../Icons");
var _CartItemVariable = _interopRequireDefault(require("./CartItemVariable"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function CartItemLicense(_ref) {
  var _selections$filter;
  let {
    onChange,
    onRemoveLicense,
    closeButtonMouseEnter: parentCloseButtonMouseEnter,
    closeButtonMouseLeave: parentCloseButtonMouseLeave,
    selections: selectionsKey
  } = _ref;
  const selections = (0, _reactRelay.useFragment)((_CartItemLicense_selection2.default.hash && _CartItemLicense_selection2.default.hash !== "8a249e2954a1c0d5279a28d666abe1c0" && console.error("The definition of 'CartItemLicense_selection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItemLicense_selection2.default), selectionsKey);
  if (selections.length === 0) return null;
  const license = selections[0].license;
  if (!license) return null;
  const [closeButtonIsHovered, setCloseButtonIsHovered] = (0, _react.useState)(false);
  const closeButtonMouseEnter = () => {
    parentCloseButtonMouseEnter();
    setCloseButtonIsHovered(true);
  };
  const closeButtonMouseLeave = () => {
    parentCloseButtonMouseLeave();
    setCloseButtonIsHovered(false);
  };
  const handleClickRemove = () => {
    closeButtonMouseLeave();
    onRemoveLicense(license.id);
  };
  return /*#__PURE__*/_react.default.createElement("li", {
    key: license.id,
    className: "cart-item__license",
    "data-close-button-hovered": closeButtonIsHovered
  }, /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "cart-item__license__delete-button",
    "aria-label": "Remove license",
    onClick: handleClickRemove,
    onMouseEnter: closeButtonMouseEnter,
    onMouseLeave: closeButtonMouseLeave
  }, /*#__PURE__*/_react.default.createElement(_Icons.X, null)), /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__license-wrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__license-name"
  }, license.name), (((_selections$filter = selections.filter(selection => selection.licenseVariable)) === null || _selections$filter === void 0 ? void 0 : _selections$filter.length) ?? 0) > 0 && /*#__PURE__*/_react.default.createElement("ul", {
    className: "cart-item__license-variables"
  }, selections.map(selection => {
    var _license$licenseVaria;
    return /*#__PURE__*/_react.default.createElement("li", {
      className: "cart-item__license-variable",
      key: selection.id
    }, /*#__PURE__*/_react.default.createElement(_CartItemVariable.default, {
      selection: selection,
      onChange: onChange,
      variable: (_license$licenseVaria = license.licenseVariables) === null || _license$licenseVaria === void 0 ? void 0 : _license$licenseVaria.find(_ref2 => {
        var _selection$licenseVar;
        let {
          id
        } = _ref2;
        return id === ((_selection$licenseVar = selection.licenseVariable) === null || _selection$licenseVar === void 0 ? void 0 : _selection$licenseVar.id);
      })
    }));
  }))));
}
