// fontdue-js/dist/components/Cart/CartItem/CartItemVariable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartItemVariable;
var _CartItemVariable_variable2 = _interopRequireDefault(require("../../../__generated__/CartItemVariable_variable.graphql"));
var _CartItemVariable_selection2 = _interopRequireDefault(require("../../../__generated__/CartItemVariable_selection.graphql"));
var _react = _interopRequireDefault(require("react"));
var _VariableTableAmounts = _interopRequireDefault(require("../../VariableTableAmounts"));
var _VariableTextInput = _interopRequireDefault(require("./VariableTextInput"));
var _Icons = require("../../Icons");
var _reactRelay = require("react-relay");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function CartItemVariable(_ref) {
  var _variableOptions, _selection$licenseOpt2, _selection$licenseOpt3, _selection$licenseOpt4, _variableOptions2;
  let {
    // selection may contain *deleted* licenseVariable and/or licenseOptions
    selection: selectionKey,
    // variable is the "public"-facing list of variables/options which do NOT included deleted stuff
    variable: variableKey,
    onChange
  } = _ref;
  const selection = (0, _reactRelay.useFragment)((_CartItemVariable_selection2.default.hash && _CartItemVariable_selection2.default.hash !== "9a658e82957a12990886e5e14dae37bc" && console.error("The definition of 'CartItemVariable_selection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItemVariable_selection2.default), selectionKey);
  const variable = (0, _reactRelay.useFragment)((_CartItemVariable_variable2.default.hash && _CartItemVariable_variable2.default.hash !== "4f53aadeee4ad6e8d84a087459ba2343" && console.error("The definition of 'CartItemVariable_variable' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItemVariable_variable2.default), variableKey);
  if (!selection.licenseVariable) return null;
  let variableOptions = (variable === null || variable === void 0 ? void 0 : variable.licenseOptions) ?? [];
  if (variableOptions && selection.licenseOption && !((_variableOptions = variableOptions) !== null && _variableOptions !== void 0 && _variableOptions.find(option => {
    var _selection$licenseOpt;
    return option.id === ((_selection$licenseOpt = selection.licenseOption) === null || _selection$licenseOpt === void 0 ? void 0 : _selection$licenseOpt.id);
  }))) {
    // show option even if it's soft-deleted (it's not found in the list of
    //   variable.variableOptions, but it is available on the selection)
    variableOptions = [selection.licenseOption, ...variableOptions];
  }
  const handleSelectChange = e => {
    var _selection$licenseVar;
    onChange({
      id: selection.id,
      licenseId: selection.license.id,
      licenseVariableId: (_selection$licenseVar = selection.licenseVariable) === null || _selection$licenseVar === void 0 ? void 0 : _selection$licenseVar.id,
      licenseOptionId: e.target.value
    });
  };
  const handleTextChange = value => {
    var _selection$licenseVar2;
    onChange({
      id: selection.id,
      licenseId: selection.license.id,
      licenseVariableId: (_selection$licenseVar2 = selection.licenseVariable) === null || _selection$licenseVar2 === void 0 ? void 0 : _selection$licenseVar2.id,
      variableText: value
    });
  };
  return /*#__PURE__*/_react.default.createElement("label", {
    className: "cart-item__license-variable__option",
    "data-variable-type": selection.licenseVariable.variableType
  }, selection.licenseVariable.variableType === 'select' || selection.licenseVariable.variableType === 'table' ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, selection.licenseVariable.variableType === 'table' ? /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__license-variable__option-name"
  }, (_selection$licenseOpt2 = selection.licenseOption) === null || _selection$licenseOpt2 === void 0 ? void 0 : _selection$licenseOpt2.name), /*#__PURE__*/_react.default.createElement(_VariableTableAmounts.default, {
    variable: selection.licenseVariable,
    option: selection.licenseOption
  })) : (_selection$licenseOpt3 = selection.licenseOption) === null || _selection$licenseOpt3 === void 0 ? void 0 : _selection$licenseOpt3.amount, /*#__PURE__*/_react.default.createElement(_Icons.CarrotDown, null), /*#__PURE__*/_react.default.createElement("select", {
    value: (_selection$licenseOpt4 = selection.licenseOption) === null || _selection$licenseOpt4 === void 0 ? void 0 : _selection$licenseOpt4.id,
    onChange: handleSelectChange
  }, (_variableOptions2 = variableOptions) === null || _variableOptions2 === void 0 ? void 0 : _variableOptions2.map((option, i) => {
    var _selection$licenseVar3;
    return /*#__PURE__*/_react.default.createElement("option", {
      key: i,
      value: option.id
    }, ((_selection$licenseVar3 = selection.licenseVariable) === null || _selection$licenseVar3 === void 0 ? void 0 : _selection$licenseVar3.variableType) === 'table' ? option.name : option.amount);
  }))) : /*#__PURE__*/_react.default.createElement(_VariableTextInput.default, {
    defaultValue: selection.variableText,
    placeholder: `${selection.licenseVariable.name} *`,
    onChange: handleTextChange
  }));
}
