// fontdue-js/dist/components/Cart/CartItem/CartItemAdditionalLicenses.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartItemAdditionalLicenses;
var _CartItemAdditionalLicenses_orderItem2 = _interopRequireDefault(require("../../../__generated__/CartItemAdditionalLicenses_orderItem.graphql"));
var _CartItemAdditionalLicenses_licenses2 = _interopRequireDefault(require("../../../__generated__/CartItemAdditionalLicenses_licenses.graphql"));
var _react = _interopRequireDefault(require("react"));
var _reactRelay = require("react-relay");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function CartItemAdditionalLicenses(_ref) {
  var _orderItem$licenseSel, _orderItem$licenseSel2;
  let {
    licenses: licensesKey,
    orderItem: orderItemKey,
    onAddLicense
  } = _ref;
  const licenses = (0, _reactRelay.useFragment)((_CartItemAdditionalLicenses_licenses2.default.hash && _CartItemAdditionalLicenses_licenses2.default.hash !== "a00d0c8ac3643f7c12911fad6f62ec4e" && console.error("The definition of 'CartItemAdditionalLicenses_licenses' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItemAdditionalLicenses_licenses2.default), licensesKey);
  const orderItem = (0, _reactRelay.useFragment)((_CartItemAdditionalLicenses_orderItem2.default.hash && _CartItemAdditionalLicenses_orderItem2.default.hash !== "52f514c1954a1d93863d7dd4c4871e3a" && console.error("The definition of 'CartItemAdditionalLicenses_orderItem' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItemAdditionalLicenses_orderItem2.default), orderItemKey);
  if (!licenses) return null;
  const selectedLicenseIds = new Set(((_orderItem$licenseSel = orderItem.licenseSelections) === null || _orderItem$licenseSel === void 0 ? void 0 : _orderItem$licenseSel.map(_ref2 => {
    let {
      license
    } = _ref2;
    return license.id;
  })) ?? []);
  const excludedLicenseIds = new Set(((_orderItem$licenseSel2 = orderItem.licenseSelections) === null || _orderItem$licenseSel2 === void 0 ? void 0 : _orderItem$licenseSel2.flatMap(_ref3 => {
    let {
      license
    } = _ref3;
    return license.excludedLicenses.map(l => l.id);
  })) ?? []);
  const additionalLicenses = licenses.filter(license => !selectedLicenseIds.has(license.id) && !excludedLicenseIds.has(license.id));
  if (additionalLicenses.length < 1) return null;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__additional-licenses"
  }, "+ Add:", ' ', additionalLicenses.map(license => /*#__PURE__*/_react.default.createElement("button", {
    type: "button",
    className: "cart-item__additional-licenses__item",
    key: license.id,
    onClick: () => onAddLicense(license.id)
  }, license.name)));
}
