// fontdue-js/dist/components/Cart/CartItem/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CartItem;
var _CartItem_node2 = _interopRequireDefault(require("../../../__generated__/CartItem_node.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Price = _interopRequireDefault(require("../../Price"));
var _CartItemLicense = _interopRequireDefault(require("./CartItemLicense"));
var _CartItemAdditionalLicenses = _interopRequireDefault(require("./CartItemAdditionalLicenses"));
var _CartItemProduct = _interopRequireDefault(require("./CartItemProduct"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function CartItem(_ref) {
  var _node$licenseSelectio, _node$licenseSelectio2, _node$sku2, _node$sku3, _node$sku3$product;
  let {
    node: nodeKey,
    onChange,
    errors
  } = _ref;
  const node = (0, _reactRelay.useFragment)((_CartItem_node2.default.hash && _CartItem_node2.default.hash !== "c17a59a874f66eaed523d8928f9f1171" && console.error("The definition of 'CartItem_node' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CartItem_node2.default), nodeKey);
  const [closeButtonIsHovered, setCloseButtonIsHovered] = (0, _react.useState)(false);
  const closeButtonMouseEnter = () => {
    setCloseButtonIsHovered(true);
  };
  const closeButtonMouseLeave = () => {
    setCloseButtonIsHovered(false);
  };
  const groupedSelections = (_node$licenseSelectio = node.licenseSelections) === null || _node$licenseSelectio === void 0 ? void 0 : _node$licenseSelectio.reduce((res, selection) => ({
    ...res,
    [selection.license.id]: (res[selection.license.id] ?? []).concat(selection)
  }), {});
  if (!groupedSelections) return null;
  const existingLicenseSelectionsInputs = ((_node$licenseSelectio2 = node.licenseSelections) === null || _node$licenseSelectio2 === void 0 ? void 0 : _node$licenseSelectio2.map(s => {
    var _s$licenseVariable, _s$licenseOption;
    return {
      id: s.id,
      licenseId: s.license.id,
      licenseVariableId: (_s$licenseVariable = s.licenseVariable) === null || _s$licenseVariable === void 0 ? void 0 : _s$licenseVariable.id,
      licenseOptionId: (_s$licenseOption = s.licenseOption) === null || _s$licenseOption === void 0 ? void 0 : _s$licenseOption.id,
      variableText: s.variableText
    };
  })) ?? [];
  const handleChangeSelection = selection => {
    onChange({
      id: node.id,
      licenseSelections: existingLicenseSelectionsInputs.filter(s => s.id !== selection.id).concat(selection)
    });
  };
  const handleRemoveLicense = licenseId => {
    onChange({
      id: node.id,
      licenseSelections: existingLicenseSelectionsInputs.filter(input => input.licenseId !== licenseId)
    });
  };
  const handleAddLicense = licenseId => {
    var _node$sku, _node$sku$product, _node$sku$product$lic, _license$licenseVaria;
    const license = (_node$sku = node.sku) === null || _node$sku === void 0 ? void 0 : (_node$sku$product = _node$sku.product) === null || _node$sku$product === void 0 ? void 0 : (_node$sku$product$lic = _node$sku$product.licenses) === null || _node$sku$product$lic === void 0 ? void 0 : _node$sku$product$lic.find(l => l.id === licenseId);
    if (!license) return;
    const newLicenseSelectionInput = (_license$licenseVaria = license.licenseVariables) !== null && _license$licenseVaria !== void 0 && _license$licenseVaria.length ? license.licenseVariables.map(variable => {
      var _variable$licenseOpti, _variable$licenseOpti2;
      return {
        licenseId,
        licenseVariableId: variable.id,
        licenseOptionId: (_variable$licenseOpti = variable.licenseOptions) === null || _variable$licenseOpti === void 0 ? void 0 : (_variable$licenseOpti2 = _variable$licenseOpti[0]) === null || _variable$licenseOpti2 === void 0 ? void 0 : _variable$licenseOpti2.id
      };
    }) : [{
      licenseId
    }];
    onChange({
      id: node.id,
      licenseSelections: existingLicenseSelectionsInputs.concat(newLicenseSelectionInput)
    });
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item",
    "data-close-button-hovered": closeButtonIsHovered,
    "data-license-count": Object.keys(groupedSelections).length
  }, /*#__PURE__*/_react.default.createElement(_CartItemProduct.default, {
    product: ((_node$sku2 = node.sku) === null || _node$sku2 === void 0 ? void 0 : _node$sku2.product) ?? null
  }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("ul", {
    className: "cart-item__licenses"
  }, Object.keys(groupedSelections).filter(licenseId => groupedSelections[licenseId].length > 0).map(licenseId => /*#__PURE__*/_react.default.createElement(_CartItemLicense.default, {
    key: licenseId,
    onChange: handleChangeSelection,
    onRemoveLicense: handleRemoveLicense,
    selections: groupedSelections[licenseId],
    closeButtonMouseEnter: closeButtonMouseEnter,
    closeButtonMouseLeave: closeButtonMouseLeave
  }))), errors !== null && errors !== void 0 && errors.licenseSelections ? /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__error"
  }, "License selections ", errors === null || errors === void 0 ? void 0 : errors.licenseSelections) : null), node.price && /*#__PURE__*/_react.default.createElement("div", {
    className: "cart-item__price"
  }, /*#__PURE__*/_react.default.createElement(_Price.default, {
    price: node.price
  })), /*#__PURE__*/_react.default.createElement(_CartItemAdditionalLicenses.default, {
    licenses: ((_node$sku3 = node.sku) === null || _node$sku3 === void 0 ? void 0 : (_node$sku3$product = _node$sku3.product) === null || _node$sku3$product === void 0 ? void 0 : _node$sku3$product.licenses) ?? null,
    orderItem: node,
    onAddLicense: handleAddLicense
  }));
}
