// fontdue-js/dist/components/Cart/CartItem/VariableTextInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = VariableTextInput;
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function VariableTextInput(_ref) {
  let {
    defaultValue,
    onChange,
    placeholder,
    ...rest
  } = _ref;
  const [value, setValue] = (0, _react.useState)(defaultValue || '');
  const [editing, setEditing] = (0, _react.useState)(!defaultValue);
  const [submitting, setSubmitting] = (0, _react.useState)(false);
  const inputRef = (0, _react.useRef)(null);
  function submit() {
    setEditing(false);
    if (value) {
      if (defaultValue !== value) {
        onChange(value);
        setSubmitting(true);
      }
    } else {
      setValue(defaultValue || '');
      setEditing(!defaultValue);
    }
  }
  (0, _react.useEffect)(() => {
    if (submitting && value === defaultValue) setSubmitting(false);
  }, [submitting, setSubmitting, value, defaultValue]);
  function handleBlur() {
    submit();
  }
  function handleSubmit(e) {
    e.preventDefault();
    submit();
  }
  function handleChange(e) {
    setValue(e.target.value);
  }
  function handleEditButtonClick() {
    setEditing(true);
  }
  (0, _react.useEffect)(() => {
    if (editing) {
      var _inputRef$current;
      (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  }, [editing]);
  if (submitting) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "cart-item__license-variable__loading"
    }, "Updating...");
  }
  if (editing) {
    return /*#__PURE__*/_react.default.createElement("form", {
      onSubmit: handleSubmit,
      className: "cart-item__license-variable__text-input-form"
    }, /*#__PURE__*/_react.default.createElement("input", _extends({
      ref: inputRef,
      className: "cart-item__license-variable__text-input",
      "data-empty": !value,
      type: "text",
      value: value,
      onChange: handleChange,
      onBlur: handleBlur,
      placeholder: `Enter ${placeholder}`
    }, rest)));
  }
  return /*#__PURE__*/_react.default.createElement("button", {
    className: "cart-item__license-variable__text-input-button",
    type: "button",
    onClick: handleEditButtonClick
  }, defaultValue, " ", /*#__PURE__*/_react.default.createElement("span", null, "Edit"));
}
