// fontdue-js/dist/components/OrderVariableSelection/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = OrderVariableSelection;
var _OrderVariableSelection_variables2 = _interopRequireDefault(require("../../__generated__/OrderVariableSelection_variables.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _Button = _interopRequireDefault(require("../elements/Button"));
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
var _Icons = require("../Icons");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// Ghost variable type

// Component props

function Variable(_ref) {
  var _variable$orderVariab2, _options, _options2;
  let {
    variable,
    selection,
    onToggle,
    ghost = false,
    onRemoveGhost,
    countries
  } = _ref;
  const {
    form: config
  } = (0, _react.useContext)(_ConfigContext.default);
  const isSelected = Boolean(selection);
  const handleClickVariable = (0, _react.useCallback)(() => {
    var _variable$orderVariab;
    // Only toggle boolean variables when clicked
    if (variable.variableType === 'boolean') {
      onToggle(variable.id, null);
    } else if (variable.variableType === 'select' && !isSelected && (_variable$orderVariab = variable.orderVariableOptions) !== null && _variable$orderVariab !== void 0 && _variable$orderVariab.length) {
      // For unselected select variables, select the first option by default
      onToggle(variable.id, variable.orderVariableOptions[0].id);
    }
    // Country variables are now static (non-editable)
  }, [variable, isSelected, onToggle]);
  const handleChangeVariableOption = (0, _react.useCallback)(e => {
    onToggle(variable.id, e.target.value);
  }, [variable.id, onToggle]);
  let options = (_variable$orderVariab2 = variable.orderVariableOptions) === null || _variable$orderVariab2 === void 0 ? void 0 : _variable$orderVariab2.map(option => ({
    ...option,
    disabled: false
  }));

  // if there are any selections NOT present in viewer.orderVariables.orderVariableOptions,
  // but they ARE present in the selection, it means they've been soft-deleted in the backend.
  // but we can include it in the list of options in a disabled state. the user can still
  // checkout with the option they previoulsy selected if they want to
  if (selection !== null && selection !== void 0 && selection.orderVariableOptionId && options && !options.find(option => (selection === null || selection === void 0 ? void 0 : selection.orderVariableOptionId) == option.id)) {
    options = options.concat({
      id: selection.orderVariableOptionId,
      amount: selection.orderVariableOptionAmount ?? null,
      disabled: true
    });
  }
  const selectedOption = (_options = options) === null || _options === void 0 ? void 0 : _options.find(o => o.id === (selection === null || selection === void 0 ? void 0 : selection.orderVariableOptionId));
  const selectedCountry = countries === null || countries === void 0 ? void 0 : countries.find(c => c.code === (selection === null || selection === void 0 ? void 0 : selection.countryCode));
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__item",
    "data-variable-type": variable.variableType,
    "data-ghost": ghost
  }, /*#__PURE__*/_react.default.createElement(_Button.default, {
    className: "store-modal__order-variable-selection__button",
    onClick: handleClickVariable,
    "data-selected": isSelected
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__label"
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "store-modal__order-variable-selection__icon-pair"
  }, variable.variableType === 'boolean' ? isSelected ? config.checkboxStyle === 'check' ? /*#__PURE__*/_react.default.createElement(_Icons.CheckboxChecked, null) : /*#__PURE__*/_react.default.createElement(_Icons.CheckboxCrossed, null) : /*#__PURE__*/_react.default.createElement(_Icons.Checkbox, null) : null, variable.name)), variable.variableType === 'select' ? /*#__PURE__*/_react.default.createElement("label", {
    className: "store-modal__order-variable-selection__select-label"
  }, selectedOption ? selectedOption.amount : `Select an option`, /*#__PURE__*/_react.default.createElement(_Icons.CarrotDown, null), /*#__PURE__*/_react.default.createElement("select", {
    className: "store-modal__order-variable-selection__select",
    value: (selection === null || selection === void 0 ? void 0 : selection.orderVariableOptionId) || '',
    onChange: handleChangeVariableOption,
    onClick: e => e.stopPropagation()
  }, /*#__PURE__*/_react.default.createElement("option", {
    value: "",
    disabled: true
  }, "Select an option"), (_options2 = options) === null || _options2 === void 0 ? void 0 : _options2.map(_ref2 => {
    let {
      id,
      amount,
      disabled
    } = _ref2;
    return /*#__PURE__*/_react.default.createElement("option", {
      key: id,
      value: id,
      disabled: disabled
    }, amount, " ", disabled ? '(Deprecated)' : null);
  }))) : null, variable.variableType === 'country' && selectedCountry ? /*#__PURE__*/_react.default.createElement("span", {
    className: "store-modal__order-variable-selection__country-value"
  }, selectedCountry.name) : null), variable.description ? /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__description",
    dangerouslySetInnerHTML: {
      __html: variable.description
    }
  }) : null, ghost && /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__ghost-container"
  }, "This variable has been removed from our system and is deprecated", ' ', /*#__PURE__*/_react.default.createElement("button", {
    className: "store-modal__order-variable-selection__ghost-remove",
    onClick: onRemoveGhost
  }, "Remove")));
}
function OrderVariableSelection(_ref3) {
  let {
    variables: variablesKey,
    selections,
    onSelectionChange,
    disabled,
    ghostVariables = [],
    onRemoveGhostVariable,
    descriptionOverrides,
    children: {
      label
    }
  } = _ref3;
  const variables = (0, _reactRelay.useFragment)((_OrderVariableSelection_variables2.default.hash && _OrderVariableSelection_variables2.default.hash !== "b73b43b61cd5edd5face7d4a5615b2f5" && console.error("The definition of 'OrderVariableSelection_variables' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _OrderVariableSelection_variables2.default), variablesKey);
  const handleSelectionChange = (0, _react.useCallback)((variableId, optionId, countryCode) => {
    var _variables$orderVaria;
    // Update the selections with the new value
    const newSelections = selections.map(selection => ({
      orderVariableId: selection.orderVariableId,
      orderVariableOptionId: selection.orderVariableOptionId,
      countryCode: selection.countryCode
    }));
    const existingIndex = newSelections.findIndex(s => s.orderVariableId === variableId);

    // Find the variable to determine its type
    const variable = (_variables$orderVaria = variables.orderVariables) === null || _variables$orderVaria === void 0 ? void 0 : _variables$orderVaria.find(v => v.id === variableId);
    const isBoolean = (variable === null || variable === void 0 ? void 0 : variable.variableType) === 'boolean';
    if (existingIndex >= 0) {
      // For boolean variables, if it's already selected, remove it entirely
      if (isBoolean && newSelections[existingIndex].orderVariableOptionId === optionId) {
        newSelections.splice(existingIndex, 1);
      } else {
        // Update existing selection
        // For country variables: clear orderVariableOptionId when setting countryCode
        // For select/boolean variables: clear countryCode when setting orderVariableOptionId
        const isCountryVariable = (variable === null || variable === void 0 ? void 0 : variable.variableType) === 'country';
        newSelections[existingIndex] = {
          ...newSelections[existingIndex],
          orderVariableOptionId: isCountryVariable ? null : optionId,
          countryCode: isCountryVariable ? countryCode ?? newSelections[existingIndex].countryCode : null
        };
      }
    } else {
      // Add new selection
      newSelections.push({
        orderVariableId: variableId,
        orderVariableOptionId: optionId,
        countryCode: countryCode
      });
    }
    onSelectionChange(newSelections);
  }, [selections, onSelectionChange, variables.orderVariables]);
  if (!variables.orderVariables || variables.orderVariables.length === 0) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "store-modal__order-variable-selection__container",
    "data-disabled": disabled
  }, /*#__PURE__*/_react.default.createElement("h3", {
    className: "store-modal__order-variable-selection__title"
  }, label), variables.orderVariables.map(variable => {
    var _variables$countries;
    const currentSelection = selections.find(s => s.orderVariableId === variable.id);

    // Use description override if provided, otherwise use the variable's description
    const effectiveVariable = (descriptionOverrides === null || descriptionOverrides === void 0 ? void 0 : descriptionOverrides[variable.id]) !== undefined ? {
      ...variable,
      description: descriptionOverrides[variable.id]
    } : variable;
    return /*#__PURE__*/_react.default.createElement(Variable, {
      key: variable.id,
      variable: effectiveVariable,
      selection: currentSelection,
      countries: ((_variables$countries = variables.countries) === null || _variables$countries === void 0 ? void 0 : _variables$countries.filter(c => c !== null)) ?? null,
      onToggle: (orderVariableId, orderVariableOptionId, countryCode) => handleSelectionChange(orderVariableId, orderVariableOptionId !== undefined ? orderVariableOptionId : null, countryCode)
    });
  }), ghostVariables.map(ghost => {
    var _ghost$orderVariableO, _ghost$orderVariableO2, _variables$countries2;
    return /*#__PURE__*/_react.default.createElement(Variable, {
      key: ghost.orderVariable.id,
      variable: {
        ...ghost.orderVariable,
        orderVariableOptions: ghost.orderVariableOption ? [ghost.orderVariableOption] : null
      },
      selection: {
        orderVariableId: ghost.orderVariable.id,
        orderVariableOptionId: ((_ghost$orderVariableO = ghost.orderVariableOption) === null || _ghost$orderVariableO === void 0 ? void 0 : _ghost$orderVariableO.id) ?? null,
        orderVariableOptionAmount: ((_ghost$orderVariableO2 = ghost.orderVariableOption) === null || _ghost$orderVariableO2 === void 0 ? void 0 : _ghost$orderVariableO2.amount) ?? null
      },
      countries: ((_variables$countries2 = variables.countries) === null || _variables$countries2 === void 0 ? void 0 : _variables$countries2.filter(c => c !== null)) ?? null,
      ghost: true,
      onToggle: () => {} // Ghost variables don't toggle
      ,
      onRemoveGhost: onRemoveGhostVariable ? () => onRemoveGhostVariable(ghost.orderVariable.id) : undefined
    });
  }));
}
