// fontdue-js/dist/components/OrderVariableSelection/OrderVariableSelectionRedux.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = OrderVariableSelectionRedux;
var _OrderVariableSelectionRedux_viewer2 = _interopRequireDefault(require("../../__generated__/OrderVariableSelectionRedux_viewer.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactRedux = require("react-redux");
var _index = _interopRequireDefault(require("./index"));
var _OrderVariableSelectionReduxRefetchQuery = _interopRequireDefault(require("../../__generated__/OrderVariableSelectionReduxRefetchQuery.graphql"));
var _useRefetchOnLicenseChanges = require("../../hooks/useRefetchOnLicenseChanges");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function OrderVariableSelectionRedux(_ref) {
  let {
    viewer: viewerKey,
    disabled,
    children
  } = _ref;
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const [viewer, refetch] = (0, _reactRelay.useRefetchableFragment)((_OrderVariableSelectionRedux_viewer2.default.hash && _OrderVariableSelectionRedux_viewer2.default.hash !== "eaa5e1fdf47cb582dd3c7bdea0a5e6b2" && console.error("The definition of 'OrderVariableSelectionRedux_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _OrderVariableSelectionRedux_viewer2.default), viewerKey);
  const refetchVariables = (0, _react.useMemo)(() => ({}), []);
  (0, _useRefetchOnLicenseChanges.useRefetchOnLicenseChanges)({
    environment,
    refetch,
    query: _OrderVariableSelectionReduxRefetchQuery.default,
    variables: refetchVariables
  });
  const dispatch = (0, _reactRedux.useDispatch)();
  const selections = (0, _reactRedux.useSelector)(state => state.orderVariableSelections);
  const handleSelectionChange = (0, _react.useCallback)(newSelections => {
    dispatch({
      type: 'UPDATE_ORDER_VARIABLE_SELECTIONS',
      selections: newSelections
    });
  }, [dispatch]);

  // Auto-populate Redux state with detected country for country-type variables
  (0, _react.useEffect)(() => {
    if (!viewer.detectedCountry || !viewer.orderVariables) return;

    // Find country-type variables that don't have a selection yet
    const countryVariablesNeedingSelection = viewer.orderVariables.filter(v => v.variableType === 'country').filter(v => !selections.find(s => s.orderVariableId === v.id));
    if (countryVariablesNeedingSelection.length === 0) return;

    // Pre-populate with detected country
    const newSelections = [...selections, ...countryVariablesNeedingSelection.map(v => ({
      orderVariableId: v.id,
      orderVariableOptionId: null,
      countryCode: viewer.detectedCountry
    }))];
    dispatch({
      type: 'UPDATE_ORDER_VARIABLE_SELECTIONS',
      selections: newSelections
    });
  }, [viewer.detectedCountry, viewer.orderVariables, selections, dispatch]);
  return /*#__PURE__*/_react.default.createElement(_index.default, {
    variables: viewer,
    selections: selections,
    onSelectionChange: handleSelectionChange,
    disabled: disabled,
    children: children
  });
}
