// fontdue-js/dist/components/CharacterViewer/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CharacterViewerPreloadedIDQueryRenderer = CharacterViewerPreloadedIDQueryRenderer;
exports.CharacterViewerPreloadedSlugQueryRenderer = CharacterViewerPreloadedSlugQueryRenderer;
exports.default = CharacterViewer;
var _CharacterViewerSlugQuery2 = _interopRequireDefault(require("../../__generated__/CharacterViewerSlugQuery.graphql"));
var _CharacterViewerIDQuery2 = _interopRequireDefault(require("../../__generated__/CharacterViewerIDQuery.graphql"));
var _CharacterViewer_style2 = _interopRequireDefault(require("../../__generated__/CharacterViewer_style.graphql"));
var _CharacterViewer_collection2 = _interopRequireDefault(require("../../__generated__/CharacterViewer_collection.graphql"));
var _CharacterViewer_family2 = _interopRequireDefault(require("../../__generated__/CharacterViewer_family.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _resizeObserver = _interopRequireDefault(require("@react-hook/resize-observer"));
var _retryImport = _interopRequireDefault(require("../../retryImport"));
var _utils = require("../../utils");
var _useFontStyle = _interopRequireDefault(require("../useFontStyle"));
var _useSerializablePreloadedQuery = _interopRequireDefault(require("../../relay/useSerializablePreloadedQuery"));
var _StyleSelect = _interopRequireDefault(require("./StyleSelect"));
var _Checkbox = _interopRequireDefault(require("../Checkbox"));
var _CharacterViewerStyleRefetchQuery = _interopRequireDefault(require("../../__generated__/CharacterViewerStyleRefetchQuery.graphql"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
function useSize() {
  const ref = (0, _react.useRef)(null);
  const [size, setSize] = _react.default.useState();
  (0, _react.useEffect)(() => {
    if (!ref.current) return;
    setSize(ref.current.getBoundingClientRect());
  }, [ref]);
  (0, _resizeObserver.default)(ref, entry => setSize(entry.contentRect), {
    polyfill: typeof window === 'undefined' ? ServerSideResizeObserverPolyfill : null
  });
  return {
    ref,
    width: size === null || size === void 0 ? void 0 : size.width,
    height: size === null || size === void 0 ? void 0 : size.height
  };
}
class ServerSideResizeObserverPolyfill {
  constructor(_callback) {}
  observe(_target, _options) {}
  unobserve(_target) {}
}
function Character(props) {
  const ref = (0, _react.useRef)(null);
  const [size, setSize] = (0, _react.useState)(1);
  (0, _resizeObserver.default)(ref, entry => {
    if (entry.target.scrollWidth > entry.target.clientWidth) {
      setSize(size + 1);
    }
  }, {
    polyfill: typeof window === 'undefined' ? ServerSideResizeObserverPolyfill : null
  });
  return /*#__PURE__*/_react.default.createElement("div", _extends({
    ref: ref
  }, props, {
    "data-size": size
  }));
}
function GlyphMeta(_ref) {
  let {
    glyph,
    featuresChecked,
    setFeaturesChecked,
    getGlyphName,
    getUnicodeName
  } = _ref;
  if (!glyph) return null;
  const unicodeName = getUnicodeName(glyph);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", null, "Glyph name"), /*#__PURE__*/_react.default.createElement("span", null, getGlyphName(glyph)), unicodeName && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", null, "Unicode name"), /*#__PURE__*/_react.default.createElement("span", {
    title: unicodeName
  }, unicodeName)), /*#__PURE__*/_react.default.createElement("span", null, "Unicode decimal"), /*#__PURE__*/_react.default.createElement("span", null, glyph.string.split('').map(str => str.charCodeAt(0).toString(10)).join(', ')), /*#__PURE__*/_react.default.createElement("span", null, "Unicode hex"), /*#__PURE__*/_react.default.createElement("span", null, glyph.string.split('').map(str => str.charCodeAt(0).toString(16).padStart(4, '0')).join(', ')), glyph.features && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", null, "OpenType feature", glyph.features.length > 1 ? 's' : ''), /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement("label", null, /*#__PURE__*/_react.default.createElement(_Checkbox.default, {
    checked: featuresChecked,
    onChange: e => setFeaturesChecked(e.target.checked)
  }), glyph.features.join(', ')))));
}
function useUnicodeData() {
  const [data, setData] = (0, _react.useState)();
  (0, _react.useEffect)(() => {
    function fetchData() {
      (0, _retryImport.default)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../../data/unicodeData')))).then(data => {
        if (!ignore) setData(data.default);
      }).catch(() => {});
    }
    let ignore = false;
    fetchData();
    return () => {
      ignore = true;
    };
  }, []);
  return data;
}
function compareGlyphs(a, b) {
  var _a$features, _b$features;
  if (!a || !b) return false;
  if (a.string !== b.string) return false;
  if (((_a$features = a.features) === null || _a$features === void 0 ? void 0 : _a$features.length) !== ((_b$features = b.features) === null || _b$features === void 0 ? void 0 : _b$features.length)) return false;
  if (a.features && b.features) {
    return a.features.every(feature => {
      var _b$features2;
      return (_b$features2 = b.features) === null || _b$features2 === void 0 ? void 0 : _b$features2.includes(feature);
    });
  }
  return true;
}
function charCode(c) {
  return c.codePointAt(0) ?? 0;
}
function fromCharCode(code) {
  return String.fromCodePoint(code);
}
function flattenCharacterList(charSet, glyphNames) {
  if (!glyphNames || !charSet || !charSet.characters) return null;
  return {
    features: charSet.features,
    characters: charSet.characters.reduce((acc, item) => {
      if (item.__typename === 'CharacterString') {
        acc.push(item.string);
      } else if (item.__typename === 'CharacterRange') {
        for (let code = charCode(item.first); code <= charCode(item.last); code++) {
          acc.push(fromCharCode(code));
        }
      }
      return acc;
    }, []).filter(string => glyphNames.some(_ref2 => {
      let {
        characters,
        features
      } = _ref2;
      return compareGlyphs({
        string,
        features: charSet.features
      }, {
        string: characters,
        features
      });
    }))
  };
}
_CharacterViewer_family2.default.hash && _CharacterViewer_family2.default.hash !== "162b1a99fe187b20ee8560018cd5b3ce" && console.error("The definition of 'CharacterViewer_family' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CharacterViewer_family2.default;
function areCombiningChars(input) {
  for (const char of input) {
    const codePoint = char.codePointAt(0);
    if (codePoint === undefined) return false;
    const isDiacriticalMark = codePoint >= 0x0300 && codePoint <= 0x036f;
    const isSpacingModifierLetter = codePoint >= 0x02b0 && codePoint <= 0x02ff;
    const isCombiningDiacriticalSupplement = codePoint >= 0x1dc0 && codePoint <= 0x1dff;
    const isCombiningDiacriticalExtended = codePoint >= 0x1ab0 && codePoint <= 0x1aff;
    const isCombiningHalfMark = codePoint >= 0xfe20 && codePoint <= 0xfe2f;
    const isCombiningSymbolMark = codePoint >= 0x20d0 && codePoint <= 0x20ff;
    const isCombiningChar = isDiacriticalMark || isSpacingModifierLetter || isCombiningDiacriticalSupplement || isCombiningDiacriticalExtended || isCombiningHalfMark || isCombiningSymbolMark;
    if (!isCombiningChar) return false;
  }
  return true;
}

// Arabic positional features that don't render correctly in Safari
// These require text shaping that Safari doesn't properly support with ZWJ
const UNSUPPORTED_POSITIONAL_FEATURES = ['init', 'medi', 'fina'];
function hasUnsupportedPositionalFeatures(features) {
  if (!features) return false;
  return features.some(f => UNSUPPORTED_POSITIONAL_FEATURES.includes(f));
}
function CharacterViewerComponent(_ref3) {
  let {
    collection: collectionKey
  } = _ref3;
  const collection = (0, _reactRelay.useFragment)((_CharacterViewer_collection2.default.hash && _CharacterViewer_collection2.default.hash !== "c2116c3c2c795e5306c3a2a0ecd60137" && console.error("The definition of 'CharacterViewer_collection' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CharacterViewer_collection2.default), collectionKey);
  const [fontStyle, refetchFontStyle] = (0, _reactRelay.useRefetchableFragment)((_CharacterViewer_style2.default.hash && _CharacterViewer_style2.default.hash !== "8cbf3328030930b88b3ec3601aaf7c45" && console.error("The definition of 'CharacterViewer_style' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CharacterViewer_style2.default), collection.featureStyle);
  let families = [];
  if (collection.collectionType === 'family') families = [collection];
  if (collection.collectionType === 'superfamily' && collection.children) families = collection.children;
  const unicodeData = useUnicodeData();
  let initialVariableSettings;
  if (fontStyle !== null && fontStyle !== void 0 && fontStyle.variableInstances) {
    initialVariableSettings = (0, _utils.variableInstanceSettings)(fontStyle.variableInstances[0]);
  }
  const [variableSettings, setVariableSettings] = (0, _react.useState)(initialVariableSettings);
  const [featuresChecked, setFeaturesChecked] = (0, _react.useState)(true);
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const handleStyleSelect = value => {
    const variables = {
      id: value.fontStyleId
    };
    // avoid Suspense using fetchQuery. future react may simplify this
    (0, _reactRelay.fetchQuery)(environment, _CharacterViewerStyleRefetchQuery.default, variables).subscribe({
      complete: () => {
        refetchFontStyle(variables, {
          fetchPolicy: 'store-only'
        });
        setVariableSettings(value.variableSettings);
      }
    });
  };
  let {
    style
  } = (0, _useFontStyle.default)({
    fontStyle: fontStyle
  });
  if (variableSettings) {
    style = {
      ...style,
      ...(0, _utils.variableSettingsCSS)(variableSettings)
    };
  }
  const [selectedGlyph, setSelectedGlyph] = (0, _react.useState)({
    string: 'A'
  });
  const [activeGlyph, setActiveGlyph] = (0, _react.useState)({
    string: 'A'
  });
  const visibleGlyph = activeGlyph ?? selectedGlyph;
  const getGlyphName = (0, _react.useCallback)(glyph => {
    var _fontStyle$glyphNames, _fontStyle$glyphNames2;
    return (fontStyle === null || fontStyle === void 0 ? void 0 : (_fontStyle$glyphNames = fontStyle.glyphNames) === null || _fontStyle$glyphNames === void 0 ? void 0 : (_fontStyle$glyphNames2 = _fontStyle$glyphNames.find(_ref4 => {
      let {
        characters,
        features
      } = _ref4;
      return compareGlyphs(glyph, {
        string: characters,
        features
      });
    })) === null || _fontStyle$glyphNames2 === void 0 ? void 0 : _fontStyle$glyphNames2.name) ?? null;
  }, [fontStyle, compareGlyphs]);
  const getUnicodeName = (0, _react.useCallback)(glyph => {
    return (unicodeData === null || unicodeData === void 0 ? void 0 : unicodeData[glyph.string.charCodeAt(0)]) ?? null;
  }, [unicodeData]);
  const handleMouseLeave = () => {
    if (selectedGlyph) setActiveGlyph(undefined);
  };
  const {
    ref: monitorRef,
    width: monitorWidth
  } = useSize();

  // Read the grid's resolved first-track width and expose it as --cell-width
  // on the blocks container. CSS uses --cell-width as min-height on every
  // cell so rows have consistent height regardless of cell composition —
  // without it, rows of all-wide span-N cells would collapse to glyph content
  // height. We read the grid track instead of measuring an individual cell so
  // a wide span-N cell at index 0 wouldn't poison the measurement.
  const blocksRef = (0, _react.useRef)(null);
  const [cellWidth, setCellWidth] = (0, _react.useState)(null);
  const measureCell = (0, _react.useCallback)(() => {
    var _blocksRef$current;
    const list = (_blocksRef$current = blocksRef.current) === null || _blocksRef$current === void 0 ? void 0 : _blocksRef$current.querySelector('.character-viewer__block__character-list');
    if (!list) return;
    const firstTrack = getComputedStyle(list).gridTemplateColumns.split(' ')[0];
    const width = parseFloat(firstTrack);
    if (!Number.isNaN(width)) setCellWidth(width);
  }, []);
  (0, _react.useEffect)(measureCell, [measureCell]);
  (0, _resizeObserver.default)(blocksRef, measureCell, {
    polyfill: typeof window === 'undefined' ? ServerSideResizeObserverPolyfill : null
  });
  if (!fontStyle) return null;
  const {
    ascender,
    descender,
    capHeight,
    xHeight,
    lineGap,
    unitsPerEm
  } = fontStyle.verticalMetrics;
  const halfGap = (lineGap ?? 0) / 2;
  const pos = offset => ascender + halfGap - offset;

  // Lines in priority order — lower-priority lines hidden when overlapping higher ones.
  // Threshold: ~18px (label height) converted to font units via current scale.
  const scale = monitorWidth ? monitorWidth / 2 / unitsPerEm : 0;
  const threshold = scale > 0 ? 18 / scale : 0;
  const lines = [{
    name: 'baseline',
    position: pos(0),
    label: 'Baseline',
    value: 0
  }, capHeight != null && {
    name: 'cap-height',
    position: pos(capHeight),
    label: 'Cap height',
    value: capHeight
  }, xHeight != null && {
    name: 'x-height',
    position: pos(xHeight),
    label: 'x-Height',
    value: xHeight
  }, {
    name: 'ascender',
    position: pos(ascender),
    label: 'Ascender',
    value: ascender
  }, {
    name: 'descender',
    position: pos(descender),
    label: 'Descender',
    value: descender
  }].filter(line => Boolean(line)).reduce((visible, line) =>
  // Skip this line if it's within the threshold of any already-accepted
  // higher-priority line — otherwise accept it
  threshold > 0 && visible.some(v => Math.abs(v.position - line.position) < threshold) ? visible : [...visible, line], []);
  const groups = (0, _react.useMemo)(() => {
    var _collection$glyphGrou;
    return (_collection$glyphGrou = collection.glyphGroups) === null || _collection$glyphGrou === void 0 ? void 0 : _collection$glyphGrou.map(_ref5 => {
      var _characterSets$filter;
      let {
        name,
        characterSets
      } = _ref5;
      return {
        name,
        characterSets: characterSets
        // Filter out character sets with Arabic positional features (init, medi, fina)
        // These don't render correctly in Safari due to text shaping limitations
        === null || characterSets
        // Filter out character sets with Arabic positional features (init, medi, fina)
        // These don't render correctly in Safari due to text shaping limitations
        === void 0 ? void 0 : (_characterSets$filter = characterSets
        // Filter out character sets with Arabic positional features (init, medi, fina)
        // These don't render correctly in Safari due to text shaping limitations
        .filter(charSet => !hasUnsupportedPositionalFeatures(charSet.features))) === null || _characterSets$filter === void 0 ? void 0 : _characterSets$filter.map(charSet => flattenCharacterList(charSet, fontStyle.glyphNames)).filter(_utils.notEmpty)
      };
    })
    // Filter out groups that have no character sets left after filtering
    .filter(group => group.characterSets && group.characterSets.length > 0);
  }, [collection, fontStyle]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer",
    style: {
      '--units-per-em': fontStyle.verticalMetrics.unitsPerEm,
      '--ascender': fontStyle.verticalMetrics.ascender,
      '--descender': fontStyle.verticalMetrics.descender,
      '--cap-height': fontStyle.verticalMetrics.capHeight ?? undefined,
      '--x-height': fontStyle.verticalMetrics.xHeight ?? undefined,
      '--line-gap': fontStyle.verticalMetrics.lineGap ?? undefined,
      '--monitor-width': monitorWidth
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__character-map"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor"
  }, visibleGlyph && /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor__container"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor__character-container"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor__character",
    style: {
      ...style,
      fontFeatureSettings: visibleGlyph.features && featuresChecked ? `${visibleGlyph.features.map(feature => `'${feature}' 1`).join(', ')}` : `'calt' off, 'liga' off`
    },
    ref: monitorRef
  }, /*#__PURE__*/_react.default.createElement("span", null, visibleGlyph.string)), lines.map(line => /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor__line",
    "data-name": line.name,
    key: line.name
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "character-viewer__monitor__line__label"
  }, line.label), /*#__PURE__*/_react.default.createElement("span", {
    className: "character-viewer__monitor__line__value"
  }, line.value)))), /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer__monitor__details"
  }, /*#__PURE__*/_react.default.createElement("span", null, "Font style"), /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement(_StyleSelect.default, {
    collections: families,
    onChange: handleStyleSelect,
    value: {
      fontStyleId: fontStyle.id,
      variableSettings
    }
  })), /*#__PURE__*/_react.default.createElement(GlyphMeta, {
    glyph: visibleGlyph,
    getGlyphName: getGlyphName,
    getUnicodeName: getUnicodeName,
    featuresChecked: featuresChecked,
    setFeaturesChecked: setFeaturesChecked
  })))), /*#__PURE__*/_react.default.createElement("div", {
    className: "character-viewer_blocks",
    ref: blocksRef,
    style: cellWidth ? {
      '--cell-width': `${cellWidth}px`
    } : undefined
  }, groups === null || groups === void 0 ? void 0 : groups.map((_ref6, i) => {
    let {
      name,
      characterSets
    } = _ref6;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "character-viewer__block",
      key: i,
      "data-name": name
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "character-viewer__block__name"
    }, name), /*#__PURE__*/_react.default.createElement("div", {
      className: "character-viewer__block__characters",
      style: style,
      onMouseLeave: handleMouseLeave
    }, characterSets === null || characterSets === void 0 ? void 0 : characterSets.map((set, j) => /*#__PURE__*/_react.default.createElement("div", {
      key: j,
      className: "character-viewer__block__character-list",
      style: {
        fontFeatureSettings: set.features ? `${set.features.map(feature => `'${feature}' 1`).join(', ')}` : undefined
      }
    }, set.characters.map((chars, k) => {
      const selection = {
        string: chars,
        features: set.features
      };
      return /*#__PURE__*/_react.default.createElement(Character, {
        key: k,
        onClick: () => setSelectedGlyph(selection),
        onMouseEnter: () => setActiveGlyph(selection),
        "data-selected": compareGlyphs(selectedGlyph, selection)
      }, areCombiningChars(chars) ? /*#__PURE__*/_react.default.createElement("span", null, '\u200C', chars) : chars);
    })))));
  }))));
}
const idQuery = (_CharacterViewerIDQuery2.default.hash && _CharacterViewerIDQuery2.default.hash !== "f90b09a4df6d95307b0a5d5fda487cdc" && console.error("The definition of 'CharacterViewerIDQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CharacterViewerIDQuery2.default);
function CharacterViewerIdQueryRenderer(_ref7) {
  let {
    collectionId
  } = _ref7;
  const data = (0, _reactRelay.useLazyLoadQuery)(idQuery, {
    collectionId
  });
  if (!data.node) return null;
  return /*#__PURE__*/_react.default.createElement(CharacterViewerComponent, {
    collection: data.node
  });
}
function CharacterViewerPreloadedIDQueryRenderer(_ref8) {
  let {
    preloadedQuery,
    ...rest
  } = _ref8;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(idQuery, queryRef);
  if (!data.node) return null;
  return /*#__PURE__*/_react.default.createElement(CharacterViewerComponent, _extends({}, rest, {
    collection: data.node
  }));
}
function CharacterViewerSlugQueryRenderer(_ref9) {
  var _data$viewer, _data$viewer$slug;
  let {
    collectionSlug
  } = _ref9;
  const data = (0, _reactRelay.useLazyLoadQuery)((_CharacterViewerSlugQuery2.default.hash && _CharacterViewerSlugQuery2.default.hash !== "afa08a8f050e0434308892fea6e3c267" && console.error("The definition of 'CharacterViewerSlugQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _CharacterViewerSlugQuery2.default), {
    collectionSlug
  });
  const collection = data === null || data === void 0 ? void 0 : (_data$viewer = data.viewer) === null || _data$viewer === void 0 ? void 0 : (_data$viewer$slug = _data$viewer.slug) === null || _data$viewer$slug === void 0 ? void 0 : _data$viewer$slug.fontCollection;
  if (!collection) return null;
  return /*#__PURE__*/_react.default.createElement(CharacterViewerComponent, {
    collection: collection
  });
}
function CharacterViewerPreloadedSlugQueryRenderer(_ref10) {
  var _data$viewer2, _data$viewer2$slug;
  let {
    preloadedQuery,
    ...rest
  } = _ref10;
  const queryRef = (0, _useSerializablePreloadedQuery.default)(preloadedQuery);
  const data = (0, _reactRelay.usePreloadedQuery)(idQuery, queryRef);
  const collection = data === null || data === void 0 ? void 0 : (_data$viewer2 = data.viewer) === null || _data$viewer2 === void 0 ? void 0 : (_data$viewer2$slug = _data$viewer2.slug) === null || _data$viewer2$slug === void 0 ? void 0 : _data$viewer2$slug.fontCollection;
  if (!collection) return null;
  return /*#__PURE__*/_react.default.createElement(CharacterViewerComponent, _extends({}, rest, {
    collection: collection
  }));
}
function CharacterViewer(_ref11) {
  let {
    collectionId,
    collectionSlug
  } = _ref11;
  if (collectionId) {
    return /*#__PURE__*/_react.default.createElement(CharacterViewerIdQueryRenderer, {
      collectionId: collectionId
    });
  }
  if (collectionSlug) {
    return /*#__PURE__*/_react.default.createElement(CharacterViewerSlugQueryRenderer, {
      collectionSlug: collectionSlug
    });
  }
  throw new Error('CharacterViewer expected either a collectionId or collectionSlug prop');
}
