// fontdue-js/dist/components/StripeProvider/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StripeProviderCreateCheckoutSessionMutation2 = _interopRequireDefault(require("../../__generated__/StripeProviderCreateCheckoutSessionMutation.graphql"));
var _StripeProvider_viewer2 = _interopRequireDefault(require("../../__generated__/StripeProvider_viewer.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _reactStripeJs = require("@stripe/react-stripe-js");
var _stripeJs = require("@stripe/stripe-js");
var _utils = require("../../utils");
var _ConfigContext = _interopRequireDefault(require("../ConfigContext"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const StripeProvider = _ref => {
  var _data$currentOrder, _data$settings, _data$settings$uiFont, _data$settings$uiFont2, _data$settings$uiFont3, _data$settings2, _data$currentOrder2, _data$currentOrder2$s, _data$themeConfig;
  let {
    viewer,
    providerType,
    children
  } = _ref;
  const data = (0, _reactRelay.useFragment)((_StripeProvider_viewer2.default.hash && _StripeProvider_viewer2.default.hash !== "96a569166948f6d2956bdf12dbd2820f" && console.error("The definition of 'StripeProvider_viewer' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StripeProvider_viewer2.default), viewer);
  const stripeKey = data.stripePublishableKey;
  // Use order's stripeAccountId if available (handles case where tenant changed accounts mid-checkout)
  const stripeAccount = ((_data$currentOrder = data.currentOrder) === null || _data$currentOrder === void 0 ? void 0 : _data$currentOrder.stripeAccountId) || data.stripeAccountId;
  const stripePromiseRef = (0, _react.useRef)(stripeKey && stripeAccount ? (0, _stripeJs.loadStripe)(stripeKey, {
    stripeAccount
  }) : null);

  // All hooks must be called before any conditional returns (React rules of hooks)
  const environment = (0, _reactRelay.useRelayEnvironment)();
  const {
    stripe: config
  } = (0, _react.useContext)(_ConfigContext.default);
  let customFontSource;
  const woffSrc = (_data$settings = data.settings) === null || _data$settings === void 0 ? void 0 : (_data$settings$uiFont = _data$settings.uiFontStyle) === null || _data$settings$uiFont === void 0 ? void 0 : (_data$settings$uiFont2 = _data$settings$uiFont.webfontSources) === null || _data$settings$uiFont2 === void 0 ? void 0 : (_data$settings$uiFont3 = _data$settings$uiFont2.find(source => (source === null || source === void 0 ? void 0 : source.format) === 'woff2')) === null || _data$settings$uiFont3 === void 0 ? void 0 : _data$settings$uiFont3.url;
  let fontFamily = (_data$settings2 = data.settings) !== null && _data$settings2 !== void 0 && _data$settings2.uiFontStyle ? `${data.settings.uiFontStyle.cssFamily} ${data.settings.uiFontStyle.name}` : undefined;
  // stripe validates that the strings match this pattern and throws an error
  // if they don't, and refuses to render the element. for example if a font
  // family has a special character in the name (which is actually valid).
  // so we just don't try to the fontFamily if it's not gonna be valid.
  // this regex is copied from the stripe js
  const stripeRegex = /^[#a-zA-Z0-9-_\s,"'().]*$/;
  if (fontFamily && !stripeRegex.test(fontFamily)) fontFamily = undefined;
  if (woffSrc && fontFamily) {
    customFontSource = {
      src: `url(${woffSrc})`,
      family: fontFamily,
      weight: '400'
    };
  }
  const clientSecret = ((_data$currentOrder2 = data.currentOrder) === null || _data$currentOrder2 === void 0 ? void 0 : (_data$currentOrder2$s = _data$currentOrder2.stripePaymentIntent) === null || _data$currentOrder2$s === void 0 ? void 0 : _data$currentOrder2$s.clientSecret) ?? undefined;
  const customProperties = (0, _utils.parseCustomProperties)((_data$themeConfig = data.themeConfig) === null || _data$themeConfig === void 0 ? void 0 : _data$themeConfig.customProperties);
  const fetchClientSecret = (0, _react.useCallback)(() => {
    return new Promise((resolve, reject) => {
      (0, _reactRelay.commitMutation)(environment, {
        mutation: (_StripeProviderCreateCheckoutSessionMutation2.default.hash && _StripeProviderCreateCheckoutSessionMutation2.default.hash !== "2ca9a7fcfed1292117e3cd6d0d95cca6" && console.error("The definition of 'StripeProviderCreateCheckoutSessionMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _StripeProviderCreateCheckoutSessionMutation2.default),
        variables: {},
        onCompleted: response => {
          var _response$createOrder, _response$createOrder2, _response$createOrder3;
          const clientSecret = (_response$createOrder = response.createOrderCheckoutSession) === null || _response$createOrder === void 0 ? void 0 : (_response$createOrder2 = _response$createOrder.order) === null || _response$createOrder2 === void 0 ? void 0 : (_response$createOrder3 = _response$createOrder2.stripeCheckoutSession) === null || _response$createOrder3 === void 0 ? void 0 : _response$createOrder3.clientSecret;
          if (clientSecret) {
            resolve(clientSecret);
          } else {
            reject("Didn't find client secret");
          }
        },
        onError: error => {
          reject(`Error creating checkout session: ${error.message}`);
        }
      });
    });
  }, [environment]);
  const options = {
    appearance: config.appearance ?? {
      theme: 'flat',
      variables: {
        borderRadius: '0',
        fontFamily,
        fontWeightLight: '400',
        fontWeightNormal: '400',
        fontWeightMedium: '400',
        fontWeightBold: '400',
        colorBackground: customProperties['--secondary_background_color'],
        colorText: customProperties['--primary_text_color'],
        colorDanger: customProperties['--error_color'],
        colorBackgroundText: customProperties['--primary_text_color'],
        colorSuccessText: customProperties['--success_color'],
        colorDangerText: customProperties['--error_color'],
        colorTextPlaceholder: customProperties['--secondary_text_color'],
        focusBoxShadow: `0 0 0 1px ${customProperties['--primary_text_color']}`,
        fontSizeBase: '16px',
        fontSmooth: 'auto',
        gridRowSpacing: '30px',
        spacingUnit: '3px',
        accordionItemSpacing: '20px'
      },
      rules: {
        '.Tab': {
          border: `1px solid ${customProperties['--button_border_color']}`,
          backgroundColor: customProperties['--button_background_color'],
          color: customProperties['--button_text_color']
        },
        '.Tab:hover': {
          border: `1px solid ${customProperties['--button_hover_border_color']}`,
          backgroundColor: customProperties['--button_hover_background_color'],
          color: customProperties['--button_hover_text_color']
        },
        '.Tab--selected,.Tab--selected:hover': {
          border: `1px solid ${customProperties['--button_selected_border_color']}`,
          backgroundColor: customProperties['--button_selected_background_color'],
          color: customProperties['--button_selected_text_color']
        },
        '.TabIcon': {
          fill: customProperties['--button_text_color']
        },
        '.TabIcon--selected': {
          fill: customProperties['--button_selected_text_color']
        },
        '.TabIcon:hover,.TabIcon--selected:hover': {
          fill: customProperties['--button_hover_text_color']
        },
        '.AccordionItem': {
          border: `1px solid ${customProperties['--button_border_color']}`,
          backgroundColor: customProperties['--button_background_color'],
          color: customProperties['--button_text_color']
        },
        '.AccordionItem:hover': {
          border: `1px solid ${customProperties['--button_hover_border_color']}`,
          backgroundColor: customProperties['--button_hover_background_color'],
          color: customProperties['--button_hover_text_color']
        },
        '.AccordionItem--selected': {
          border: `1px solid ${customProperties['--button_selected_border_color']}`,
          backgroundColor: customProperties['--button_selected_background_color'],
          color: customProperties['--button_selected_text_color']
        },
        '.Label': {
          color: customProperties['--secondary_text_color'],
          fontSize: '1em'
        },
        '.Input': {
          marginTop: '3px',
          fontSize: '20px'
        },
        '.Input--invalid': {
          boxShadow: 'none'
        }
      }
    },
    fonts: customFontSource ? [customFontSource] : undefined,
    clientSecret,
    syncAddressCheckbox: 'none'
  };
  if (providerType === 'checkout') {
    return /*#__PURE__*/_react.default.createElement(_reactStripeJs.EmbeddedCheckoutProvider, {
      stripe: stripePromiseRef.current,
      options: {
        fetchClientSecret
      }
    }, children);
  }
  if (providerType === 'elements') {
    return /*#__PURE__*/_react.default.createElement(_reactStripeJs.Elements, {
      stripe: stripePromiseRef.current,
      options: options
    }, children);
  }
  return null;
};
var _default = StripeProvider;
exports.default = _default;
