// fontdue-js/dist/components/NodePasswordForm/index.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NodePasswordForm;
var _NodePasswordFormSlugQuery2 = _interopRequireDefault(require("../../__generated__/NodePasswordFormSlugQuery.graphql"));
var _NodePasswordFormIDQuery2 = _interopRequireDefault(require("../../__generated__/NodePasswordFormIDQuery.graphql"));
var _NodePasswordFormAccessNodeMutation2 = _interopRequireDefault(require("../../__generated__/NodePasswordFormAccessNodeMutation.graphql"));
var _react = _interopRequireWildcard(require("react"));
var _reactRelay = require("react-relay");
var _TextField = _interopRequireDefault(require("../TextField"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const accessNodeMutation = (_NodePasswordFormAccessNodeMutation2.default.hash && _NodePasswordFormAccessNodeMutation2.default.hash !== "d0beeb729c358ef4eebd1c8be47ec68e" && console.error("The definition of 'NodePasswordFormAccessNodeMutation' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _NodePasswordFormAccessNodeMutation2.default);
function NodePasswordFormComponent(props) {
  const [password, setPassword] = (0, _react.useState)('');
  const [error, setError] = (0, _react.useState)(null);
  const [commitMutation, submitting] = (0, _reactRelay.useMutation)(accessNodeMutation);
  const disabled = submitting || !password;
  const handleSubmit = e => {
    e.preventDefault();
    setError(null);
    commitMutation({
      variables: {
        input: {
          password,
          id: props.collectionId,
          slug: props.collectionSlug
        }
      },
      onCompleted: res => {
        var _res$accessNode;
        if ((_res$accessNode = res.accessNode) !== null && _res$accessNode !== void 0 && _res$accessNode.success) {
          location.reload();
        }
      },
      onError: error => {
        setError(error.message);
      }
    });
  };
  return /*#__PURE__*/_react.default.createElement("form", {
    onSubmit: handleSubmit,
    className: "node-password-form"
  }, error && /*#__PURE__*/_react.default.createElement("p", {
    className: "node-password-form__error"
  }, error), /*#__PURE__*/_react.default.createElement("div", {
    className: "node-password-form__section"
  }, /*#__PURE__*/_react.default.createElement(_TextField.default, {
    label: "Password *",
    placeholder: "Enter password",
    autocomplete: "off",
    type: "password",
    value: password,
    onChange: value => setPassword(value)
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "node-password-form__section"
  }, /*#__PURE__*/_react.default.createElement("button", {
    disabled: disabled,
    className: "submit-button"
  }, "Submit")));
}
const idQuery = (_NodePasswordFormIDQuery2.default.hash && _NodePasswordFormIDQuery2.default.hash !== "2af86f8f6df3d2e6061ea1c9dd777a26" && console.error("The definition of 'NodePasswordFormIDQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _NodePasswordFormIDQuery2.default);
function NodePasswordFormIDQueryRenderer(_ref) {
  let {
    collectionId,
    ...rest
  } = _ref;
  const data = (0, _reactRelay.useLazyLoadQuery)(idQuery, {
    collectionId
  });
  // this is counterintuitive, but we've added the @catch directive on the collection,
  // and we're expecting it to respond with an error if the node is inaccessible due to being
  // password-protected. so, when we encounter the error we're safe to render the password form.
  // (we don't want to show the form if the user has already gained access)
  if (data.collection.ok === false) return /*#__PURE__*/_react.default.createElement(NodePasswordFormComponent, _extends({}, rest, {
    collectionId: collectionId
  }));
}
const slugQuery = (_NodePasswordFormSlugQuery2.default.hash && _NodePasswordFormSlugQuery2.default.hash !== "c40faf067007027fa9ba6673350c7202" && console.error("The definition of 'NodePasswordFormSlugQuery' appears to have changed. Run `relay-compiler` to update the generated files to receive the expected data."), _NodePasswordFormSlugQuery2.default);
function NodePasswordFormSlugQueryRenderer(_ref2) {
  var _data$viewer$slug;
  let {
    collectionSlug,
    ...rest
  } = _ref2;
  const data = (0, _reactRelay.useLazyLoadQuery)(slugQuery, {
    collectionSlug
  });
  // see note in NodePasswordFormIDQueryRenderer
  if (((_data$viewer$slug = data.viewer.slug) === null || _data$viewer$slug === void 0 ? void 0 : _data$viewer$slug.collection.ok) === false) return /*#__PURE__*/_react.default.createElement(NodePasswordFormComponent, _extends({}, rest, {
    collectionSlug: collectionSlug
  }));
  return null;
}

// default client export
function NodePasswordForm(_ref3) {
  let {
    collectionId,
    collectionSlug,
    ...rest
  } = _ref3;
  if (collectionId) {
    return /*#__PURE__*/_react.default.createElement(NodePasswordFormIDQueryRenderer, _extends({
      collectionId: collectionId
    }, rest));
  }
  if (collectionSlug) {
    return /*#__PURE__*/_react.default.createElement(NodePasswordFormSlugQueryRenderer, _extends({
      collectionSlug: collectionSlug
    }, rest));
  }
  throw new Error('NodePasswordForm expected either a collectionId or collectionSlug prop');
}
